// @abp/ng.feature-management/esm2015/lib/directives/free-text-input.directive.js
import { Directive, Input, HostBinding } from '@angular/core';
export const INPUT_TYPES = {
    numeric: 'number',
    default: 'text',
};
export class FreeTextInputDirective {
    set feature(val) {
        this._feature = val;
        this.setInputType();
    }
    get feature() {
        return this._feature;
    }
    setInputType() {
        var _a, _b, _c, _d;
        const validatorType = (_c = (_b = (_a = this.feature) === null || _a === void 0 ? void 0 : _a.valueType) === null || _b === void 0 ? void 0 : _b.validator) === null || _c === void 0 ? void 0 : _c.name.toLowerCase();
        this.type = (_d = INPUT_TYPES[validatorType]) !== null && _d !== void 0 ? _d : INPUT_TYPES.default;
    }
}
FreeTextInputDirective.decorators = [
    { type: Directive, args: [{
                selector: 'input[abpFeatureManagementFreeText]',
                exportAs: 'inputAbpFeatureManagementFreeText',
            },] }
];
FreeTextInputDirective.propDecorators = {
    feature: [{ type: Input, args: ['abpFeatureManagementFreeText',] }],
    type: [{ type: HostBinding, args: ['type',] }]
};
