// @abp/ng.core/esm2015/testing/lib/services/mock-rest.service.js
import { CORE_OPTIONS, EnvironmentService, RestService } from '@abp/ng.core';
import { HttpClient } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@abp/ng.core";
import * as i2 from "@angular/common/http";
export class MockRestService extends RestService {
    constructor(options, http, environment) {
        super(options, http, environment, null);
        this.options = options;
        this.http = http;
        this.environment = environment;
    }
    handleError(err) {
        return throwError(err);
    }
}
MockRestService.ɵprov = i0.ɵɵdefineInjectable({ factory: function MockRestService_Factory() { return new MockRestService(i0.ɵɵinject(i1.CORE_OPTIONS), i0.ɵɵinject(i2.HttpClient), i0.ɵɵinject(i1.EnvironmentService)); }, token: MockRestService, providedIn: "root" });
MockRestService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
MockRestService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CORE_OPTIONS,] }] },
    { type: HttpClient },
    { type: EnvironmentService }
];
