// @abp/ng.account/esm2015/lib/components/reset-password/reset-password.component.js
import { getPasswordValidators } from '@abp/ng.theme.shared';
import { Component, Injector } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { comparePasswords } from '@ngx-validate/core';
import { finalize } from 'rxjs/operators';
import { AccountService } from '../../proxy/account/account.service';
const PASSWORD_FIELDS = ['password', 'confirmPassword'];
export class ResetPasswordComponent {
    constructor(fb, accountService, route, router, injector) {
        this.fb = fb;
        this.accountService = accountService;
        this.route = route;
        this.router = router;
        this.injector = injector;
        this.inProgress = false;
        this.isPasswordReset = false;
        this.mapErrorsFn = (errors, groupErrors, control) => {
            if (PASSWORD_FIELDS.indexOf(String(control.name)) < 0)
                return errors;
            return errors.concat(groupErrors.filter(({ key }) => key === 'passwordMismatch'));
        };
    }
    ngOnInit() {
        this.route.queryParams.subscribe(({ userId, resetToken }) => {
            if (!userId || !resetToken)
                this.router.navigateByUrl('/account/login');
            this.form = this.fb.group({
                userId: [userId, [Validators.required]],
                resetToken: [resetToken, [Validators.required]],
                password: ['', [Validators.required, ...getPasswordValidators(this.injector)]],
                confirmPassword: ['', [Validators.required, ...getPasswordValidators(this.injector)]],
            }, {
                validators: [comparePasswords(PASSWORD_FIELDS)],
            });
        });
    }
    onSubmit() {
        if (this.form.invalid || this.inProgress)
            return;
        this.inProgress = true;
        this.accountService
            .resetPassword({
            userId: this.form.get('userId').value,
            resetToken: this.form.get('resetToken').value,
            password: this.form.get('password').value,
        })
            .pipe(finalize(() => (this.inProgress = false)))
            .subscribe(() => {
            this.isPasswordReset = true;
        });
    }
}
ResetPasswordComponent.decorators = [
    { type: Component, args: [{
                selector: 'abp-reset-password',
                template: "<h4>{{ 'AbpAccount::ResetPassword' | abpLocalization }}</h4>\n\n<form\n  *ngIf=\"!isPasswordReset; else passwordResetTemplate\"\n  [formGroup]=\"form\"\n  [mapErrorsFn]=\"mapErrorsFn\"\n  (ngSubmit)=\"onSubmit()\"\n  validateOnSubmit\n>\n  <p>{{ 'AbpAccount::ResetPassword_Information' | abpLocalization }}</p>\n  <div class=\"form-group\">\n    <label for=\"input-password\">{{ 'AbpAccount::Password' | abpLocalization }}</label\n    ><span> * </span>\n    <input type=\"password\" id=\"input-password\" class=\"form-control\" formControlName=\"password\" />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"input-confirm-password\">{{ 'AbpAccount::ConfirmPassword' | abpLocalization }}</label\n    ><span> * </span>\n    <input\n      type=\"password\"\n      id=\"input-confirm-password\"\n      class=\"form-control\"\n      formControlName=\"confirmPassword\"\n    />\n  </div>\n  <button class=\"mr-2 btn btn-secondary\" type=\"button\" routerLink=\"/account/login\">\n    {{ 'AbpAccount::Cancel' | abpLocalization }}\n  </button>\n  <abp-button\n    buttonType=\"submit\"\n    buttonClass=\"mr-2 btn btn-primary\"\n    [loading]=\"inProgress\"\n    (click)=\"onSubmit()\"\n  >\n    {{ 'AbpAccount::Submit' | abpLocalization }}\n  </abp-button>\n</form>\n\n<ng-template #passwordResetTemplate>\n  <p>\n    {{ 'AbpAccount::YourPasswordIsSuccessfullyReset' | abpLocalization }}\n  </p>\n\n  <a routerLink=\"/account/login\">\n    <button class=\"d-block mt-2 mb-3 btn btn-primary\">\n      {{ 'AbpAccount::BackToLogin' | abpLocalization }}\n    </button>\n  </a>\n</ng-template>\n"
            },] }
];
ResetPasswordComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: AccountService },
    { type: ActivatedRoute },
    { type: Router },
    { type: Injector }
];
