// @abp/ng.tenant-management/esm2015/lib/defaults/default-tenants-entity-actions.js
import { EntityAction } from '@abp/ng.theme.shared/extensions';
import { TenantsComponent } from '../components/tenants/tenants.component';
const ɵ0 = data => {
    const component = data.getInjected(TenantsComponent);
    component.editTenant(data.record.id);
}, ɵ1 = data => {
    const component = data.getInjected(TenantsComponent);
    component.openFeaturesModal(data.record.id);
}, ɵ2 = data => {
    const component = data.getInjected(TenantsComponent);
    component.delete(data.record.id, data.record.name);
};
export const DEFAULT_TENANTS_ENTITY_ACTIONS = EntityAction.createMany([
    {
        text: 'AbpTenantManagement::Edit',
        action: ɵ0,
        permission: 'AbpTenantManagement.Tenants.Update',
    },
    {
        text: 'AbpTenantManagement::Permission:ManageFeatures',
        action: ɵ1,
        permission: 'AbpTenantManagement.Tenants.ManageFeatures',
    },
    {
        text: 'AbpTenantManagement::Delete',
        action: ɵ2,
        permission: 'AbpTenantManagement.Tenants.Delete',
    },
]);
export { ɵ0, ɵ1, ɵ2 };
