// @abp/ng.tenant-management/esm2015/lib/defaults/default-tenants-form-props.js
import { getPasswordValidators } from '@abp/ng.theme.shared';
import { FormProp } from '@abp/ng.theme.shared/extensions';
import { Validators } from '@angular/forms';
const ɵ0 = () => [Validators.required, Validators.maxLength(256)], ɵ1 = () => [Validators.required, Validators.maxLength(256), Validators.email], ɵ2 = data => [Validators.required, ...getPasswordValidators({ get: data.getInjected })];
export const DEFAULT_TENANTS_CREATE_FORM_PROPS = FormProp.createMany([
    {
        type: "string" /* String */,
        name: 'name',
        id: 'name',
        displayName: 'AbpTenantManagement::TenantName',
        validators: ɵ0,
    },
    {
        type: "email" /* Email */,
        name: 'adminEmailAddress',
        displayName: 'AbpTenantManagement::DisplayName:AdminEmailAddress',
        id: 'admin-email-address',
        validators: ɵ1,
    },
    {
        type: "password" /* Password */,
        name: 'adminPassword',
        displayName: 'AbpTenantManagement::DisplayName:AdminPassword',
        id: 'admin-password',
        autocomplete: 'new-password',
        validators: ɵ2,
    },
]);
export const DEFAULT_TENANTS_EDIT_FORM_PROPS = DEFAULT_TENANTS_CREATE_FORM_PROPS.slice(0, 1);
export { ɵ0, ɵ1, ɵ2 };
