// @abp/ng.theme.basic/fesm2015/abp-ng.theme.basic-testing.js
import { BASIC_THEME_NAV_ITEM_PROVIDERS, BASIC_THEME_STYLES_PROVIDERS, ValidationErrorComponent, BaseThemeBasicModule } from '@abp/ng.theme.basic';
import { NgModule } from '@angular/core';
import { VALIDATION_ERROR_TEMPLATE, VALIDATION_TARGET_SELECTOR, VALIDATION_INVALID_CLASSES } from '@ngx-validate/core';

class ThemeBasicTestingModule {
    static withConfig() {
        return {
            ngModule: ThemeBasicTestingModule,
            providers: [
                BASIC_THEME_NAV_ITEM_PROVIDERS,
                BASIC_THEME_STYLES_PROVIDERS,
                {
                    provide: VALIDATION_ERROR_TEMPLATE,
                    useValue: ValidationErrorComponent,
                },
                {
                    provide: VALIDATION_TARGET_SELECTOR,
                    useValue: '.form-group',
                },
                {
                    provide: VALIDATION_INVALID_CLASSES,
                    useValue: 'is-invalid',
                },
            ],
        };
    }
}
ThemeBasicTestingModule.decorators = [
    { type: NgModule, args: [{
                exports: [BaseThemeBasicModule],
                imports: [BaseThemeBasicModule],
            },] }
];

/**
 * Generated bundle index. Do not edit.
 */

export { ThemeBasicTestingModule };
