// @abp/ng.core/esm2015/lib/handlers/oauth-configuration.handler.js
import { Inject, Injectable } from '@angular/core';
import { OAuthService } from 'angular-oauth2-oidc';
import compare from 'just-compare';
import { filter, map } from 'rxjs/operators';
import { EnvironmentService } from '../services/environment.service';
import { CORE_OPTIONS } from '../tokens/options.token';
import * as i0 from "@angular/core";
import * as i1 from "angular-oauth2-oidc";
import * as i2 from "../services/environment.service";
import * as i3 from "../tokens/options.token";
export class OAuthConfigurationHandler {
    constructor(oAuthService, environmentService, options) {
        this.oAuthService = oAuthService;
        this.environmentService = environmentService;
        this.options = options;
        this.listenToSetEnvironment();
    }
    listenToSetEnvironment() {
        this.environmentService
            .createOnUpdateStream(state => state)
            .pipe(map(environment => environment.oAuthConfig), filter(config => !compare(config, this.options.environment.oAuthConfig)))
            .subscribe(config => {
            this.oAuthService.configure(config);
        });
    }
}
OAuthConfigurationHandler.ɵprov = i0.ɵɵdefineInjectable({ factory: function OAuthConfigurationHandler_Factory() { return new OAuthConfigurationHandler(i0.ɵɵinject(i1.OAuthService), i0.ɵɵinject(i2.EnvironmentService), i0.ɵɵinject(i3.CORE_OPTIONS)); }, token: OAuthConfigurationHandler, providedIn: "root" });
OAuthConfigurationHandler.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
OAuthConfigurationHandler.ctorParameters = () => [
    { type: OAuthService },
    { type: EnvironmentService },
    { type: undefined, decorators: [{ type: Inject, args: [CORE_OPTIONS,] }] }
];
