// @abp/ng.theme.shared/fesm2015/abp-ng.theme.shared-testing.js
import { THEME_SHARED_ROUTE_PROVIDERS, DateParserFormatter, DEFAULT_VALIDATION_BLUEPRINTS, BaseThemeSharedModule } from '@abp/ng.theme.shared';
import { NgModule } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbDateParserFormatter } from '@ng-bootstrap/ng-bootstrap';
import { VALIDATION_BLUEPRINTS, VALIDATION_MAP_ERRORS_FN, defaultMapErrorsFn, VALIDATION_VALIDATE_ON_SUBMIT } from '@ngx-validate/core';

var index = /*#__PURE__*/Object.freeze({
    __proto__: null
});

/**
 * ThemeSharedTestingModule is the module that will be used in tests
 */
class ThemeSharedTestingModule {
    static withConfig({ validation = {} } = {}) {
        return {
            ngModule: ThemeSharedTestingModule,
            providers: [
                THEME_SHARED_ROUTE_PROVIDERS,
                { provide: NgbDateParserFormatter, useClass: DateParserFormatter },
                {
                    provide: VALIDATION_BLUEPRINTS,
                    useValue: Object.assign(Object.assign({}, DEFAULT_VALIDATION_BLUEPRINTS), (validation.blueprints || {})),
                },
                {
                    provide: VALIDATION_MAP_ERRORS_FN,
                    useValue: validation.mapErrorsFn || defaultMapErrorsFn,
                },
                {
                    provide: VALIDATION_VALIDATE_ON_SUBMIT,
                    useValue: validation.validateOnSubmit,
                },
            ],
        };
    }
}
ThemeSharedTestingModule.decorators = [
    { type: NgModule, args: [{
                exports: [RouterTestingModule, BaseThemeSharedModule],
                imports: [RouterTestingModule, BaseThemeSharedModule],
            },] }
];

/**
 * Generated bundle index. Do not edit.
 */

export { index as ThemeSharedTesting, ThemeSharedTestingModule };
