// @abp/ng.core/esm2015/lib/proxy/pages/abp/multi-tenancy/abp-tenant.service.js
import { Injectable } from '@angular/core';
import { RestService } from '../../../../services/rest.service';
import * as i0 from "@angular/core";
import * as i1 from "../../../../services/rest.service";
export class AbpTenantService {
    constructor(restService) {
        this.restService = restService;
        this.apiName = 'abp';
        this.findTenantById = (id, headers) => this.restService.request({
            method: 'GET',
            url: `/api/abp/multi-tenancy/tenants/by-id/${id}`,
            headers,
        }, { apiName: this.apiName });
        this.findTenantByName = (name, headers) => this.restService.request({
            method: 'GET',
            url: `/api/abp/multi-tenancy/tenants/by-name/${name}`,
            headers,
        }, { apiName: this.apiName });
    }
}
AbpTenantService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AbpTenantService_Factory() { return new AbpTenantService(i0.ɵɵinject(i1.RestService)); }, token: AbpTenantService, providedIn: "root" });
AbpTenantService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AbpTenantService.ctorParameters = () => [
    { type: RestService }
];
