// @abp/ng.account/esm2015/lib/components/login/login.component.js
import { ConfigStateService, AuthService } from '@abp/ng.core';
import { ToasterService } from '@abp/ng.theme.shared';
import { Component, Injector } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import { throwError } from 'rxjs';
import { catchError, finalize } from 'rxjs/operators';
import snq from 'snq';
import { getRedirectUrl } from '../../utils/auth-utils';
const { maxLength, required } = Validators;
export class LoginComponent {
    constructor(injector, fb, toasterService, authService, configState) {
        this.injector = injector;
        this.fb = fb;
        this.toasterService = toasterService;
        this.authService = authService;
        this.configState = configState;
        this.isSelfRegistrationEnabled = true;
        this.authWrapperKey = "Account.AuthWrapperComponent" /* AuthWrapper */;
    }
    ngOnInit() {
        this.init();
        this.buildForm();
    }
    init() {
        this.isSelfRegistrationEnabled =
            (this.configState.getSetting('Abp.Account.IsSelfRegistrationEnabled') || '').toLowerCase() !== 'false';
    }
    buildForm() {
        this.form = this.fb.group({
            username: ['', [required, maxLength(255)]],
            password: ['', [required, maxLength(128)]],
            rememberMe: [false],
        });
    }
    onSubmit() {
        if (this.form.invalid)
            return;
        this.inProgress = true;
        const { username, password, rememberMe } = this.form.value;
        const redirectUrl = getRedirectUrl(this.injector);
        this.authService
            .login({ username, password, rememberMe, redirectUrl })
            .pipe(catchError(err => {
            this.toasterService.error(snq(() => err.error.error_description) ||
                snq(() => err.error.error.message, 'AbpAccount::DefaultErrorMessage'), 'Error', { life: 7000 });
            return throwError(err);
        }), finalize(() => (this.inProgress = false)))
            .subscribe();
    }
}
LoginComponent.decorators = [
    { type: Component, args: [{
                selector: 'abp-login',
                template: "<h4>{{ 'AbpAccount::Login' | abpLocalization }}</h4>\n<strong *ngIf=\"isSelfRegistrationEnabled\">\n  {{ 'AbpAccount::AreYouANewUser' | abpLocalization }}\n  <a class=\"text-decoration-none\" routerLink=\"/account/register\" queryParamsHandling=\"preserve\">{{\n    'AbpAccount::Register' | abpLocalization\n  }}</a>\n</strong>\n<form [formGroup]=\"form\" (ngSubmit)=\"onSubmit()\" validateOnSubmit class=\"mt-4\">\n  <div class=\"form-group\">\n    <label for=\"login-input-user-name-or-email-address\">{{\n      'AbpAccount::UserNameOrEmailAddress' | abpLocalization\n    }}</label>\n    <input\n      class=\"form-control\"\n      type=\"text\"\n      id=\"login-input-user-name-or-email-address\"\n      formControlName=\"username\"\n      autocomplete=\"username\"\n      autofocus\n    />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"login-input-password\">{{ 'AbpAccount::Password' | abpLocalization }}</label>\n    <input\n      class=\"form-control\"\n      type=\"password\"\n      id=\"login-input-password\"\n      formControlName=\"password\"\n      autocomplete=\"current-password\"\n    />\n  </div>\n\n  <div class=\"row\">\n    <div class=\"col\">\n      <label class=\"custom-checkbox custom-control mb-2\" for=\"login-input-remember-me\">\n        <input\n          class=\"form-check-input\"\n          type=\"checkbox\"\n          id=\"login-input-remember-me\"\n          formControlName=\"rememberMe\"\n        />\n        {{ 'AbpAccount::RememberMe' | abpLocalization }}\n      </label>\n    </div>\n    <div class=\"text-right col\">\n      <a routerLink=\"/account/forgot-password\">{{\n        'AbpAccount::ForgotPassword' | abpLocalization\n      }}</a>\n    </div>\n  </div>\n\n  <abp-button\n    [loading]=\"inProgress\"\n    buttonType=\"submit\"\n    name=\"Action\"\n    buttonClass=\"btn-block btn-lg mt-3 btn btn-primary\"\n  >\n    {{ 'AbpAccount::Login' | abpLocalization }}\n  </abp-button>\n</form>\n"
            },] }
];
LoginComponent.ctorParameters = () => [
    { type: Injector },
    { type: FormBuilder },
    { type: ToasterService },
    { type: AuthService },
    { type: ConfigStateService }
];
