// @abp/ng.permission-management/esm2015/lib/states/permission-management.state.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Action, Selector, State } from '@ngxs/store';
import { tap } from 'rxjs/operators';
import { GetPermissions, UpdatePermissions } from '../actions/permission-management.actions';
import { PermissionsService } from '../proxy/permissions.service';
let PermissionManagementState = class PermissionManagementState {
    constructor(service) {
        this.service = service;
    }
    static getPermissionGroups({ permissionRes }) {
        return permissionRes.groups || [];
    }
    static getEntityDisplayName({ permissionRes }) {
        return permissionRes.entityDisplayName;
    }
    permissionManagementGet({ patchState }, { payload: { providerKey, providerName } = {} }) {
        return this.service.get(providerName, providerKey).pipe(tap(permissionResponse => patchState({
            permissionRes: permissionResponse,
        })));
    }
    permissionManagementUpdate(_, { payload }) {
        return this.service.update(payload.providerName, payload.providerKey, {
            permissions: payload.permissions,
        });
    }
};
PermissionManagementState.decorators = [
    { type: Injectable }
];
PermissionManagementState.ctorParameters = () => [
    { type: PermissionsService }
];
__decorate([
    Action(GetPermissions)
], PermissionManagementState.prototype, "permissionManagementGet", null);
__decorate([
    Action(UpdatePermissions)
], PermissionManagementState.prototype, "permissionManagementUpdate", null);
__decorate([
    Selector()
], PermissionManagementState, "getPermissionGroups", null);
__decorate([
    Selector()
], PermissionManagementState, "getEntityDisplayName", null);
PermissionManagementState = __decorate([
    State({
        name: 'PermissionManagementState',
        defaults: { permissionRes: {} },
    })
], PermissionManagementState);
export { PermissionManagementState };
