// @abp/ng.tenant-management/__ivy_ngcc__/fesm2015/abp-ng.tenant-management-config.js
import { RoutesService } from '@abp/ng.core';
import { APP_INITIALIZER, NgModule } from '@angular/core';

import * as ɵngcc0 from '@angular/core';
const TENANT_MANAGEMENT_ROUTE_PROVIDERS = [
    { provide: APP_INITIALIZER, useFactory: configureRoutes, deps: [RoutesService], multi: true },
];
function configureRoutes(routes) {
    return () => {
        routes.add([
            {
                path: undefined,
                name: "AbpTenantManagement::Menu:TenantManagement" /* TenantManagement */,
                parentName: "AbpUiNavigation::Menu:Administration" /* Administration */,
                requiredPolicy: "AbpTenantManagement.Tenants" /* TenantManagement */,
                layout: "application" /* application */,
                iconClass: 'fa fa-users',
                order: 2,
            },
            {
                path: '/tenant-management/tenants',
                name: "AbpTenantManagement::Tenants" /* Tenants */,
                parentName: "AbpTenantManagement::Menu:TenantManagement" /* TenantManagement */,
                requiredPolicy: "AbpTenantManagement.Tenants" /* Tenants */,
                order: 1,
            },
        ]);
    };
}

class TenantManagementConfigModule {
    static forRoot() {
        return {
            ngModule: TenantManagementConfigModule,
            providers: [TENANT_MANAGEMENT_ROUTE_PROVIDERS],
        };
    }
}
TenantManagementConfigModule.ɵfac = function TenantManagementConfigModule_Factory(t) { return new (t || TenantManagementConfigModule)(); };
TenantManagementConfigModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: TenantManagementConfigModule });
TenantManagementConfigModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TenantManagementConfigModule, [{
        type: NgModule
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { TENANT_MANAGEMENT_ROUTE_PROVIDERS, TenantManagementConfigModule, configureRoutes, TENANT_MANAGEMENT_ROUTE_PROVIDERS as ɵa, configureRoutes as ɵb };
