// @abp/ng.identity/esm2015/config/providers/route.provider.js
import { RoutesService } from '@abp/ng.core';
import { APP_INITIALIZER } from '@angular/core';
export const IDENTITY_ROUTE_PROVIDERS = [
    { provide: APP_INITIALIZER, useFactory: configureRoutes, deps: [RoutesService], multi: true },
];
export function configureRoutes(routesService) {
    return () => {
        routesService.add([
            {
                path: undefined,
                name: "AbpIdentity::Menu:IdentityManagement" /* IdentityManagement */,
                parentName: "AbpUiNavigation::Menu:Administration" /* Administration */,
                requiredPolicy: "AbpIdentity.Roles || AbpIdentity.Users" /* IdentityManagement */,
                iconClass: 'fa fa-id-card-o',
                layout: "application" /* application */,
                order: 1,
            },
            {
                path: '/identity/roles',
                name: "AbpIdentity::Roles" /* Roles */,
                parentName: "AbpIdentity::Menu:IdentityManagement" /* IdentityManagement */,
                requiredPolicy: "AbpIdentity.Roles" /* Roles */,
                order: 1,
            },
            {
                path: '/identity/users',
                name: "AbpIdentity::Users" /* Users */,
                parentName: "AbpIdentity::Menu:IdentityManagement" /* IdentityManagement */,
                requiredPolicy: "AbpIdentity.Users" /* Users */,
                order: 2,
            },
        ]);
    };
}
