// @abp/ng.account/bundles/abp-ng.account-config.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@abp/ng.core'), require('@angular/router')) :
    typeof define === 'function' && define.amd ? define('@abp/ng.account/config', ['exports', '@angular/core', '@abp/ng.core', '@angular/router'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.abp = global.abp || {}, global.abp.ng = global.abp.ng || {}, global.abp.ng.account = global.abp.ng.account || {}, global.abp.ng.account.config = {}), global.ng.core, global.ng_core, global.ng.router));
}(this, (function (exports, core, ng_core, router) { 'use strict';

    var ACCOUNT_ROUTE_PROVIDERS = [
        { provide: core.APP_INITIALIZER, useFactory: configureRoutes, deps: [ng_core.RoutesService], multi: true },
    ];
    function configureRoutes(routes) {
        return function () {
            routes.add([
                {
                    path: undefined,
                    name: "AbpAccount::Menu:Account" /* Account */,
                    invisible: true,
                    layout: "account" /* account */,
                    order: 1,
                },
                {
                    path: '/account/login',
                    name: "AbpAccount::Login" /* Login */,
                    parentName: "AbpAccount::Menu:Account" /* Account */,
                    order: 1,
                },
                {
                    path: '/account/register',
                    name: "AbpAccount::Register" /* Register */,
                    parentName: "AbpAccount::Menu:Account" /* Account */,
                    order: 2,
                },
                {
                    path: '/account/manage',
                    name: "AbpAccount::MyAccount" /* ManageProfile */,
                    parentName: "AbpAccount::Menu:Account" /* Account */,
                    layout: "application" /* application */,
                    order: 3,
                },
                {
                    path: '/account/forgot-password',
                    parentName: "AbpAccount::Menu:Account" /* Account */,
                    name: "AbpAccount::ForgotPassword" /* ForgotPassword */,
                    invisible: true,
                },
                {
                    path: '/account/reset-password',
                    parentName: "AbpAccount::Menu:Account" /* Account */,
                    name: "AbpAccount::ResetPassword" /* ResetPassword */,
                    invisible: true,
                },
            ]);
        };
    }

    function navigateToManageProfileFactory(injector) {
        return function () {
            var router$1 = injector.get(router.Router);
            var routes = injector.get(ng_core.RoutesService);
            var path = routes.find(function (item) { return item.name === "AbpAccount::MyAccount"; } /* ManageProfile */).path;
            router$1.navigateByUrl(path);
        };
    }

    var AccountConfigModule = /** @class */ (function () {
        function AccountConfigModule() {
        }
        AccountConfigModule.forRoot = function () {
            return {
                ngModule: AccountConfigModule,
                providers: [
                    ACCOUNT_ROUTE_PROVIDERS,
                    {
                        provide: ng_core.NAVIGATE_TO_MANAGE_PROFILE,
                        useFactory: navigateToManageProfileFactory,
                        deps: [core.Injector],
                    },
                ],
            };
        };
        return AccountConfigModule;
    }());
    AccountConfigModule.decorators = [
        { type: core.NgModule }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ACCOUNT_ROUTE_PROVIDERS = ACCOUNT_ROUTE_PROVIDERS;
    exports.AccountConfigModule = AccountConfigModule;
    exports.configureRoutes = configureRoutes;
    exports.ɵa = ACCOUNT_ROUTE_PROVIDERS;
    exports.ɵb = configureRoutes;
    exports.ɵc = navigateToManageProfileFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
