// @abp/ng.identity/esm2015/lib/defaults/default-roles-entity-actions.js
import { EntityAction } from '@abp/ng.theme.shared/extensions';
import { RolesComponent } from '../components/roles/roles.component';
const ɵ0 = data => {
    const component = data.getInjected(RolesComponent);
    component.edit(data.record.id);
}, ɵ1 = data => {
    const component = data.getInjected(RolesComponent);
    component.openPermissionsModal(data.record.name);
}, ɵ2 = data => {
    const component = data.getInjected(RolesComponent);
    component.delete(data.record.id, data.record.name);
}, ɵ3 = data => !data.record.isStatic;
export const DEFAULT_ROLES_ENTITY_ACTIONS = EntityAction.createMany([
    {
        text: 'AbpIdentity::Edit',
        action: ɵ0,
        permission: 'AbpIdentity.Roles.Update',
    },
    {
        text: 'AbpIdentity::Permissions',
        action: ɵ1,
        permission: 'AbpIdentity.Roles.ManagePermissions',
    },
    {
        text: 'AbpIdentity::Delete',
        action: ɵ2,
        permission: 'AbpIdentity.Roles.Delete',
        visible: ɵ3,
    },
]);
export { ɵ0, ɵ1, ɵ2, ɵ3 };
