// @abp/ng.account/fesm2015/abp-ng.account-config.js
import { APP_INITIALIZER, Injector, NgModule } from '@angular/core';
import { RoutesService, NAVIGATE_TO_MANAGE_PROFILE } from '@abp/ng.core';
import { Router } from '@angular/router';

const ACCOUNT_ROUTE_PROVIDERS = [
    { provide: APP_INITIALIZER, useFactory: configureRoutes, deps: [RoutesService], multi: true },
];
function configureRoutes(routes) {
    return () => {
        routes.add([
            {
                path: undefined,
                name: "AbpAccount::Menu:Account" /* Account */,
                invisible: true,
                layout: "account" /* account */,
                order: 1,
            },
            {
                path: '/account/login',
                name: "AbpAccount::Login" /* Login */,
                parentName: "AbpAccount::Menu:Account" /* Account */,
                order: 1,
            },
            {
                path: '/account/register',
                name: "AbpAccount::Register" /* Register */,
                parentName: "AbpAccount::Menu:Account" /* Account */,
                order: 2,
            },
            {
                path: '/account/manage',
                name: "AbpAccount::MyAccount" /* ManageProfile */,
                parentName: "AbpAccount::Menu:Account" /* Account */,
                layout: "application" /* application */,
                order: 3,
            },
            {
                path: '/account/forgot-password',
                parentName: "AbpAccount::Menu:Account" /* Account */,
                name: "AbpAccount::ForgotPassword" /* ForgotPassword */,
                invisible: true,
            },
            {
                path: '/account/reset-password',
                parentName: "AbpAccount::Menu:Account" /* Account */,
                name: "AbpAccount::ResetPassword" /* ResetPassword */,
                invisible: true,
            },
        ]);
    };
}

function navigateToManageProfileFactory(injector) {
    return () => {
        const router = injector.get(Router);
        const routes = injector.get(RoutesService);
        const { path } = routes.find(item => item.name === "AbpAccount::MyAccount" /* ManageProfile */);
        router.navigateByUrl(path);
    };
}

class AccountConfigModule {
    static forRoot() {
        return {
            ngModule: AccountConfigModule,
            providers: [
                ACCOUNT_ROUTE_PROVIDERS,
                {
                    provide: NAVIGATE_TO_MANAGE_PROFILE,
                    useFactory: navigateToManageProfileFactory,
                    deps: [Injector],
                },
            ],
        };
    }
}
AccountConfigModule.decorators = [
    { type: NgModule }
];

/**
 * Generated bundle index. Do not edit.
 */

export { ACCOUNT_ROUTE_PROVIDERS, AccountConfigModule, configureRoutes, ACCOUNT_ROUTE_PROVIDERS as ɵa, configureRoutes as ɵb, navigateToManageProfileFactory as ɵc };
