// @abp/ng.setting-management/esm2015/config/components/email-setting-group/email-setting-group.component.js
import { collapse, ToasterService } from '@abp/ng.theme.shared';
import { Component } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import { finalize } from 'rxjs/operators';
import { EmailSettingsService } from '../../proxy/email-settings.service';
export class EmailSettingGroupComponent {
    constructor(emailSettingsService, fb, toasterService) {
        this.emailSettingsService = emailSettingsService;
        this.fb = fb;
        this.toasterService = toasterService;
        this.saving = false;
    }
    ngOnInit() {
        this.getData();
    }
    getData() {
        this.emailSettingsService.get().subscribe(res => {
            this.buildForm(res);
        });
    }
    buildForm(emailSettings) {
        this.form = this.fb.group({
            defaultFromDisplayName: [emailSettings.defaultFromDisplayName, [Validators.required]],
            defaultFromAddress: [emailSettings.defaultFromAddress, [Validators.required]],
            smtpHost: [emailSettings.smtpHost],
            smtpPort: [emailSettings.smtpPort, [Validators.required]],
            smtpEnableSsl: [emailSettings.smtpEnableSsl],
            smtpUseDefaultCredentials: [emailSettings.smtpUseDefaultCredentials],
            smtpDomain: [emailSettings.smtpDomain],
            smtpUserName: [emailSettings.smtpUserName],
            smtpPassword: [emailSettings.smtpPassword],
        });
    }
    submit() {
        if (this.saving || this.form.invalid)
            return;
        this.saving = true;
        this.emailSettingsService
            .update(this.form.value)
            .pipe(finalize(() => (this.saving = false)))
            .subscribe(() => {
            this.toasterService.success('AbpSettingManagement::SuccessfullySaved');
            this.getData();
        });
    }
}
EmailSettingGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'abp-email-setting-group',
                template: "<h2>{{ 'AbpSettingManagement::Menu:Emailing' | abpLocalization }}</h2>\n\n<hr class=\"my-3\" />\n\n<form *ngIf=\"form\" [formGroup]=\"form\" (ngSubmit)=\"submit()\" validateOnSubmit>\n  <div class=\"form-group\">\n    <label>{{ 'AbpSettingManagement::DefaultFromDisplayName' | abpLocalization }}</label>\n    <input type=\"text\" class=\"form-control\" formControlName=\"defaultFromDisplayName\" />\n  </div>\n  <div class=\"form-group\">\n    <label>{{ 'AbpSettingManagement::DefaultFromAddress' | abpLocalization }}</label>\n    <input type=\"text\" class=\"form-control\" formControlName=\"defaultFromAddress\" />\n  </div>\n  <div class=\"form-group\">\n    <label>{{ 'AbpSettingManagement::SmtpHost' | abpLocalization }}</label>\n    <input type=\"text\" class=\"form-control\" formControlName=\"smtpHost\" />\n  </div>\n  <div class=\"form-group\">\n    <label>{{ 'AbpSettingManagement::SmtpPort' | abpLocalization }}</label>\n    <input type=\"number\" class=\"form-control\" formControlName=\"smtpPort\" />\n  </div>\n\n  <div class=\"custom-checkbox custom-control mb-2\">\n    <input\n      type=\"checkbox\"\n      id=\"smtp-enable-ssl\"\n      class=\"custom-control-input\"\n      formControlName=\"smtpEnableSsl\"\n    />\n    <label class=\"custom-control-label\" for=\"smtp-enable-ssl\">{{\n      'AbpSettingManagement::SmtpEnableSsl' | abpLocalization\n    }}</label>\n  </div>\n  <div class=\"custom-checkbox custom-control mb-2\">\n    <input\n      type=\"checkbox\"\n      id=\"smtp-use-default-credentials\"\n      class=\"custom-control-input\"\n      formControlName=\"smtpUseDefaultCredentials\"\n    />\n    <label class=\"custom-control-label\" for=\"smtp-use-default-credentials\">{{\n      'AbpSettingManagement::SmtpUseDefaultCredentials' | abpLocalization\n    }}</label>\n  </div>\n\n  <div\n    [@collapse]=\"{\n      value: form.get('smtpUseDefaultCredentials').value ? 'collapsed' : 'expanded',\n      params: { time: '200ms', easing: 'linear' }\n    }\"\n  >\n    <div class=\"form-group\">\n      <label>{{ 'AbpSettingManagement::SmtpDomain' | abpLocalization }}</label>\n      <input type=\"text\" class=\"form-control\" formControlName=\"smtpDomain\" />\n    </div>\n\n    <div class=\"form-group\">\n      <label>{{ 'AbpSettingManagement::SmtpUserName' | abpLocalization }}</label>\n      <input type=\"text\" class=\"form-control\" formControlName=\"smtpUserName\" />\n    </div>\n\n    <div class=\"form-group\">\n      <label>{{ 'AbpSettingManagement::SmtpPassword' | abpLocalization }}</label>\n      <input type=\"text\" class=\"form-control\" formControlName=\"smtpPassword\" />\n    </div>\n  </div>\n\n  <hr />\n\n  <button type=\"submit\" class=\"btn btn-primary\">\n    {{ 'AbpSettingManagement::Save' | abpLocalization }}\n  </button>\n</form>\n",
                animations: [collapse]
            },] }
];
EmailSettingGroupComponent.ctorParameters = () => [
    { type: EmailSettingsService },
    { type: FormBuilder },
    { type: ToasterService }
];
