// @abp/ng.core/esm2015/lib/services/application-configuration.service.js
import { Injectable } from '@angular/core';
import { RestService } from './rest.service';
import * as i0 from "@angular/core";
import * as i1 from "./rest.service";
/**
 * @deprecated Use AbpApplicationConfigurationService instead. To be deleted in v5.0.
 */
export class ApplicationConfigurationService {
    constructor(rest) {
        this.rest = rest;
    }
    getConfiguration() {
        const request = {
            method: 'GET',
            url: '/api/abp/application-configuration',
        };
        return this.rest.request(request, {});
    }
}
ApplicationConfigurationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ApplicationConfigurationService_Factory() { return new ApplicationConfigurationService(i0.ɵɵinject(i1.RestService)); }, token: ApplicationConfigurationService, providedIn: "root" });
ApplicationConfigurationService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ApplicationConfigurationService.ctorParameters = () => [
    { type: RestService }
];
