// @abp/ng.core/esm2015/lib/services/auth.service.js
import { __awaiter } from "tslib";
import { Injectable, Injector } from '@angular/core';
import { from } from 'rxjs';
import { filter, map, switchMap, take, tap } from 'rxjs/operators';
import { AUTH_FLOW_STRATEGY } from '../strategies/auth-flow.strategy';
import { EnvironmentService } from './environment.service';
import * as i0 from "@angular/core";
export class AuthService {
    constructor(injector) {
        this.injector = injector;
    }
    get isInternalAuth() {
        return this.strategy.isInternalAuth;
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            const environmentService = this.injector.get(EnvironmentService);
            return environmentService
                .getEnvironment$()
                .pipe(map(env => env === null || env === void 0 ? void 0 : env.oAuthConfig), filter(oAuthConfig => !!oAuthConfig), tap(oAuthConfig => {
                this.strategy =
                    oAuthConfig.responseType === 'code'
                        ? AUTH_FLOW_STRATEGY.Code(this.injector)
                        : AUTH_FLOW_STRATEGY.Password(this.injector);
            }), switchMap(() => from(this.strategy.init())), take(1))
                .toPromise();
        });
    }
    logout(queryParams) {
        return this.strategy.logout(queryParams);
    }
    /**
     * @deprecated Use navigateToLogin method instead. To be deleted in v5.0
     */
    initLogin() {
        this.strategy.navigateToLogin();
    }
    navigateToLogin(queryParams) {
        this.strategy.navigateToLogin(queryParams);
    }
    login(params) {
        return this.strategy.login(params);
    }
}
AuthService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AuthService_Factory() { return new AuthService(i0.ɵɵinject(i0.INJECTOR)); }, token: AuthService, providedIn: "root" });
AuthService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AuthService.ctorParameters = () => [
    { type: Injector }
];
