// @abp/ng.core/esm2015/lib/directives/permission.directive.js
import { ChangeDetectorRef, Directive, ElementRef, Input, Optional, Renderer2, TemplateRef, ViewContainerRef, } from '@angular/core';
import { distinctUntilChanged } from 'rxjs/operators';
import { PermissionService } from '../services/permission.service';
export class PermissionDirective {
    constructor(elRef, renderer, templateRef, vcRef, permissionService, cdRef) {
        this.elRef = elRef;
        this.renderer = renderer;
        this.templateRef = templateRef;
        this.vcRef = vcRef;
        this.permissionService = permissionService;
        this.cdRef = cdRef;
    }
    check() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
        this.subscription = this.permissionService
            .getGrantedPolicy$(this.condition)
            .pipe(distinctUntilChanged())
            .subscribe(isGranted => {
            if (this.templateRef)
                this.initStructural(isGranted);
            else
                this.initAttribute(isGranted);
            this.cdRef.detectChanges();
        });
    }
    initStructural(isGranted) {
        this.vcRef.clear();
        if (isGranted)
            this.vcRef.createEmbeddedView(this.templateRef);
    }
    /**
     * @deprecated Will be deleted in v5.0
     */
    initAttribute(isGranted) {
        if (!isGranted) {
            this.renderer.removeChild(this.elRef.nativeElement.parentElement, this.elRef.nativeElement);
        }
    }
    ngOnDestroy() {
        if (this.subscription)
            this.subscription.unsubscribe();
    }
    ngOnChanges() {
        this.check();
    }
}
PermissionDirective.decorators = [
    { type: Directive, args: [{
                selector: '[abpPermission]',
            },] }
];
PermissionDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TemplateRef, decorators: [{ type: Optional }] },
    { type: ViewContainerRef },
    { type: PermissionService },
    { type: ChangeDetectorRef }
];
PermissionDirective.propDecorators = {
    condition: [{ type: Input, args: ['abpPermission',] }]
};
