// @abp/ng.account/__ivy_ngcc__/fesm2015/abp-ng.account.js
import * as i1 from '@abp/ng.core';
import { RestService, AuthService, ConfigStateService, InternalStore, ProfileService, DynamicLayoutComponent, ReplaceableRouteContainerComponent, AuthGuard, LazyModuleFactory, CoreModule, mapEnumToOptions } from '@abp/ng.core';
import { ToasterService, fadeIn, getPasswordValidators, ThemeSharedModule } from '@abp/ng.theme.shared';
import * as i0 from '@angular/core';
import { Injectable, Component, InjectionToken, Injector, NgModule } from '@angular/core';
import { NgbDropdownModule } from '@ng-bootstrap/ng-bootstrap';
import { comparePasswords, NgxValidateCoreModule } from '@ngx-validate/core';
import { ActivatedRoute, Router, RouterModule } from '@angular/router';
import { Validators, FormBuilder, FormControl } from '@angular/forms';
import { finalize, catchError, switchMap } from 'rxjs/operators';
import { throwError } from 'rxjs';
import snq from 'snq';
import { trigger, transition, useAnimation } from '@angular/animations';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@abp/ng.core';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@ngx-validate/core';
import * as ɵngcc5 from '@abp/ng.theme.shared';
import * as ɵngcc6 from '@angular/router';

function ForgotPasswordComponent_form_3_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 2);
    ɵngcc0.ɵɵlistener("ngSubmit", function ForgotPasswordComponent_form_3_Template_form_ngSubmit_0_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onSubmit(); });
    ɵngcc0.ɵɵelementStart(1, "p");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 3);
    ɵngcc0.ɵɵelementStart(5, "label", 4);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "span");
    ɵngcc0.ɵɵtext(9, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(10, "input", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "abp-button", 6);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "a", 7);
    ɵngcc0.ɵɵelement(15, "i", 8);
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("formGroup", ctx_r0.form);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 7, "AbpAccount::SendPasswordResetLink_Information"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 9, "AbpAccount::EmailAddress"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("loading", ctx_r0.inProgress)("disabled", ctx_r0.form == null ? null : ctx_r0.form.invalid);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 11, "AbpAccount::Submit"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(17, 13, "AbpAccount::Login"));
} }
function ForgotPasswordComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "a", 7);
    ɵngcc0.ɵɵelementStart(4, "button", 9);
    ɵngcc0.ɵɵelement(5, "i", 8);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 2, "AbpAccount::PasswordResetMailSentMessage"), " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 4, "AbpAccount::BackToLogin"), " ");
} }
function LoginComponent_strong_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "strong");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "abpLocalization");
    ɵngcc0.ɵɵelementStart(3, "a", 14);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 2, "AbpAccount::AreYouANewUser"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 4, "AbpAccount::Register"));
} }
const _c0 = function (a0) { return { active: a0 }; };
function ManageProfileComponent_li_7_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 10);
    ɵngcc0.ɵɵlistener("click", function ManageProfileComponent_li_7_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.selectedTab = 0; });
    ɵngcc0.ɵɵelementStart(1, "a", 8);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c0, ctx_r0.selectedTab === 0));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, "AbpUi::ChangePassword"));
} }
function ManageProfileComponent_div_13_div_1_abp_change_password_form_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "abp-change-password-form");
} }
const _c1 = function (a0) { return { componentKey: a0 }; };
function ManageProfileComponent_div_13_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelementStart(2, "h4");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "abpLocalization");
    ɵngcc0.ɵɵelement(5, "hr");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, ManageProfileComponent_div_13_div_1_abp_change_password_form_6_Template, 1, 0, "abp-change-password-form", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("@fadeIn", undefined);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 3, "AbpIdentity::ChangePassword"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("abpReplaceableTemplate", ɵngcc0.ɵɵpureFunction1(5, _c1, ctx_r4.changePasswordKey));
} }
function ManageProfileComponent_div_13_div_2_abp_personal_settings_form_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "abp-personal-settings-form");
} }
function ManageProfileComponent_div_13_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelementStart(2, "h4");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "abpLocalization");
    ɵngcc0.ɵɵelement(5, "hr");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, ManageProfileComponent_div_13_div_2_abp_personal_settings_form_6_Template, 1, 0, "abp-personal-settings-form", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("@fadeIn", undefined);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 3, "AbpIdentity::PersonalSettings"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("abpReplaceableTemplate", ɵngcc0.ɵɵpureFunction1(5, _c1, ctx_r5.personalSettingsKey));
} }
function ManageProfileComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵtemplate(1, ManageProfileComponent_div_13_div_1_Template, 7, 7, "div", 12);
    ɵngcc0.ɵɵtemplate(2, ManageProfileComponent_div_13_div_2_Template, 7, 7, "div", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.selectedTab === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.selectedTab === 1);
} }
function RegisterComponent_form_9_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 2);
    ɵngcc0.ɵɵlistener("ngSubmit", function RegisterComponent_form_9_Template_form_ngSubmit_0_listener() { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.onSubmit(); });
    ɵngcc0.ɵɵelementStart(1, "div", 3);
    ɵngcc0.ɵɵelementStart(2, "label", 4);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "span");
    ɵngcc0.ɵɵtext(6, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(7, "input", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 3);
    ɵngcc0.ɵɵelementStart(9, "label", 6);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "span");
    ɵngcc0.ɵɵtext(13, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(14, "input", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 3);
    ɵngcc0.ɵɵelementStart(16, "label", 8);
    ɵngcc0.ɵɵtext(17);
    ɵngcc0.ɵɵpipe(18, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(19, "span");
    ɵngcc0.ɵɵtext(20, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(21, "input", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(22, "abp-button", 10);
    ɵngcc0.ɵɵtext(23);
    ɵngcc0.ɵɵpipe(24, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("formGroup", ctx_r0.form);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 6, "AbpAccount::UserName"));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 8, "AbpAccount::EmailAddress"));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(18, 10, "AbpAccount::Password"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("loading", ctx_r0.inProgress);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(24, 12, "AbpAccount::Register"), " ");
} }
function ResetPasswordComponent_form_3_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 2);
    ɵngcc0.ɵɵlistener("ngSubmit", function ResetPasswordComponent_form_3_Template_form_ngSubmit_0_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onSubmit(); });
    ɵngcc0.ɵɵelementStart(1, "p");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 3);
    ɵngcc0.ɵɵelementStart(5, "label", 4);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "span");
    ɵngcc0.ɵɵtext(9, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(10, "input", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 3);
    ɵngcc0.ɵɵelementStart(12, "label", 6);
    ɵngcc0.ɵɵtext(13);
    ɵngcc0.ɵɵpipe(14, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "span");
    ɵngcc0.ɵɵtext(16, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(17, "input", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(18, "button", 8);
    ɵngcc0.ɵɵtext(19);
    ɵngcc0.ɵɵpipe(20, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "abp-button", 9);
    ɵngcc0.ɵɵlistener("click", function ResetPasswordComponent_form_3_Template_abp_button_click_21_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.onSubmit(); });
    ɵngcc0.ɵɵtext(22);
    ɵngcc0.ɵɵpipe(23, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("formGroup", ctx_r0.form)("mapErrorsFn", ctx_r0.mapErrorsFn);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 8, "AbpAccount::ResetPassword_Information"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 10, "AbpAccount::Password"));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(14, 12, "AbpAccount::ConfirmPassword"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 14, "AbpAccount::Cancel"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("loading", ctx_r0.inProgress);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(23, 16, "AbpAccount::Submit"), " ");
} }
function ResetPasswordComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "a", 10);
    ɵngcc0.ɵɵelementStart(4, "button", 11);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 2, "AbpAccount::YourPasswordIsSuccessfullyReset"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 4, "AbpAccount::BackToLogin"), " ");
} }
function ChangePasswordComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "label", 8);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "input", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, "AbpIdentity::DisplayName:CurrentPassword"));
} }
function PersonalSettingsComponent_form_0_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 1);
    ɵngcc0.ɵɵlistener("ngSubmit", function PersonalSettingsComponent_form_0_Template_form_ngSubmit_0_listener() { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.submit(); });
    ɵngcc0.ɵɵelementStart(1, "div", 2);
    ɵngcc0.ɵɵelementStart(2, "label", 3);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "span");
    ɵngcc0.ɵɵtext(6, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "input", 4);
    ɵngcc0.ɵɵlistener("keydown.space", function PersonalSettingsComponent_form_0_Template_input_keydown_space_7_listener($event) { return $event.preventDefault(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 5);
    ɵngcc0.ɵɵelementStart(9, "div", 6);
    ɵngcc0.ɵɵelementStart(10, "div", 2);
    ɵngcc0.ɵɵelementStart(11, "label", 7);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(14, "input", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 6);
    ɵngcc0.ɵɵelementStart(16, "div", 2);
    ɵngcc0.ɵɵelementStart(17, "label", 9);
    ɵngcc0.ɵɵtext(18);
    ɵngcc0.ɵɵpipe(19, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(20, "input", 10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "div", 2);
    ɵngcc0.ɵɵelementStart(22, "label", 11);
    ɵngcc0.ɵɵtext(23);
    ɵngcc0.ɵɵpipe(24, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "span");
    ɵngcc0.ɵɵtext(26, " * ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(27, "input", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(28, "div", 2);
    ɵngcc0.ɵɵelementStart(29, "label", 13);
    ɵngcc0.ɵɵtext(30);
    ɵngcc0.ɵɵpipe(31, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(32, "input", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(33, "abp-button", 15);
    ɵngcc0.ɵɵtext(34);
    ɵngcc0.ɵɵpipe(35, "abpLocalization");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("formGroup", ctx_r0.form);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 9, "AbpIdentity::DisplayName:UserName"));
    ɵngcc0.ɵɵadvance(9);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(13, 11, "AbpIdentity::DisplayName:Name"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(19, 13, "AbpIdentity::DisplayName:Surname"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(24, 15, "AbpIdentity::DisplayName:Email"));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(31, 17, "AbpIdentity::DisplayName:PhoneNumber"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("loading", ctx_r0.inProgress)("disabled", ctx_r0.form == null ? null : ctx_r0.form.invalid);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(35, 19, "AbpIdentity::Save"), "");
} }
class AccountService$1 {
    constructor(restService) {
        this.restService = restService;
        this.apiName = 'AbpAccount';
        this.register = (input) => this.restService.request({
            method: 'POST',
            url: '/api/account/register',
            body: input,
        }, { apiName: this.apiName });
        this.resetPassword = (input) => this.restService.request({
            method: 'POST',
            url: '/api/account/reset-password',
            body: input,
        }, { apiName: this.apiName });
        this.sendPasswordResetCode = (input) => this.restService.request({
            method: 'POST',
            url: '/api/account/send-password-reset-code',
            body: input,
        }, { apiName: this.apiName });
    }
}
AccountService$1.ɵfac = function AccountService$1_Factory(t) { return new (t || AccountService$1)(ɵngcc0.ɵɵinject(ɵngcc1.RestService)); };
AccountService$1.ɵprov = i0.ɵɵdefineInjectable({ factory: function AccountService_Factory() { return new AccountService$1(i0.ɵɵinject(i1.RestService)); }, token: AccountService$1, providedIn: "root" });
AccountService$1.ctorParameters = () => [
    { type: RestService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AccountService$1, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.RestService }]; }, null); })();

class ForgotPasswordComponent {
    constructor(fb, accountService) {
        this.fb = fb;
        this.accountService = accountService;
        this.isEmailSent = false;
        this.form = this.fb.group({
            email: ['', [Validators.required, Validators.email]],
        });
    }
    onSubmit() {
        if (this.form.invalid)
            return;
        this.inProgress = true;
        this.accountService
            .sendPasswordResetCode({ email: this.form.get('email').value, appName: 'Angular' })
            .pipe(finalize(() => (this.inProgress = false)))
            .subscribe(() => {
            this.isEmailSent = true;
        });
    }
}
ForgotPasswordComponent.ɵfac = function ForgotPasswordComponent_Factory(t) { return new (t || ForgotPasswordComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(AccountService$1)); };
ForgotPasswordComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ForgotPasswordComponent, selectors: [["abp-forgot-password"]], decls: 6, vars: 5, consts: [["validateOnSubmit", "", 3, "formGroup", "ngSubmit", 4, "ngIf", "ngIfElse"], ["emailSentTemplate", ""], ["validateOnSubmit", "", 3, "formGroup", "ngSubmit"], [1, "form-group"], ["for", "input-email-address"], ["type", "email", "id", "input-email-address", "formControlName", "email", 1, "form-control"], ["buttonClass", "mt-2 mb-3 btn btn-primary btn-block", "buttonType", "submit", 1, "d-block", 3, "loading", "disabled"], ["routerLink", "/account/login"], [1, "fa", "fa-long-arrow-left", "mr-1"], [1, "d-block", "mt-2", "mb-3", "btn", "btn-primary", "btn-block"]], template: function ForgotPasswordComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "h4");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, ForgotPasswordComponent_form_3_Template, 18, 15, "form", 0);
        ɵngcc0.ɵɵtemplate(4, ForgotPasswordComponent_ng_template_4_Template, 8, 6, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(5);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 3, "AbpAccount::ForgotPassword"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isEmailSent)("ngIfElse", _r1);
    } }, directives: [ɵngcc3.NgIf, ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc1.FormSubmitDirective, ɵngcc2.FormGroupDirective, ɵngcc4.ValidationGroupDirective, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc4.ValidationDirective, ɵngcc5.ButtonComponent, ɵngcc6.RouterLinkWithHref], pipes: [ɵngcc1.LocalizationPipe], encapsulation: 2 });
ForgotPasswordComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: AccountService$1 }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ForgotPasswordComponent, [{
        type: Component,
        args: [{
                selector: 'abp-forgot-password',
                template: "<h4>{{ 'AbpAccount::ForgotPassword' | abpLocalization }}</h4>\n\n<form\n  *ngIf=\"!isEmailSent; else emailSentTemplate\"\n  [formGroup]=\"form\"\n  (ngSubmit)=\"onSubmit()\"\n  validateOnSubmit\n>\n  <p>{{ 'AbpAccount::SendPasswordResetLink_Information' | abpLocalization }}</p>\n  <div class=\"form-group\">\n    <label for=\"input-email-address\">{{ 'AbpAccount::EmailAddress' | abpLocalization }}</label\n    ><span> * </span>\n    <input type=\"email\" id=\"input-email-address\" class=\"form-control\" formControlName=\"email\" />\n  </div>\n  <abp-button\n    class=\"d-block\"\n    buttonClass=\"mt-2 mb-3 btn btn-primary btn-block\"\n    [loading]=\"inProgress\"\n    buttonType=\"submit\"\n    [disabled]=\"form?.invalid\"\n  >\n    {{ 'AbpAccount::Submit' | abpLocalization }}\n  </abp-button>\n  <a routerLink=\"/account/login\"\n    ><i class=\"fa fa-long-arrow-left mr-1\"></i>{{ 'AbpAccount::Login' | abpLocalization }}</a\n  >\n</form>\n\n<ng-template #emailSentTemplate>\n  <p>\n    {{ 'AbpAccount::PasswordResetMailSentMessage' | abpLocalization }}\n  </p>\n\n  <a routerLink=\"/account/login\">\n    <button class=\"d-block mt-2 mb-3 btn btn-primary btn-block\">\n      <i class=\"fa fa-long-arrow-left mr-1\"></i>\n      {{ 'AbpAccount::BackToLogin' | abpLocalization }}\n    </button>\n  </a>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc2.FormBuilder }, { type: AccountService$1 }]; }, null); })();

const ACCOUNT_CONFIG_OPTIONS = new InjectionToken('ACCOUNT_CONFIG_OPTIONS');

function getRedirectUrl(injector) {
    const route = injector.get(ActivatedRoute);
    const options = injector.get(ACCOUNT_CONFIG_OPTIONS);
    return route.snapshot.queryParams.returnUrl || options.redirectUrl || '/';
}

const { maxLength: maxLength$2, required: required$3 } = Validators;
class LoginComponent {
    constructor(injector, fb, toasterService, authService, configState) {
        this.injector = injector;
        this.fb = fb;
        this.toasterService = toasterService;
        this.authService = authService;
        this.configState = configState;
        this.isSelfRegistrationEnabled = true;
        this.authWrapperKey = "Account.AuthWrapperComponent" /* AuthWrapper */;
    }
    ngOnInit() {
        this.init();
        this.buildForm();
    }
    init() {
        this.isSelfRegistrationEnabled =
            (this.configState.getSetting('Abp.Account.IsSelfRegistrationEnabled') || '').toLowerCase() !== 'false';
    }
    buildForm() {
        this.form = this.fb.group({
            username: ['', [required$3, maxLength$2(255)]],
            password: ['', [required$3, maxLength$2(128)]],
            rememberMe: [false],
        });
    }
    onSubmit() {
        if (this.form.invalid)
            return;
        this.inProgress = true;
        const { username, password, rememberMe } = this.form.value;
        const redirectUrl = getRedirectUrl(this.injector);
        this.authService
            .login({ username, password, rememberMe, redirectUrl })
            .pipe(catchError(err => {
            this.toasterService.error(snq(() => err.error.error_description) ||
                snq(() => err.error.error.message, 'AbpAccount::DefaultErrorMessage'), 'Error', { life: 7000 });
            return throwError(err);
        }), finalize(() => (this.inProgress = false)))
            .subscribe();
    }
}
LoginComponent.ɵfac = function LoginComponent_Factory(t) { return new (t || LoginComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.ToasterService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AuthService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ConfigStateService)); };
LoginComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LoginComponent, selectors: [["abp-login"]], decls: 28, vars: 21, consts: [[4, "ngIf"], ["validateOnSubmit", "", 1, "mt-4", 3, "formGroup", "ngSubmit"], [1, "form-group"], ["for", "login-input-user-name-or-email-address"], ["type", "text", "id", "login-input-user-name-or-email-address", "formControlName", "username", "autocomplete", "username", "autofocus", "", 1, "form-control"], ["for", "login-input-password"], ["type", "password", "id", "login-input-password", "formControlName", "password", "autocomplete", "current-password", 1, "form-control"], [1, "row"], [1, "col"], ["for", "login-input-remember-me", 1, "custom-checkbox", "custom-control", "mb-2"], ["type", "checkbox", "id", "login-input-remember-me", "formControlName", "rememberMe", 1, "form-check-input"], [1, "text-right", "col"], ["routerLink", "/account/forgot-password"], ["buttonType", "submit", "name", "Action", "buttonClass", "btn-block btn-lg mt-3 btn btn-primary", 3, "loading"], ["routerLink", "/account/register", "queryParamsHandling", "preserve", 1, "text-decoration-none"]], template: function LoginComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "h4");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, LoginComponent_strong_3_Template, 6, 6, "strong", 0);
        ɵngcc0.ɵɵelementStart(4, "form", 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function LoginComponent_Template_form_ngSubmit_4_listener() { return ctx.onSubmit(); });
        ɵngcc0.ɵɵelementStart(5, "div", 2);
        ɵngcc0.ɵɵelementStart(6, "label", 3);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(9, "input", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 2);
        ɵngcc0.ɵɵelementStart(11, "label", 5);
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(14, "input", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "div", 7);
        ɵngcc0.ɵɵelementStart(16, "div", 8);
        ɵngcc0.ɵɵelementStart(17, "label", 9);
        ɵngcc0.ɵɵelement(18, "input", 10);
        ɵngcc0.ɵɵtext(19);
        ɵngcc0.ɵɵpipe(20, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "div", 11);
        ɵngcc0.ɵɵelementStart(22, "a", 12);
        ɵngcc0.ɵɵtext(23);
        ɵngcc0.ɵɵpipe(24, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "abp-button", 13);
        ɵngcc0.ɵɵtext(26);
        ɵngcc0.ɵɵpipe(27, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 9, "AbpAccount::Login"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSelfRegistrationEnabled);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("formGroup", ctx.form);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 11, "AbpAccount::UserNameOrEmailAddress"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(13, 13, "AbpAccount::Password"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 15, "AbpAccount::RememberMe"), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(24, 17, "AbpAccount::ForgotPassword"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("loading", ctx.inProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(27, 19, "AbpAccount::Login"), " ");
    } }, directives: [ɵngcc3.NgIf, ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc1.FormSubmitDirective, ɵngcc2.FormGroupDirective, ɵngcc4.ValidationGroupDirective, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc4.ValidationDirective, ɵngcc1.AutofocusDirective, ɵngcc2.CheckboxControlValueAccessor, ɵngcc6.RouterLinkWithHref, ɵngcc5.ButtonComponent], pipes: [ɵngcc1.LocalizationPipe], encapsulation: 2 });
LoginComponent.ctorParameters = () => [
    { type: Injector },
    { type: FormBuilder },
    { type: ToasterService },
    { type: AuthService },
    { type: ConfigStateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LoginComponent, [{
        type: Component,
        args: [{
                selector: 'abp-login',
                template: "<h4>{{ 'AbpAccount::Login' | abpLocalization }}</h4>\n<strong *ngIf=\"isSelfRegistrationEnabled\">\n  {{ 'AbpAccount::AreYouANewUser' | abpLocalization }}\n  <a class=\"text-decoration-none\" routerLink=\"/account/register\" queryParamsHandling=\"preserve\">{{\n    'AbpAccount::Register' | abpLocalization\n  }}</a>\n</strong>\n<form [formGroup]=\"form\" (ngSubmit)=\"onSubmit()\" validateOnSubmit class=\"mt-4\">\n  <div class=\"form-group\">\n    <label for=\"login-input-user-name-or-email-address\">{{\n      'AbpAccount::UserNameOrEmailAddress' | abpLocalization\n    }}</label>\n    <input\n      class=\"form-control\"\n      type=\"text\"\n      id=\"login-input-user-name-or-email-address\"\n      formControlName=\"username\"\n      autocomplete=\"username\"\n      autofocus\n    />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"login-input-password\">{{ 'AbpAccount::Password' | abpLocalization }}</label>\n    <input\n      class=\"form-control\"\n      type=\"password\"\n      id=\"login-input-password\"\n      formControlName=\"password\"\n      autocomplete=\"current-password\"\n    />\n  </div>\n\n  <div class=\"row\">\n    <div class=\"col\">\n      <label class=\"custom-checkbox custom-control mb-2\" for=\"login-input-remember-me\">\n        <input\n          class=\"form-check-input\"\n          type=\"checkbox\"\n          id=\"login-input-remember-me\"\n          formControlName=\"rememberMe\"\n        />\n        {{ 'AbpAccount::RememberMe' | abpLocalization }}\n      </label>\n    </div>\n    <div class=\"text-right col\">\n      <a routerLink=\"/account/forgot-password\">{{\n        'AbpAccount::ForgotPassword' | abpLocalization\n      }}</a>\n    </div>\n  </div>\n\n  <abp-button\n    [loading]=\"inProgress\"\n    buttonType=\"submit\"\n    name=\"Action\"\n    buttonClass=\"btn-block btn-lg mt-3 btn btn-primary\"\n  >\n    {{ 'AbpAccount::Login' | abpLocalization }}\n  </abp-button>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc0.Injector }, { type: ɵngcc2.FormBuilder }, { type: ɵngcc5.ToasterService }, { type: ɵngcc1.AuthService }, { type: ɵngcc1.ConfigStateService }]; }, null); })();

class ManageProfileStateService {
    constructor() {
        this.store = new InternalStore({});
    }
    get createOnUpdateStream() {
        return this.store.sliceUpdate;
    }
    getProfile$() {
        return this.store.sliceState(state => state.profile);
    }
    getProfile() {
        return this.store.state.profile;
    }
    setProfile(profile) {
        this.store.patch({ profile });
    }
}
ManageProfileStateService.ɵfac = function ManageProfileStateService_Factory(t) { return new (t || ManageProfileStateService)(); };
ManageProfileStateService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ManageProfileStateService_Factory() { return new ManageProfileStateService(); }, token: ManageProfileStateService, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ManageProfileStateService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();

class ManageProfileComponent {
    constructor(profileService, manageProfileState) {
        this.profileService = profileService;
        this.manageProfileState = manageProfileState;
        this.selectedTab = 0;
        this.changePasswordKey = "Account.ChangePasswordComponent" /* ChangePassword */;
        this.personalSettingsKey = "Account.PersonalSettingsComponent" /* PersonalSettings */;
        this.profile$ = this.manageProfileState.getProfile$();
    }
    ngOnInit() {
        this.profileService.get().subscribe(profile => {
            this.manageProfileState.setProfile(profile);
            if (profile.isExternal) {
                this.hideChangePasswordTab = true;
                this.selectedTab = 1;
            }
        });
    }
}
ManageProfileComponent.ɵfac = function ManageProfileComponent_Factory(t) { return new (t || ManageProfileComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ProfileService), ɵngcc0.ɵɵdirectiveInject(ManageProfileStateService)); };
ManageProfileComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ManageProfileComponent, selectors: [["abp-manage-profile"]], decls: 15, vars: 15, consts: [["id", "AbpContentToolbar"], [1, "card", "border-0", "shadow-sm", "min-h-400", 3, "abpLoading"], [1, "card-body"], [1, "row"], [1, "col-12", "col-md-3"], ["id", "nav-tab", "role", "tablist", 1, "nav", "flex-column", "nav-pills"], ["class", "nav-item", 3, "click", 4, "ngIf"], [1, "nav-item", "mb-2", 3, "click"], ["role", "tab", "href", "javascript:void(0)", 1, "nav-link", 3, "ngClass"], ["class", "col-12 col-md-9", 4, "ngIf"], [1, "nav-item", 3, "click"], [1, "col-12", "col-md-9"], ["class", "tab-content", 4, "ngIf"], [1, "tab-content"], ["role", "tabpanel", 1, "tab-pane", "active"], [4, "abpReplaceableTemplate"]], template: function ManageProfileComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵelementStart(3, "div", 2);
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "ul", 5);
        ɵngcc0.ɵɵtemplate(7, ManageProfileComponent_li_7_Template, 4, 6, "li", 6);
        ɵngcc0.ɵɵpipe(8, "async");
        ɵngcc0.ɵɵelementStart(9, "li", 7);
        ɵngcc0.ɵɵlistener("click", function ManageProfileComponent_Template_li_click_9_listener() { return ctx.selectedTab = 1; });
        ɵngcc0.ɵɵelementStart(10, "a", 8);
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵpipe(12, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(13, ManageProfileComponent_div_13_Template, 3, 2, "div", 9);
        ɵngcc0.ɵɵpipe(14, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("abpLoading", !ɵngcc0.ɵɵpipeBind1(2, 5, ctx.profile$));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.hideChangePasswordTab && ɵngcc0.ɵɵpipeBind1(8, 7, ctx.profile$));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(13, _c0, ctx.selectedTab === 1));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(12, 9, "AbpAccount::PersonalSettings"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(14, 11, ctx.profile$));
    } }, directives: function () { return [ɵngcc5.LoadingDirective, ɵngcc3.NgIf, ɵngcc3.NgClass, ɵngcc1.ReplaceableTemplateDirective, ChangePasswordComponent, PersonalSettingsComponent]; }, pipes: function () { return [ɵngcc3.AsyncPipe, ɵngcc1.LocalizationPipe]; }, styles: [".min-h-400[_ngcontent-%COMP%] {\n        min-height: 400px;\n      }"], data: { animation: [trigger('fadeIn', [transition(':enter', useAnimation(fadeIn))])] } });
ManageProfileComponent.ctorParameters = () => [
    { type: ProfileService },
    { type: ManageProfileStateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ManageProfileComponent, [{
        type: Component,
        args: [{
                selector: 'abp-manage-profile',
                template: "<div id=\"AbpContentToolbar\"></div>\n\n<div class=\"card border-0 shadow-sm min-h-400\" [abpLoading]=\"!(profile$ | async)\">\n  <div class=\"card-body\">\n    <div class=\"row\">\n      <div class=\"col-12 col-md-3\">\n        <ul class=\"nav flex-column nav-pills\" id=\"nav-tab\" role=\"tablist\">\n          <li\n            *ngIf=\"!hideChangePasswordTab && (profile$ | async)\"\n            class=\"nav-item\"\n            (click)=\"selectedTab = 0\"\n          >\n            <a\n              class=\"nav-link\"\n              [ngClass]=\"{ active: selectedTab === 0 }\"\n              role=\"tab\"\n              href=\"javascript:void(0)\"\n              >{{ 'AbpUi::ChangePassword' | abpLocalization }}</a\n            >\n          </li>\n          <li class=\"nav-item mb-2\" (click)=\"selectedTab = 1\">\n            <a\n              class=\"nav-link\"\n              [ngClass]=\"{ active: selectedTab === 1 }\"\n              role=\"tab\"\n              href=\"javascript:void(0)\"\n              >{{ 'AbpAccount::PersonalSettings' | abpLocalization }}</a\n            >\n          </li>\n        </ul>\n      </div>\n      <div *ngIf=\"profile$ | async\" class=\"col-12 col-md-9\">\n        <div class=\"tab-content\" *ngIf=\"selectedTab === 0\" [@fadeIn]>\n          <div class=\"tab-pane active\" role=\"tabpanel\">\n            <h4>\n              {{ 'AbpIdentity::ChangePassword' | abpLocalization }}\n              <hr />\n            </h4>\n            <abp-change-password-form\n              *abpReplaceableTemplate=\"{\n                componentKey: changePasswordKey\n              }\"\n            ></abp-change-password-form>\n          </div>\n        </div>\n        <div class=\"tab-content\" *ngIf=\"selectedTab === 1\" [@fadeIn]>\n          <div class=\"tab-pane active\" role=\"tabpanel\">\n            <h4>\n              {{ 'AbpIdentity::PersonalSettings' | abpLocalization }}\n              <hr />\n            </h4>\n            <abp-personal-settings-form\n              *abpReplaceableTemplate=\"{\n                componentKey: personalSettingsKey\n              }\"\n            ></abp-personal-settings-form>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n",
                animations: [trigger('fadeIn', [transition(':enter', useAnimation(fadeIn))])],
                styles: [`
      .min-h-400 {
        min-height: 400px;
      }
    `]
            }]
    }], function () { return [{ type: ɵngcc1.ProfileService }, { type: ManageProfileStateService }]; }, null); })();

const { maxLength: maxLength$1, required: required$2, email: email$1 } = Validators;
class RegisterComponent {
    constructor(fb, accountService, configState, toasterService, authService, injector) {
        this.fb = fb;
        this.accountService = accountService;
        this.configState = configState;
        this.toasterService = toasterService;
        this.authService = authService;
        this.injector = injector;
        this.isSelfRegistrationEnabled = true;
        this.authWrapperKey = "Account.AuthWrapperComponent" /* AuthWrapper */;
    }
    ngOnInit() {
        this.init();
        this.buildForm();
    }
    init() {
        this.isSelfRegistrationEnabled =
            (this.configState.getSetting('Abp.Account.IsSelfRegistrationEnabled') || '').toLowerCase() !==
                'false';
        if (!this.isSelfRegistrationEnabled) {
            this.toasterService.warn({
                key: 'AbpAccount::SelfRegistrationDisabledMessage',
                defaultValue: 'Self registration is disabled.',
            }, null, { life: 10000 });
            return;
        }
    }
    buildForm() {
        this.form = this.fb.group({
            username: ['', [required$2, maxLength$1(255)]],
            password: ['', [required$2, ...getPasswordValidators(this.injector)]],
            email: ['', [required$2, email$1]],
        });
    }
    onSubmit() {
        if (this.form.invalid)
            return;
        this.inProgress = true;
        const newUser = {
            userName: this.form.get('username').value,
            password: this.form.get('password').value,
            emailAddress: this.form.get('email').value,
            appName: 'Angular',
        };
        this.accountService
            .register(newUser)
            .pipe(switchMap(() => this.authService.login({
            username: newUser.userName,
            password: newUser.password,
            redirectUrl: getRedirectUrl(this.injector),
        })), catchError(err => {
            this.toasterService.error(snq(() => err.error.error_description) ||
                snq(() => err.error.error.message, 'AbpAccount::DefaultErrorMessage'), 'Error', { life: 7000 });
            return throwError(err);
        }), finalize(() => (this.inProgress = false)))
            .subscribe();
    }
}
RegisterComponent.ɵfac = function RegisterComponent_Factory(t) { return new (t || RegisterComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(AccountService$1), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ConfigStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.ToasterService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AuthService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector)); };
RegisterComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: RegisterComponent, selectors: [["abp-register"]], decls: 10, vars: 10, consts: [["routerLink", "/account/login", 1, "text-decoration-none"], ["validateOnSubmit", "", "class", "mt-4", 3, "formGroup", "ngSubmit", 4, "ngIf"], ["validateOnSubmit", "", 1, "mt-4", 3, "formGroup", "ngSubmit"], [1, "form-group"], ["for", "input-user-name"], ["autofocus", "", "type", "text", "id", "input-user-name", "formControlName", "username", "autocomplete", "username", 1, "form-control"], ["for", "input-email-address"], ["type", "email", "id", "input-email-address", "formControlName", "email", 1, "form-control"], ["for", "input-password"], ["type", "password", "id", "input-password", "formControlName", "password", "autocomplete", "current-password", 1, "form-control"], ["buttonType", "submit", "name", "Action", "buttonClass", "btn-block btn-lg mt-3 btn btn-primary", 3, "loading"]], template: function RegisterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "h4");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "strong");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "abpLocalization");
        ɵngcc0.ɵɵelementStart(6, "a", 0);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, RegisterComponent_form_9_Template, 25, 14, "form", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 4, "AbpAccount::Register"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 6, "AbpAccount::AlreadyRegistered"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 8, "AbpAccount::Login"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSelfRegistrationEnabled);
    } }, directives: [ɵngcc6.RouterLinkWithHref, ɵngcc3.NgIf, ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc1.FormSubmitDirective, ɵngcc2.FormGroupDirective, ɵngcc4.ValidationGroupDirective, ɵngcc2.DefaultValueAccessor, ɵngcc1.AutofocusDirective, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc4.ValidationDirective, ɵngcc5.ButtonComponent], pipes: [ɵngcc1.LocalizationPipe], encapsulation: 2 });
RegisterComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: AccountService$1 },
    { type: ConfigStateService },
    { type: ToasterService },
    { type: AuthService },
    { type: Injector }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RegisterComponent, [{
        type: Component,
        args: [{
                selector: 'abp-register',
                template: "<h4>{{ 'AbpAccount::Register' | abpLocalization }}</h4>\n<strong>\n  {{ 'AbpAccount::AlreadyRegistered' | abpLocalization }}\n  <a class=\"text-decoration-none\" routerLink=\"/account/login\">{{\n    'AbpAccount::Login' | abpLocalization\n  }}</a>\n</strong>\n<form\n  *ngIf=\"isSelfRegistrationEnabled\"\n  [formGroup]=\"form\"\n  (ngSubmit)=\"onSubmit()\"\n  validateOnSubmit\n  class=\"mt-4\"\n>\n  <div class=\"form-group\">\n    <label for=\"input-user-name\">{{ 'AbpAccount::UserName' | abpLocalization }}</label\n    ><span> * </span\n    ><input\n      autofocus\n      type=\"text\"\n      id=\"input-user-name\"\n      class=\"form-control\"\n      formControlName=\"username\"\n      autocomplete=\"username\"\n    />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"input-email-address\">{{ 'AbpAccount::EmailAddress' | abpLocalization }}</label\n    ><span> * </span\n    ><input type=\"email\" id=\"input-email-address\" class=\"form-control\" formControlName=\"email\" />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"input-password\">{{ 'AbpAccount::Password' | abpLocalization }}</label\n    ><span> * </span\n    ><input\n      type=\"password\"\n      id=\"input-password\"\n      class=\"form-control\"\n      formControlName=\"password\"\n      autocomplete=\"current-password\"\n    />\n  </div>\n  <abp-button\n    [loading]=\"inProgress\"\n    buttonType=\"submit\"\n    name=\"Action\"\n    buttonClass=\"btn-block btn-lg mt-3 btn btn-primary\"\n  >\n    {{ 'AbpAccount::Register' | abpLocalization }}\n  </abp-button>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc2.FormBuilder }, { type: AccountService$1 }, { type: ɵngcc1.ConfigStateService }, { type: ɵngcc5.ToasterService }, { type: ɵngcc1.AuthService }, { type: ɵngcc0.Injector }]; }, null); })();

const PASSWORD_FIELDS$1 = ['password', 'confirmPassword'];
class ResetPasswordComponent {
    constructor(fb, accountService, route, router, injector) {
        this.fb = fb;
        this.accountService = accountService;
        this.route = route;
        this.router = router;
        this.injector = injector;
        this.inProgress = false;
        this.isPasswordReset = false;
        this.mapErrorsFn = (errors, groupErrors, control) => {
            if (PASSWORD_FIELDS$1.indexOf(String(control.name)) < 0)
                return errors;
            return errors.concat(groupErrors.filter(({ key }) => key === 'passwordMismatch'));
        };
    }
    ngOnInit() {
        this.route.queryParams.subscribe(({ userId, resetToken }) => {
            if (!userId || !resetToken)
                this.router.navigateByUrl('/account/login');
            this.form = this.fb.group({
                userId: [userId, [Validators.required]],
                resetToken: [resetToken, [Validators.required]],
                password: ['', [Validators.required, ...getPasswordValidators(this.injector)]],
                confirmPassword: ['', [Validators.required, ...getPasswordValidators(this.injector)]],
            }, {
                validators: [comparePasswords(PASSWORD_FIELDS$1)],
            });
        });
    }
    onSubmit() {
        if (this.form.invalid || this.inProgress)
            return;
        this.inProgress = true;
        this.accountService
            .resetPassword({
            userId: this.form.get('userId').value,
            resetToken: this.form.get('resetToken').value,
            password: this.form.get('password').value,
        })
            .pipe(finalize(() => (this.inProgress = false)))
            .subscribe(() => {
            this.isPasswordReset = true;
        });
    }
}
ResetPasswordComponent.ɵfac = function ResetPasswordComponent_Factory(t) { return new (t || ResetPasswordComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(AccountService$1), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector)); };
ResetPasswordComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ResetPasswordComponent, selectors: [["abp-reset-password"]], decls: 6, vars: 5, consts: [["validateOnSubmit", "", 3, "formGroup", "mapErrorsFn", "ngSubmit", 4, "ngIf", "ngIfElse"], ["passwordResetTemplate", ""], ["validateOnSubmit", "", 3, "formGroup", "mapErrorsFn", "ngSubmit"], [1, "form-group"], ["for", "input-password"], ["type", "password", "id", "input-password", "formControlName", "password", 1, "form-control"], ["for", "input-confirm-password"], ["type", "password", "id", "input-confirm-password", "formControlName", "confirmPassword", 1, "form-control"], ["type", "button", "routerLink", "/account/login", 1, "mr-2", "btn", "btn-secondary"], ["buttonType", "submit", "buttonClass", "mr-2 btn btn-primary", 3, "loading", "click"], ["routerLink", "/account/login"], [1, "d-block", "mt-2", "mb-3", "btn", "btn-primary"]], template: function ResetPasswordComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "h4");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, ResetPasswordComponent_form_3_Template, 24, 18, "form", 0);
        ɵngcc0.ɵɵtemplate(4, ResetPasswordComponent_ng_template_4_Template, 7, 6, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(5);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 3, "AbpAccount::ResetPassword"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isPasswordReset)("ngIfElse", _r1);
    } }, directives: [ɵngcc3.NgIf, ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc1.FormSubmitDirective, ɵngcc2.FormGroupDirective, ɵngcc4.ValidationGroupDirective, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc4.ValidationDirective, ɵngcc6.RouterLink, ɵngcc5.ButtonComponent, ɵngcc6.RouterLinkWithHref], pipes: [ɵngcc1.LocalizationPipe], encapsulation: 2 });
ResetPasswordComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: AccountService$1 },
    { type: ActivatedRoute },
    { type: Router },
    { type: Injector }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ResetPasswordComponent, [{
        type: Component,
        args: [{
                selector: 'abp-reset-password',
                template: "<h4>{{ 'AbpAccount::ResetPassword' | abpLocalization }}</h4>\n\n<form\n  *ngIf=\"!isPasswordReset; else passwordResetTemplate\"\n  [formGroup]=\"form\"\n  [mapErrorsFn]=\"mapErrorsFn\"\n  (ngSubmit)=\"onSubmit()\"\n  validateOnSubmit\n>\n  <p>{{ 'AbpAccount::ResetPassword_Information' | abpLocalization }}</p>\n  <div class=\"form-group\">\n    <label for=\"input-password\">{{ 'AbpAccount::Password' | abpLocalization }}</label\n    ><span> * </span>\n    <input type=\"password\" id=\"input-password\" class=\"form-control\" formControlName=\"password\" />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"input-confirm-password\">{{ 'AbpAccount::ConfirmPassword' | abpLocalization }}</label\n    ><span> * </span>\n    <input\n      type=\"password\"\n      id=\"input-confirm-password\"\n      class=\"form-control\"\n      formControlName=\"confirmPassword\"\n    />\n  </div>\n  <button class=\"mr-2 btn btn-secondary\" type=\"button\" routerLink=\"/account/login\">\n    {{ 'AbpAccount::Cancel' | abpLocalization }}\n  </button>\n  <abp-button\n    buttonType=\"submit\"\n    buttonClass=\"mr-2 btn btn-primary\"\n    [loading]=\"inProgress\"\n    (click)=\"onSubmit()\"\n  >\n    {{ 'AbpAccount::Submit' | abpLocalization }}\n  </abp-button>\n</form>\n\n<ng-template #passwordResetTemplate>\n  <p>\n    {{ 'AbpAccount::YourPasswordIsSuccessfullyReset' | abpLocalization }}\n  </p>\n\n  <a routerLink=\"/account/login\">\n    <button class=\"d-block mt-2 mb-3 btn btn-primary\">\n      {{ 'AbpAccount::BackToLogin' | abpLocalization }}\n    </button>\n  </a>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc2.FormBuilder }, { type: AccountService$1 }, { type: ɵngcc6.ActivatedRoute }, { type: ɵngcc6.Router }, { type: ɵngcc0.Injector }]; }, null); })();

class AuthenticationFlowGuard {
    constructor(authService) {
        this.authService = authService;
    }
    canActivate() {
        if (this.authService.isInternalAuth)
            return true;
        this.authService.navigateToLogin();
        return false;
    }
}
AuthenticationFlowGuard.ɵfac = function AuthenticationFlowGuard_Factory(t) { return new (t || AuthenticationFlowGuard)(ɵngcc0.ɵɵinject(ɵngcc1.AuthService)); };
AuthenticationFlowGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: AuthenticationFlowGuard, factory: AuthenticationFlowGuard.ɵfac });
AuthenticationFlowGuard.ctorParameters = () => [
    { type: AuthService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AuthenticationFlowGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.AuthService }]; }, null); })();

const ɵ0 = {
    replaceableComponent: {
        key: "Account.LoginComponent" /* Login */,
        defaultComponent: LoginComponent,
    },
}, ɵ1 = {
    replaceableComponent: {
        key: "Account.RegisterComponent" /* Register */,
        defaultComponent: RegisterComponent,
    },
}, ɵ2 = {
    replaceableComponent: {
        key: "Account.ForgotPasswordComponent" /* ForgotPassword */,
        defaultComponent: ForgotPasswordComponent,
    },
}, ɵ3 = {
    tenantBoxVisible: false,
    replaceableComponent: {
        key: "Account.ResetPasswordComponent" /* ResetPassword */,
        defaultComponent: ResetPasswordComponent,
    },
}, ɵ4 = {
    replaceableComponent: {
        key: "Account.ManageProfileComponent" /* ManageProfile */,
        defaultComponent: ManageProfileComponent,
    },
};
const routes = [
    { path: '', pathMatch: 'full', redirectTo: 'login' },
    {
        path: '',
        component: DynamicLayoutComponent,
        children: [
            {
                path: 'login',
                component: ReplaceableRouteContainerComponent,
                canActivate: [AuthenticationFlowGuard],
                data: ɵ0,
            },
            {
                path: 'register',
                component: ReplaceableRouteContainerComponent,
                canActivate: [AuthenticationFlowGuard],
                data: ɵ1,
            },
            {
                path: 'forgot-password',
                component: ReplaceableRouteContainerComponent,
                canActivate: [AuthenticationFlowGuard],
                data: ɵ2,
            },
            {
                path: 'reset-password',
                component: ReplaceableRouteContainerComponent,
                canActivate: [],
                data: ɵ3,
            },
            {
                path: 'manage',
                component: ReplaceableRouteContainerComponent,
                canActivate: [AuthGuard],
                data: ɵ4,
            },
        ],
    },
];
class AccountRoutingModule {
}
AccountRoutingModule.ɵfac = function AccountRoutingModule_Factory(t) { return new (t || AccountRoutingModule)(); };
AccountRoutingModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AccountRoutingModule });
AccountRoutingModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[RouterModule.forChild(routes)], RouterModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AccountRoutingModule, [{
        type: NgModule,
        args: [{
                imports: [RouterModule.forChild(routes)],
                exports: [RouterModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AccountRoutingModule, { imports: function () { return [ɵngcc6.RouterModule]; }, exports: function () { return [RouterModule]; } }); })();

const { required: required$1 } = Validators;
const PASSWORD_FIELDS = ['newPassword', 'repeatNewPassword'];
class ChangePasswordComponent {
    constructor(fb, injector, toasterService, profileService, manageProfileState) {
        this.fb = fb;
        this.injector = injector;
        this.toasterService = toasterService;
        this.profileService = profileService;
        this.manageProfileState = manageProfileState;
        this.mapErrorsFn = (errors, groupErrors, control) => {
            if (PASSWORD_FIELDS.indexOf(String(control.name)) < 0)
                return errors;
            return errors.concat(groupErrors.filter(({ key }) => key === 'passwordMismatch'));
        };
    }
    ngOnInit() {
        var _a;
        this.hideCurrentPassword = !((_a = this.manageProfileState.getProfile()) === null || _a === void 0 ? void 0 : _a.hasPassword);
        const passwordValidations = getPasswordValidators(this.injector);
        this.form = this.fb.group({
            password: ['', required$1],
            newPassword: [
                '',
                {
                    validators: [required$1, ...passwordValidations],
                },
            ],
            repeatNewPassword: [
                '',
                {
                    validators: [required$1, ...passwordValidations],
                },
            ],
        }, {
            validators: [comparePasswords(PASSWORD_FIELDS)],
        });
        if (this.hideCurrentPassword)
            this.form.removeControl('password');
    }
    onSubmit() {
        if (this.form.invalid)
            return;
        this.inProgress = true;
        this.profileService
            .changePassword(Object.assign(Object.assign({}, (!this.hideCurrentPassword && { currentPassword: this.form.get('password').value })), { newPassword: this.form.get('newPassword').value }))
            .pipe(finalize(() => (this.inProgress = false)))
            .subscribe({
            next: () => {
                this.form.reset();
                this.toasterService.success('AbpAccount::PasswordChangedMessage', '', {
                    life: 5000,
                });
                if (this.hideCurrentPassword) {
                    this.hideCurrentPassword = false;
                    this.form.addControl('password', new FormControl('', [required$1]));
                }
            },
            error: err => {
                this.toasterService.error(snq(() => err.error.error.message, 'AbpAccount::DefaultErrorMessage'));
            },
        });
    }
}
ChangePasswordComponent.ɵfac = function ChangePasswordComponent_Factory(t) { return new (t || ChangePasswordComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.ToasterService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ProfileService), ɵngcc0.ɵɵdirectiveInject(ManageProfileStateService)); };
ChangePasswordComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ChangePasswordComponent, selectors: [["abp-change-password-form"]], exportAs: ["abpChangePasswordForm"], decls: 19, vars: 14, consts: [["validateOnSubmit", "", 3, "formGroup", "mapErrorsFn", "ngSubmit"], ["class", "form-group", 4, "ngIf"], [1, "form-group"], ["for", "new-password"], ["type", "password", "id", "new-password", "formControlName", "newPassword", "autocomplete", "new-password", 1, "form-control"], ["for", "confirm-new-password"], ["type", "password", "id", "confirm-new-password", "formControlName", "repeatNewPassword", "autocomplete", "new-password", 1, "form-control"], ["iconClass", "fa fa-check", "buttonClass", "btn btn-primary color-white", "buttonType", "submit", 3, "loading", "disabled"], ["for", "current-password"], ["type", "password", "id", "current-password", "formControlName", "password", "autofocus", "", "autocomplete", "current-password", 1, "form-control"]], template: function ChangePasswordComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0);
        ɵngcc0.ɵɵlistener("ngSubmit", function ChangePasswordComponent_Template_form_ngSubmit_0_listener() { return ctx.onSubmit(); });
        ɵngcc0.ɵɵtemplate(1, ChangePasswordComponent_div_1_Template, 7, 3, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "label", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "span");
        ɵngcc0.ɵɵtext(7, " * ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(8, "input", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 2);
        ɵngcc0.ɵɵelementStart(10, "label", 5);
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵpipe(12, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "span");
        ɵngcc0.ɵɵtext(14, " * ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(15, "input", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "abp-button", 7);
        ɵngcc0.ɵɵtext(17);
        ɵngcc0.ɵɵpipe(18, "abpLocalization");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("formGroup", ctx.form)("mapErrorsFn", ctx.mapErrorsFn);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.hideCurrentPassword);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 8, "AbpIdentity::DisplayName:NewPassword"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(12, 10, "AbpIdentity::DisplayName:NewPasswordConfirm"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("loading", ctx.inProgress)("disabled", ctx.form == null ? null : ctx.form.invalid);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(18, 12, "AbpIdentity::Save"));
    } }, directives: [ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc1.FormSubmitDirective, ɵngcc2.FormGroupDirective, ɵngcc4.ValidationGroupDirective, ɵngcc3.NgIf, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc4.ValidationDirective, ɵngcc5.ButtonComponent, ɵngcc1.AutofocusDirective], pipes: [ɵngcc1.LocalizationPipe], encapsulation: 2 });
ChangePasswordComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: Injector },
    { type: ToasterService },
    { type: ProfileService },
    { type: ManageProfileStateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ChangePasswordComponent, [{
        type: Component,
        args: [{
                selector: 'abp-change-password-form',
                template: "<form [formGroup]=\"form\" (ngSubmit)=\"onSubmit()\" [mapErrorsFn]=\"mapErrorsFn\" validateOnSubmit>\n  <div *ngIf=\"!hideCurrentPassword\" class=\"form-group\">\n    <label for=\"current-password\">{{\n      'AbpIdentity::DisplayName:CurrentPassword' | abpLocalization\n    }}</label\n    ><span> * </span\n    ><input\n      type=\"password\"\n      id=\"current-password\"\n      class=\"form-control\"\n      formControlName=\"password\"\n      autofocus\n      autocomplete=\"current-password\"\n    />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"new-password\">{{ 'AbpIdentity::DisplayName:NewPassword' | abpLocalization }}</label\n    ><span> * </span\n    ><input\n      type=\"password\"\n      id=\"new-password\"\n      class=\"form-control\"\n      formControlName=\"newPassword\"\n      autocomplete=\"new-password\"\n    />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"confirm-new-password\">{{\n      'AbpIdentity::DisplayName:NewPasswordConfirm' | abpLocalization\n    }}</label\n    ><span> * </span\n    ><input\n      type=\"password\"\n      id=\"confirm-new-password\"\n      class=\"form-control\"\n      formControlName=\"repeatNewPassword\"\n      autocomplete=\"new-password\"\n    />\n  </div>\n  <abp-button\n    iconClass=\"fa fa-check\"\n    buttonClass=\"btn btn-primary color-white\"\n    buttonType=\"submit\"\n    [loading]=\"inProgress\"\n    [disabled]=\"form?.invalid\"\n    >{{ 'AbpIdentity::Save' | abpLocalization }}</abp-button\n  >\n</form>\n",
                exportAs: 'abpChangePasswordForm'
            }]
    }], function () { return [{ type: ɵngcc2.FormBuilder }, { type: ɵngcc0.Injector }, { type: ɵngcc5.ToasterService }, { type: ɵngcc1.ProfileService }, { type: ManageProfileStateService }]; }, null); })();

const { maxLength, required, email } = Validators;
class PersonalSettingsComponent {
    constructor(fb, toasterService, profileService, manageProfileState) {
        this.fb = fb;
        this.toasterService = toasterService;
        this.profileService = profileService;
        this.manageProfileState = manageProfileState;
    }
    ngOnInit() {
        this.buildForm();
    }
    buildForm() {
        const profile = this.manageProfileState.getProfile();
        this.form = this.fb.group({
            userName: [profile.userName, [required, maxLength(256)]],
            email: [profile.email, [required, email, maxLength(256)]],
            name: [profile.name || '', [maxLength(64)]],
            surname: [profile.surname || '', [maxLength(64)]],
            phoneNumber: [profile.phoneNumber || '', [maxLength(16)]],
        });
    }
    submit() {
        if (this.form.invalid)
            return;
        this.inProgress = true;
        this.profileService
            .update(this.form.value)
            .pipe(finalize(() => (this.inProgress = false)))
            .subscribe(profile => {
            this.manageProfileState.setProfile(profile);
            this.toasterService.success('AbpAccount::PersonalSettingsSaved', 'Success', { life: 5000 });
        });
    }
}
PersonalSettingsComponent.ɵfac = function PersonalSettingsComponent_Factory(t) { return new (t || PersonalSettingsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.ToasterService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ProfileService), ɵngcc0.ɵɵdirectiveInject(ManageProfileStateService)); };
PersonalSettingsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PersonalSettingsComponent, selectors: [["abp-personal-settings-form"]], exportAs: ["abpPersonalSettingsForm"], decls: 1, vars: 1, consts: [["validateOnSubmit", "", 3, "formGroup", "ngSubmit", 4, "ngIf"], ["validateOnSubmit", "", 3, "formGroup", "ngSubmit"], [1, "form-group"], ["for", "username"], ["type", "text", "id", "username", "formControlName", "userName", "autofocus", "", 1, "form-control", 3, "keydown.space"], [1, "row"], [1, "col", "col-md-6"], ["for", "name"], ["type", "text", "id", "name", "formControlName", "name", 1, "form-control"], ["for", "surname"], ["type", "text", "id", "surname", "formControlName", "surname", 1, "form-control"], ["for", "email-address"], ["type", "text", "id", "email-address", "formControlName", "email", 1, "form-control"], ["for", "phone-number"], ["type", "text", "id", "phone-number", "formControlName", "phoneNumber", 1, "form-control"], ["buttonType", "submit", "iconClass", "fa fa-check", "buttonClass", "btn btn-primary color-white", 3, "loading", "disabled"]], template: function PersonalSettingsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, PersonalSettingsComponent_form_0_Template, 36, 21, "form", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.form);
    } }, directives: [ɵngcc3.NgIf, ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc1.FormSubmitDirective, ɵngcc2.FormGroupDirective, ɵngcc4.ValidationGroupDirective, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc4.ValidationDirective, ɵngcc1.AutofocusDirective, ɵngcc5.ButtonComponent], pipes: [ɵngcc1.LocalizationPipe], encapsulation: 2 });
PersonalSettingsComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: ToasterService },
    { type: ProfileService },
    { type: ManageProfileStateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PersonalSettingsComponent, [{
        type: Component,
        args: [{
                selector: 'abp-personal-settings-form',
                template: "<form validateOnSubmit *ngIf=\"form\" [formGroup]=\"form\" (ngSubmit)=\"submit()\">\n  <div class=\"form-group\">\n    <label for=\"username\">{{ 'AbpIdentity::DisplayName:UserName' | abpLocalization }}</label\n    ><span> * </span\n    ><input\n      type=\"text\"\n      id=\"username\"\n      class=\"form-control\"\n      formControlName=\"userName\"\n      autofocus\n      (keydown.space)=\"$event.preventDefault()\"\n    />\n  </div>\n  <div class=\"row\">\n    <div class=\"col col-md-6\">\n      <div class=\"form-group\">\n        <label for=\"name\">{{ 'AbpIdentity::DisplayName:Name' | abpLocalization }}</label\n        ><input type=\"text\" id=\"name\" class=\"form-control\" formControlName=\"name\" />\n      </div>\n    </div>\n    <div class=\"col col-md-6\">\n      <div class=\"form-group\">\n        <label for=\"surname\">{{ 'AbpIdentity::DisplayName:Surname' | abpLocalization }}</label\n        ><input type=\"text\" id=\"surname\" class=\"form-control\" formControlName=\"surname\" />\n      </div>\n    </div>\n  </div>\n  <div class=\"form-group\">\n    <label for=\"email-address\">{{ 'AbpIdentity::DisplayName:Email' | abpLocalization }}</label\n    ><span> * </span\n    ><input type=\"text\" id=\"email-address\" class=\"form-control\" formControlName=\"email\" />\n  </div>\n  <div class=\"form-group\">\n    <label for=\"phone-number\">{{ 'AbpIdentity::DisplayName:PhoneNumber' | abpLocalization }}</label\n    ><input type=\"text\" id=\"phone-number\" class=\"form-control\" formControlName=\"phoneNumber\" />\n  </div>\n  <abp-button\n    buttonType=\"submit\"\n    iconClass=\"fa fa-check\"\n    buttonClass=\"btn btn-primary color-white\"\n    [loading]=\"inProgress\"\n    [disabled]=\"form?.invalid\"\n  >\n    {{ 'AbpIdentity::Save' | abpLocalization }}</abp-button\n  >\n</form>\n",
                exportAs: 'abpPersonalSettingsForm'
            }]
    }], function () { return [{ type: ɵngcc2.FormBuilder }, { type: ɵngcc5.ToasterService }, { type: ɵngcc1.ProfileService }, { type: ManageProfileStateService }]; }, null); })();

function accountConfigOptionsFactory(options) {
    return Object.assign({ redirectUrl: '/' }, options);
}

const declarations = [
    LoginComponent,
    RegisterComponent,
    ChangePasswordComponent,
    ManageProfileComponent,
    PersonalSettingsComponent,
    ForgotPasswordComponent,
    ResetPasswordComponent,
];
class AccountModule {
    static forChild(options = {}) {
        return {
            ngModule: AccountModule,
            providers: [
                AuthenticationFlowGuard,
                { provide: ACCOUNT_CONFIG_OPTIONS, useValue: options },
                {
                    provide: 'ACCOUNT_OPTIONS',
                    useFactory: accountConfigOptionsFactory,
                    deps: [ACCOUNT_CONFIG_OPTIONS],
                },
            ],
        };
    }
    static forLazy(options = {}) {
        return new LazyModuleFactory(AccountModule.forChild(options));
    }
}
AccountModule.ɵfac = function AccountModule_Factory(t) { return new (t || AccountModule)(); };
AccountModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AccountModule });
AccountModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CoreModule,
            AccountRoutingModule,
            ThemeSharedModule,
            NgbDropdownModule,
            NgxValidateCoreModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AccountModule, [{
        type: NgModule,
        args: [{
                declarations: [...declarations],
                imports: [
                    CoreModule,
                    AccountRoutingModule,
                    ThemeSharedModule,
                    NgbDropdownModule,
                    NgxValidateCoreModule,
                ],
                exports: [...declarations]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AccountModule, { declarations: function () { return [LoginComponent, RegisterComponent, ChangePasswordComponent, ManageProfileComponent, PersonalSettingsComponent, ForgotPasswordComponent, ResetPasswordComponent]; }, imports: function () { return [CoreModule, AccountRoutingModule, ThemeSharedModule,
        NgbDropdownModule,
        NgxValidateCoreModule]; }, exports: function () { return [LoginComponent, RegisterComponent, ChangePasswordComponent, ManageProfileComponent, PersonalSettingsComponent, ForgotPasswordComponent, ResetPasswordComponent]; } }); })();

var LoginResultType;
(function (LoginResultType) {
    LoginResultType[LoginResultType["Success"] = 1] = "Success";
    LoginResultType[LoginResultType["InvalidUserNameOrPassword"] = 2] = "InvalidUserNameOrPassword";
    LoginResultType[LoginResultType["NotAllowed"] = 3] = "NotAllowed";
    LoginResultType[LoginResultType["LockedOut"] = 4] = "LockedOut";
    LoginResultType[LoginResultType["RequiresTwoFactor"] = 5] = "RequiresTwoFactor";
})(LoginResultType || (LoginResultType = {}));
const loginResultTypeOptions = mapEnumToOptions(LoginResultType);

var index$4 = /*#__PURE__*/Object.freeze({
    __proto__: null,
    get LoginResultType () { return LoginResultType; },
    loginResultTypeOptions: loginResultTypeOptions
});

class AccountService {
    constructor(restService) {
        this.restService = restService;
        this.apiName = 'AbpAccount';
        this.checkPasswordByLogin = (login) => this.restService.request({
            method: 'POST',
            url: '/api/account/check-password',
            body: login,
        }, { apiName: this.apiName });
        this.loginByLogin = (login) => this.restService.request({
            method: 'POST',
            url: '/api/account/login',
            body: login,
        }, { apiName: this.apiName });
        this.logout = () => this.restService.request({
            method: 'GET',
            url: '/api/account/logout',
        }, { apiName: this.apiName });
    }
}
AccountService.ɵfac = function AccountService_Factory(t) { return new (t || AccountService)(ɵngcc0.ɵɵinject(ɵngcc1.RestService)); };
AccountService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AccountService_Factory() { return new AccountService(i0.ɵɵinject(i1.RestService)); }, token: AccountService, providedIn: "root" });
AccountService.ctorParameters = () => [
    { type: RestService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AccountService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.RestService }]; }, null); })();

var index$3 = /*#__PURE__*/Object.freeze({
    __proto__: null,
    Models: index$4,
    AccountService: AccountService
});

var index$2 = /*#__PURE__*/Object.freeze({
    __proto__: null,
    Controllers: index$3
});

var index$1 = /*#__PURE__*/Object.freeze({
    __proto__: null,
    Account: index$2
});

var index = /*#__PURE__*/Object.freeze({
    __proto__: null,
    Areas: index$1
});

/**
 * Generated bundle index. Do not edit.
 */

export { ACCOUNT_CONFIG_OPTIONS, AccountModule, AccountService$1 as AccountService, AuthenticationFlowGuard, ChangePasswordComponent, ForgotPasswordComponent, LoginComponent, ManageProfileComponent, ManageProfileStateService, PersonalSettingsComponent, RegisterComponent, ResetPasswordComponent, index as Web, LoginComponent as ɵa, RegisterComponent as ɵb, AccountService$1 as ɵc, ChangePasswordComponent as ɵd, ManageProfileStateService as ɵe, ManageProfileComponent as ɵf, PersonalSettingsComponent as ɵg, ForgotPasswordComponent as ɵh, ResetPasswordComponent as ɵi, AccountRoutingModule as ɵj, AuthenticationFlowGuard as ɵk, ACCOUNT_CONFIG_OPTIONS as ɵm, accountConfigOptionsFactory as ɵn };
