// @abp/ng.components/esm2015/page/page-part.directive.js
import { Directive, TemplateRef, ViewContainerRef, Input, InjectionToken, Optional, Inject, Injector, } from '@angular/core';
import { Observable, of } from 'rxjs';
export const PAGE_RENDER_STRATEGY = new InjectionToken('PAGE_RENDER_STRATEGY');
export class PagePartDirective {
    constructor(templateRef, viewContainer, renderLogic, injector) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.renderLogic = renderLogic;
        this.injector = injector;
        this.hasRendered = false;
        this.render = (shouldRender) => {
            if (shouldRender && !this.hasRendered) {
                this.viewContainer.createEmbeddedView(this.templateRef);
                this.hasRendered = true;
            }
            else if (!shouldRender && this.hasRendered) {
                this.viewContainer.clear();
                this.hasRendered = false;
            }
        };
    }
    set abpPagePart(type) {
        this.type = type;
        this.createRenderStream(type);
    }
    ngOnChanges({ context }) {
        var _a;
        if ((_a = this.renderLogic) === null || _a === void 0 ? void 0 : _a.onContextUpdate) {
            this.renderLogic.onContextUpdate(context);
        }
    }
    ngOnInit() {
        var _a;
        if ((_a = this.renderLogic) === null || _a === void 0 ? void 0 : _a.onInit) {
            this.renderLogic.onInit(this.type, this.injector, this.context);
        }
    }
    ngOnDestroy() {
        var _a;
        this.clearSubscription();
        if ((_a = this.renderLogic) === null || _a === void 0 ? void 0 : _a.onDestroy) {
            this.renderLogic.onDestroy(this.type, this.injector, this.context);
        }
    }
    shouldRender(type) {
        if (this.renderLogic) {
            const willRender = this.renderLogic.shouldRender(type);
            return willRender instanceof Observable ? willRender : of(willRender);
        }
        return of(true);
    }
    createRenderStream(type) {
        this.clearSubscription();
        this.subscription = this.shouldRender(type).subscribe(this.render);
    }
    clearSubscription() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
}
PagePartDirective.decorators = [
    { type: Directive, args: [{ selector: '[abpPagePart]' },] }
];
PagePartDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [PAGE_RENDER_STRATEGY,] }] },
    { type: Injector }
];
PagePartDirective.propDecorators = {
    context: [{ type: Input, args: ['abpPagePartContext',] }],
    abpPagePart: [{ type: Input }]
};
