// @abp/ng.components/esm2015/page/page-parts.component.js
import { Component, ViewEncapsulation } from '@angular/core';
export var PageParts;
(function (PageParts) {
    PageParts["title"] = "PageTitleContainerComponent";
    PageParts["breadcrumb"] = "PageBreadcrumbContainerComponent";
    PageParts["toolbar"] = "PageToolbarContainerComponent";
})(PageParts || (PageParts = {}));
export class PageTitleContainerComponent {
}
PageTitleContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'abp-page-title-container',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
export class PageBreadcrumbContainerComponent {
}
PageBreadcrumbContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'abp-page-breadcrumb-container',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
export class PageToolbarContainerComponent {
}
PageToolbarContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'abp-page-toolbar-container',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
