// @abp/ng.components/esm2015/page/page.component.js
import { Component, Input, ViewEncapsulation, ContentChild } from '@angular/core';
import { PageTitleContainerComponent, PageBreadcrumbContainerComponent, PageToolbarContainerComponent, PageParts, } from './page-parts.component';
export class PageComponent {
    constructor() {
        this.toolbarVisible = false;
        this.breadcrumbVisible = true;
        this.pageParts = {
            title: PageParts.title,
            breadcrumb: PageParts.breadcrumb,
            toolbar: PageParts.toolbar,
        };
    }
    set toolbarData(val) {
        this._toolbarData = val;
        this.toolbarVisible = true;
    }
    get toolbarData() {
        return this._toolbarData;
    }
    get shouldRenderRow() {
        return !!(this.title ||
            this.toolbarVisible ||
            this.breadcrumbVisible ||
            this.customTitle ||
            this.customBreadcrumb ||
            this.customToolbar);
    }
}
PageComponent.decorators = [
    { type: Component, args: [{
                selector: 'abp-page',
                template: "<div class=\"row entry-row\" *ngIf=\"shouldRenderRow\">\n  <ng-container *ngIf=\"customTitle; else defaultTitleTemplate\">\n    <ng-content select=\"abp-page-title-container\"></ng-content>\n  </ng-container>\n\n  <ng-container *ngIf=\"customBreadcrumb; else defaultBreadcrumbTemplate\">\n    <ng-content select=\"abp-page-breadcrumb-container\"></ng-content>\n  </ng-container>\n\n  <ng-container *ngIf=\"customToolbar; else defaultPageToolbarTemplate\">\n    <ng-content select=\"abp-page-toolbar-container\"></ng-content>\n  </ng-container>\n</div>\n\n<ng-content></ng-content>\n\n<ng-template #defaultTitleTemplate>\n  <ng-container *ngIf=\"title\">\n    <div class=\"col-auto\" *abpPagePart=\"pageParts.title\">\n      <h1 class=\"content-header-title\">\n        {{ title }}\n      </h1>\n    </div>\n  </ng-container>\n</ng-template>\n\n<ng-template #defaultBreadcrumbTemplate>\n  <ng-container *ngIf=\"breadcrumbVisible\">\n    <div class=\"col-lg-auto pl-lg-0\" *abpPagePart=\"pageParts.breadcrumb\">\n      <abp-breadcrumb></abp-breadcrumb>\n    </div>\n  </ng-container>\n</ng-template>\n\n<ng-template #defaultPageToolbarTemplate>\n  <ng-container *ngIf=\"toolbarVisible\">\n    <div class=\"col\" *abpPagePart=\"pageParts.toolbar; context: toolbarData\">\n      <abp-page-toolbar [record]=\"toolbarData\"></abp-page-toolbar>\n    </div>\n  </ng-container>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None
            },] }
];
PageComponent.propDecorators = {
    title: [{ type: Input }],
    toolbarData: [{ type: Input, args: ['toolbar',] }],
    breadcrumbVisible: [{ type: Input, args: ['breadcrumb',] }],
    customTitle: [{ type: ContentChild, args: [PageTitleContainerComponent,] }],
    customBreadcrumb: [{ type: ContentChild, args: [PageBreadcrumbContainerComponent,] }],
    customToolbar: [{ type: ContentChild, args: [PageToolbarContainerComponent,] }]
};
