// @abp/ng.setting-management/esm2015/config/proxy/email-settings.service.js
import { RestService } from '@abp/ng.core';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@abp/ng.core";
export class EmailSettingsService {
    constructor(restService) {
        this.restService = restService;
        this.apiName = 'SettingManagement';
        this.get = () => this.restService.request({
            method: 'GET',
            url: '/api/setting-management/emailing',
        }, { apiName: this.apiName });
        this.update = (input) => this.restService.request({
            method: 'POST',
            url: '/api/setting-management/emailing',
            body: input,
        }, { apiName: this.apiName });
    }
}
EmailSettingsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function EmailSettingsService_Factory() { return new EmailSettingsService(i0.ɵɵinject(i1.RestService)); }, token: EmailSettingsService, providedIn: "root" });
EmailSettingsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
EmailSettingsService.ctorParameters = () => [
    { type: RestService }
];
