// @abp/ng.identity/__ivy_ngcc__/fesm2015/abp-ng.identity-config.js
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { RoutesService } from '@abp/ng.core';

import * as ɵngcc0 from '@angular/core';
const IDENTITY_ROUTE_PROVIDERS = [
    { provide: APP_INITIALIZER, useFactory: configureRoutes, deps: [RoutesService], multi: true },
];
function configureRoutes(routesService) {
    return () => {
        routesService.add([
            {
                path: undefined,
                name: "AbpIdentity::Menu:IdentityManagement" /* IdentityManagement */,
                parentName: "AbpUiNavigation::Menu:Administration" /* Administration */,
                requiredPolicy: "AbpIdentity.Roles || AbpIdentity.Users" /* IdentityManagement */,
                iconClass: 'fa fa-id-card-o',
                layout: "application" /* application */,
                order: 1,
            },
            {
                path: '/identity/roles',
                name: "AbpIdentity::Roles" /* Roles */,
                parentName: "AbpIdentity::Menu:IdentityManagement" /* IdentityManagement */,
                requiredPolicy: "AbpIdentity.Roles" /* Roles */,
                order: 1,
            },
            {
                path: '/identity/users',
                name: "AbpIdentity::Users" /* Users */,
                parentName: "AbpIdentity::Menu:IdentityManagement" /* IdentityManagement */,
                requiredPolicy: "AbpIdentity.Users" /* Users */,
                order: 2,
            },
        ]);
    };
}

class IdentityConfigModule {
    static forRoot() {
        return {
            ngModule: IdentityConfigModule,
            providers: [IDENTITY_ROUTE_PROVIDERS],
        };
    }
}
IdentityConfigModule.ɵfac = function IdentityConfigModule_Factory(t) { return new (t || IdentityConfigModule)(); };
IdentityConfigModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: IdentityConfigModule });
IdentityConfigModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(IdentityConfigModule, [{
        type: NgModule
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { IDENTITY_ROUTE_PROVIDERS, IdentityConfigModule, configureRoutes, IDENTITY_ROUTE_PROVIDERS as ɵa, configureRoutes as ɵb };
