// @abp/ng.components/__ivy_ngcc__/fesm2015/abp-ng.components-page.js
import { Component, ViewEncapsulation, Input, ContentChild, InjectionToken, Directive, TemplateRef, ViewContainerRef, Optional, Inject, Injector, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { UiExtensionsModule } from '@abp/ng.theme.shared/extensions';
import { ThemeSharedModule } from '@abp/ng.theme.shared';
import { CoreModule } from '@abp/ng.core';
import { Observable, of } from 'rxjs';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@abp/ng.theme.shared';
import * as ɵngcc3 from '@abp/ng.theme.shared/extensions';

const _c0 = ["*"];
function PageComponent_div_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵprojection(1, 1);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function PageComponent_div_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵprojection(1, 2);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function PageComponent_div_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵprojection(1, 3);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function PageComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵtemplate(1, PageComponent_div_0_ng_container_1_Template, 2, 0, "ng-container", 5);
    ɵngcc0.ɵɵtemplate(2, PageComponent_div_0_ng_container_2_Template, 2, 0, "ng-container", 5);
    ɵngcc0.ɵɵtemplate(3, PageComponent_div_0_ng_container_3_Template, 2, 0, "ng-container", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    const _r1 = ɵngcc0.ɵɵreference(3);
    const _r3 = ɵngcc0.ɵɵreference(5);
    const _r5 = ɵngcc0.ɵɵreference(7);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customTitle)("ngIfElse", _r1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customBreadcrumb)("ngIfElse", _r3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customToolbar)("ngIfElse", _r5);
} }
function PageComponent_ng_template_2_ng_container_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵelementStart(1, "h1", 9);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r11.title, " ");
} }
function PageComponent_ng_template_2_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, PageComponent_ng_template_2_ng_container_0_div_1_Template, 3, 1, "div", 7);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("abpPagePart", ctx_r10.pageParts.title);
} }
function PageComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, PageComponent_ng_template_2_ng_container_0_Template, 2, 1, "ng-container", 6);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.title);
} }
function PageComponent_ng_template_4_ng_container_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "abp-breadcrumb");
    ɵngcc0.ɵɵelementEnd();
} }
function PageComponent_ng_template_4_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, PageComponent_ng_template_4_ng_container_0_div_1_Template, 2, 0, "div", 10);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("abpPagePart", ctx_r12.pageParts.breadcrumb);
} }
function PageComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, PageComponent_ng_template_4_ng_container_0_Template, 2, 1, "ng-container", 6);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.breadcrumbVisible);
} }
function PageComponent_ng_template_6_ng_container_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelement(1, "abp-page-toolbar", 14);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("record", ctx_r15.toolbarData);
} }
function PageComponent_ng_template_6_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, PageComponent_ng_template_6_ng_container_0_div_1_Template, 2, 1, "div", 12);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("abpPagePart", ctx_r14.pageParts.toolbar)("abpPagePartContext", ctx_r14.toolbarData);
} }
function PageComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, PageComponent_ng_template_6_ng_container_0_Template, 2, 2, "ng-container", 6);
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.toolbarVisible);
} }
const _c1 = ["*", [["abp-page-title-container"]], [["abp-page-breadcrumb-container"]], [["abp-page-toolbar-container"]]];
const _c2 = ["*", "abp-page-title-container", "abp-page-breadcrumb-container", "abp-page-toolbar-container"];
var PageParts;
(function (PageParts) {
    PageParts["title"] = "PageTitleContainerComponent";
    PageParts["breadcrumb"] = "PageBreadcrumbContainerComponent";
    PageParts["toolbar"] = "PageToolbarContainerComponent";
})(PageParts || (PageParts = {}));
class PageTitleContainerComponent {
}
PageTitleContainerComponent.ɵfac = function PageTitleContainerComponent_Factory(t) { return new (t || PageTitleContainerComponent)(); };
PageTitleContainerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PageTitleContainerComponent, selectors: [["abp-page-title-container"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function PageTitleContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PageTitleContainerComponent, [{
        type: Component,
        args: [{
                selector: 'abp-page-title-container',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            }]
    }], null, null); })();
class PageBreadcrumbContainerComponent {
}
PageBreadcrumbContainerComponent.ɵfac = function PageBreadcrumbContainerComponent_Factory(t) { return new (t || PageBreadcrumbContainerComponent)(); };
PageBreadcrumbContainerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PageBreadcrumbContainerComponent, selectors: [["abp-page-breadcrumb-container"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function PageBreadcrumbContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PageBreadcrumbContainerComponent, [{
        type: Component,
        args: [{
                selector: 'abp-page-breadcrumb-container',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            }]
    }], null, null); })();
class PageToolbarContainerComponent {
}
PageToolbarContainerComponent.ɵfac = function PageToolbarContainerComponent_Factory(t) { return new (t || PageToolbarContainerComponent)(); };
PageToolbarContainerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PageToolbarContainerComponent, selectors: [["abp-page-toolbar-container"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function PageToolbarContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PageToolbarContainerComponent, [{
        type: Component,
        args: [{
                selector: 'abp-page-toolbar-container',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            }]
    }], null, null); })();

class PageComponent {
    constructor() {
        this.toolbarVisible = false;
        this.breadcrumbVisible = true;
        this.pageParts = {
            title: PageParts.title,
            breadcrumb: PageParts.breadcrumb,
            toolbar: PageParts.toolbar,
        };
    }
    set toolbarData(val) {
        this._toolbarData = val;
        this.toolbarVisible = true;
    }
    get toolbarData() {
        return this._toolbarData;
    }
    get shouldRenderRow() {
        return !!(this.title ||
            this.toolbarVisible ||
            this.breadcrumbVisible ||
            this.customTitle ||
            this.customBreadcrumb ||
            this.customToolbar);
    }
}
PageComponent.ɵfac = function PageComponent_Factory(t) { return new (t || PageComponent)(); };
PageComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PageComponent, selectors: [["abp-page"]], contentQueries: function PageComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, PageTitleContainerComponent, 5);
        ɵngcc0.ɵɵcontentQuery(dirIndex, PageBreadcrumbContainerComponent, 5);
        ɵngcc0.ɵɵcontentQuery(dirIndex, PageToolbarContainerComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.customTitle = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.customBreadcrumb = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.customToolbar = _t.first);
    } }, inputs: { breadcrumbVisible: ["breadcrumb", "breadcrumbVisible"], toolbarData: ["toolbar", "toolbarData"], title: "title" }, ngContentSelectors: _c2, decls: 8, vars: 1, consts: [["class", "row entry-row", 4, "ngIf"], ["defaultTitleTemplate", ""], ["defaultBreadcrumbTemplate", ""], ["defaultPageToolbarTemplate", ""], [1, "row", "entry-row"], [4, "ngIf", "ngIfElse"], [4, "ngIf"], ["class", "col-auto", 4, "abpPagePart"], [1, "col-auto"], [1, "content-header-title"], ["class", "col-lg-auto pl-lg-0", 4, "abpPagePart"], [1, "col-lg-auto", "pl-lg-0"], ["class", "col", 4, "abpPagePart", "abpPagePartContext"], [1, "col"], [3, "record"]], template: function PageComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c1);
        ɵngcc0.ɵɵtemplate(0, PageComponent_div_0_Template, 4, 6, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵtemplate(2, PageComponent_ng_template_2_Template, 1, 1, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(4, PageComponent_ng_template_4_Template, 1, 1, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(6, PageComponent_ng_template_6_Template, 1, 1, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.shouldRenderRow);
    } }, directives: function () { return [ɵngcc1.NgIf, PagePartDirective, ɵngcc2.BreadcrumbComponent, ɵngcc3.PageToolbarComponent]; }, encapsulation: 2 });
PageComponent.propDecorators = {
    title: [{ type: Input }],
    toolbarData: [{ type: Input, args: ['toolbar',] }],
    breadcrumbVisible: [{ type: Input, args: ['breadcrumb',] }],
    customTitle: [{ type: ContentChild, args: [PageTitleContainerComponent,] }],
    customBreadcrumb: [{ type: ContentChild, args: [PageBreadcrumbContainerComponent,] }],
    customToolbar: [{ type: ContentChild, args: [PageToolbarContainerComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PageComponent, [{
        type: Component,
        args: [{
                selector: 'abp-page',
                template: "<div class=\"row entry-row\" *ngIf=\"shouldRenderRow\">\n  <ng-container *ngIf=\"customTitle; else defaultTitleTemplate\">\n    <ng-content select=\"abp-page-title-container\"></ng-content>\n  </ng-container>\n\n  <ng-container *ngIf=\"customBreadcrumb; else defaultBreadcrumbTemplate\">\n    <ng-content select=\"abp-page-breadcrumb-container\"></ng-content>\n  </ng-container>\n\n  <ng-container *ngIf=\"customToolbar; else defaultPageToolbarTemplate\">\n    <ng-content select=\"abp-page-toolbar-container\"></ng-content>\n  </ng-container>\n</div>\n\n<ng-content></ng-content>\n\n<ng-template #defaultTitleTemplate>\n  <ng-container *ngIf=\"title\">\n    <div class=\"col-auto\" *abpPagePart=\"pageParts.title\">\n      <h1 class=\"content-header-title\">\n        {{ title }}\n      </h1>\n    </div>\n  </ng-container>\n</ng-template>\n\n<ng-template #defaultBreadcrumbTemplate>\n  <ng-container *ngIf=\"breadcrumbVisible\">\n    <div class=\"col-lg-auto pl-lg-0\" *abpPagePart=\"pageParts.breadcrumb\">\n      <abp-breadcrumb></abp-breadcrumb>\n    </div>\n  </ng-container>\n</ng-template>\n\n<ng-template #defaultPageToolbarTemplate>\n  <ng-container *ngIf=\"toolbarVisible\">\n    <div class=\"col\" *abpPagePart=\"pageParts.toolbar; context: toolbarData\">\n      <abp-page-toolbar [record]=\"toolbarData\"></abp-page-toolbar>\n    </div>\n  </ng-container>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return []; }, { breadcrumbVisible: [{
            type: Input,
            args: ['breadcrumb']
        }], toolbarData: [{
            type: Input,
            args: ['toolbar']
        }], title: [{
            type: Input
        }], customTitle: [{
            type: ContentChild,
            args: [PageTitleContainerComponent]
        }], customBreadcrumb: [{
            type: ContentChild,
            args: [PageBreadcrumbContainerComponent]
        }], customToolbar: [{
            type: ContentChild,
            args: [PageToolbarContainerComponent]
        }] }); })();

const PAGE_RENDER_STRATEGY = new InjectionToken('PAGE_RENDER_STRATEGY');
class PagePartDirective {
    constructor(templateRef, viewContainer, renderLogic, injector) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.renderLogic = renderLogic;
        this.injector = injector;
        this.hasRendered = false;
        this.render = (shouldRender) => {
            if (shouldRender && !this.hasRendered) {
                this.viewContainer.createEmbeddedView(this.templateRef);
                this.hasRendered = true;
            }
            else if (!shouldRender && this.hasRendered) {
                this.viewContainer.clear();
                this.hasRendered = false;
            }
        };
    }
    set abpPagePart(type) {
        this.type = type;
        this.createRenderStream(type);
    }
    ngOnChanges({ context }) {
        var _a;
        if ((_a = this.renderLogic) === null || _a === void 0 ? void 0 : _a.onContextUpdate) {
            this.renderLogic.onContextUpdate(context);
        }
    }
    ngOnInit() {
        var _a;
        if ((_a = this.renderLogic) === null || _a === void 0 ? void 0 : _a.onInit) {
            this.renderLogic.onInit(this.type, this.injector, this.context);
        }
    }
    ngOnDestroy() {
        var _a;
        this.clearSubscription();
        if ((_a = this.renderLogic) === null || _a === void 0 ? void 0 : _a.onDestroy) {
            this.renderLogic.onDestroy(this.type, this.injector, this.context);
        }
    }
    shouldRender(type) {
        if (this.renderLogic) {
            const willRender = this.renderLogic.shouldRender(type);
            return willRender instanceof Observable ? willRender : of(willRender);
        }
        return of(true);
    }
    createRenderStream(type) {
        this.clearSubscription();
        this.subscription = this.shouldRender(type).subscribe(this.render);
    }
    clearSubscription() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
}
PagePartDirective.ɵfac = function PagePartDirective_Factory(t) { return new (t || PagePartDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(PAGE_RENDER_STRATEGY, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector)); };
PagePartDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: PagePartDirective, selectors: [["", "abpPagePart", ""]], inputs: { abpPagePart: "abpPagePart", context: ["abpPagePartContext", "context"] }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
PagePartDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [PAGE_RENDER_STRATEGY,] }] },
    { type: Injector }
];
PagePartDirective.propDecorators = {
    context: [{ type: Input, args: ['abpPagePartContext',] }],
    abpPagePart: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PagePartDirective, [{
        type: Directive,
        args: [{ selector: '[abpPagePart]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }, { type: ɵngcc0.ViewContainerRef }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [PAGE_RENDER_STRATEGY]
            }] }, { type: ɵngcc0.Injector }]; }, { abpPagePart: [{
            type: Input
        }], context: [{
            type: Input,
            args: ['abpPagePartContext']
        }] }); })();

const exportedDeclarations = [
    PageComponent,
    PageTitleContainerComponent,
    PageBreadcrumbContainerComponent,
    PageToolbarContainerComponent,
    PagePartDirective,
];
class PageModule {
}
PageModule.ɵfac = function PageModule_Factory(t) { return new (t || PageModule)(); };
PageModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: PageModule });
PageModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, UiExtensionsModule, CoreModule, ThemeSharedModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PageModule, [{
        type: NgModule,
        args: [{
                declarations: [...exportedDeclarations],
                imports: [CommonModule, UiExtensionsModule, CoreModule, ThemeSharedModule],
                exports: [...exportedDeclarations]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PageModule, { declarations: function () { return [PageComponent, PageTitleContainerComponent, PageBreadcrumbContainerComponent, PageToolbarContainerComponent, PagePartDirective]; }, imports: function () { return [CommonModule, UiExtensionsModule, CoreModule, ThemeSharedModule]; }, exports: function () { return [PageComponent, PageTitleContainerComponent, PageBreadcrumbContainerComponent, PageToolbarContainerComponent, PagePartDirective]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { PAGE_RENDER_STRATEGY, PageBreadcrumbContainerComponent, PageComponent, PageModule, PagePartDirective, PageParts, PageTitleContainerComponent, PageToolbarContainerComponent };
