// @abp/ng.theme.basic/bundles/abp-ng.theme.basic-testing.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@abp/ng.theme.basic'), require('@angular/core'), require('@ngx-validate/core')) :
    typeof define === 'function' && define.amd ? define('@abp/ng.theme.basic/testing', ['exports', '@abp/ng.theme.basic', '@angular/core', '@ngx-validate/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.abp = global.abp || {}, global.abp.ng = global.abp.ng || {}, global.abp.ng.theme = global.abp.ng.theme || {}, global.abp.ng.theme.basic = global.abp.ng.theme.basic || {}, global.abp.ng.theme.basic.testing = {}), global.abp.ng.theme.basic, global.ng.core, global.core));
}(this, (function (exports, ng_theme_basic, core$1, core) { 'use strict';

    var ThemeBasicTestingModule = /** @class */ (function () {
        function ThemeBasicTestingModule() {
        }
        ThemeBasicTestingModule.withConfig = function () {
            return {
                ngModule: ThemeBasicTestingModule,
                providers: [
                    ng_theme_basic.BASIC_THEME_NAV_ITEM_PROVIDERS,
                    ng_theme_basic.BASIC_THEME_STYLES_PROVIDERS,
                    {
                        provide: core.VALIDATION_ERROR_TEMPLATE,
                        useValue: ng_theme_basic.ValidationErrorComponent,
                    },
                    {
                        provide: core.VALIDATION_TARGET_SELECTOR,
                        useValue: '.form-group',
                    },
                    {
                        provide: core.VALIDATION_INVALID_CLASSES,
                        useValue: 'is-invalid',
                    },
                ],
            };
        };
        return ThemeBasicTestingModule;
    }());
    ThemeBasicTestingModule.decorators = [
        { type: core$1.NgModule, args: [{
                    exports: [ng_theme_basic.BaseThemeBasicModule],
                    imports: [ng_theme_basic.BaseThemeBasicModule],
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThemeBasicTestingModule = ThemeBasicTestingModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
