// @abp/ng.theme.basic/bundles/abp-ng.theme.basic.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@abp/ng.core'), require('rxjs'), require('rxjs/operators'), require('@abp/ng.account.core'), require('@abp/ng.theme.shared'), require('snq'), require('@ngx-validate/core'), require('@ng-bootstrap/ng-bootstrap')) :
    typeof define === 'function' && define.amd ? define('@abp/ng.theme.basic', ['exports', '@angular/core', '@abp/ng.core', 'rxjs', 'rxjs/operators', '@abp/ng.account.core', '@abp/ng.theme.shared', 'snq', '@ngx-validate/core', '@ng-bootstrap/ng-bootstrap'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.abp = global.abp || {}, global.abp.ng = global.abp.ng || {}, global.abp.ng.theme = global.abp.ng.theme || {}, global.abp.ng.theme.basic = {}), global.ng.core, global.ng_core, global.rxjs, global.rxjs.operators, global.ng_account_core, global.ng_theme_shared, global.snq, global.core$1, global.ngBootstrap));
}(this, (function (exports, core, ng_core, rxjs, operators, ng_account_core, ng_theme_shared, snq, core$1, ngBootstrap) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var snq__default = /*#__PURE__*/_interopDefaultLegacy(snq);

    var LayoutService = /** @class */ (function () {
        function LayoutService(subscription, cdRef) {
            this.subscription = subscription;
            this.cdRef = cdRef;
            this.isCollapsed = true;
            this.logoComponentKey = "Theme.LogoComponent" /* Logo */;
            this.routesComponentKey = "Theme.RoutesComponent" /* Routes */;
            this.navItemsComponentKey = "Theme.NavItemsComponent" /* NavItems */;
        }
        LayoutService.prototype.checkWindowWidth = function () {
            var _this = this;
            var isSmallScreen = window.innerWidth < 992;
            if (isSmallScreen && this.smallScreen === false) {
                this.isCollapsed = false;
                setTimeout(function () {
                    _this.isCollapsed = true;
                }, 100);
            }
            this.smallScreen = isSmallScreen;
            this.cdRef.detectChanges();
        };
        LayoutService.prototype.subscribeWindowSize = function () {
            var _this = this;
            this.checkWindowWidth();
            var resize$ = rxjs.fromEvent(window, 'resize').pipe(operators.debounceTime(150));
            this.subscription.addOne(resize$, function () { return _this.checkWindowWidth(); });
        };
        return LayoutService;
    }());
    LayoutService.decorators = [
        { type: core.Injectable }
    ];
    LayoutService.ctorParameters = function () { return [
        { type: ng_core.SubscriptionService },
        { type: core.ChangeDetectorRef }
    ]; };

    var AccountLayoutComponent = /** @class */ (function () {
        function AccountLayoutComponent(service) {
            this.service = service;
            this.authWrapperKey = 'Account.AuthWrapperComponent';
        }
        AccountLayoutComponent.prototype.ngAfterViewInit = function () {
            this.service.subscribeWindowSize();
        };
        return AccountLayoutComponent;
    }());
    // required for dynamic component
    AccountLayoutComponent.type = "account" /* account */;
    AccountLayoutComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-layout-account',
                    template: "<nav\n  class=\"navbar navbar-expand-lg navbar-dark bg-dark shadow-sm flex-column flex-md-row mb-4\"\n  id=\"main-navbar\"\n  style=\"min-height: 4rem\"\n>\n  <div class=\"container\">\n    <abp-logo *abpReplaceableTemplate=\"{ componentKey: service.logoComponentKey }\"></abp-logo>\n    <button\n      class=\"navbar-toggler\"\n      type=\"button\"\n      [attr.aria-expanded]=\"!service.isCollapsed\"\n      (click)=\"service.isCollapsed = !service.isCollapsed\"\n    >\n      <span class=\"navbar-toggler-icon\"></span>\n    </button>\n    <div\n      class=\"navbar-collapse\"\n      [class.overflow-hidden]=\"service.smallScreen\"\n      id=\"main-navbar-collapse\"\n    >\n      <ng-container *ngTemplateOutlet=\"!service.smallScreen ? navigations : null\"></ng-container>\n\n      <div\n        *ngIf=\"service.smallScreen\"\n        [@collapseWithMargin]=\"service.isCollapsed ? 'collapsed' : 'expanded'\"\n      >\n        <ng-container *ngTemplateOutlet=\"navigations\"></ng-container>\n      </div>\n\n      <ng-template #navigations>\n        <abp-routes\n          *abpReplaceableTemplate=\"{\n            componentKey: service.routesComponentKey,\n            inputs: {\n              smallScreen: { value: service.smallScreen }\n            }\n          }\"\n          class=\"mx-auto\"\n          [smallScreen]=\"service.smallScreen\"\n        ></abp-routes>\n\n        <abp-nav-items\n          *abpReplaceableTemplate=\"{\n            componentKey: service.navItemsComponentKey\n          }\"\n        ></abp-nav-items>\n      </ng-template>\n    </div>\n  </div>\n</nav>\n\n<!-- [@slideFromBottom]=\"outlet.isActivated && outlet.activatedRoute?.routeConfig?.path\" TODO: throws ExpressionChangedAfterItHasBeenCheck when animation is active. It should be fixed -->\n<div class=\"container\">\n  <abp-page-alert-container></abp-page-alert-container>\n  <abp-auth-wrapper\n    *abpReplaceableTemplate=\"{\n      componentKey: authWrapperKey\n    }\"\n  >\n    <router-outlet #outlet=\"outlet\"></router-outlet>\n  </abp-auth-wrapper>\n</div>\n",
                    providers: [LayoutService, ng_core.SubscriptionService]
                },] }
    ];
    AccountLayoutComponent.ctorParameters = function () { return [
        { type: LayoutService }
    ]; };

    var AuthWrapperComponent = /** @class */ (function () {
        function AuthWrapperComponent(service) {
            this.service = service;
        }
        AuthWrapperComponent.prototype.ngOnInit = function () { };
        return AuthWrapperComponent;
    }());
    AuthWrapperComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-auth-wrapper',
                    template: "<div class=\"row\">\n  <div class=\"mx-auto col col-md-5\">\n    <ng-container *ngIf=\"(service.isMultiTenancyEnabled$ | async) && service.isTenantBoxVisible\">\n      <abp-tenant-box\n        *abpReplaceableTemplate=\"{ componentKey: service.tenantBoxKey }\"\n      ></abp-tenant-box>\n    </ng-container>\n\n    <div class=\"abp-account-container\">\n      <div\n        *ngIf=\"service.enableLocalLogin$ | async; else disableLocalLoginTemplate\"\n        class=\"card mt-3 shadow-sm rounded\"\n      >\n        <div class=\"card-body p-5\">\n          <ng-content></ng-content>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n\n<ng-template #disableLocalLoginTemplate>\n  <div class=\"alert alert-warning\">\n    <strong>{{ 'AbpAccount::InvalidLoginRequest' | abpLocalization }}</strong>\n    {{ 'AbpAccount::ThereAreNoLoginSchemesConfiguredForThisClient' | abpLocalization }}\n  </div>\n</ng-template>\n",
                    providers: [ng_account_core.AuthWrapperService]
                },] }
    ];
    AuthWrapperComponent.ctorParameters = function () { return [
        { type: ng_account_core.AuthWrapperService }
    ]; };

    var TenantBoxComponent = /** @class */ (function () {
        function TenantBoxComponent(service) {
            this.service = service;
        }
        TenantBoxComponent.prototype.ngOnInit = function () { };
        return TenantBoxComponent;
    }());
    TenantBoxComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-tenant-box',
                    template: "<ng-container *ngIf=\"(service.currentTenant$ | async) || {} as currentTenant\">\n  <div class=\"card shadow-sm rounded mb-3\">\n    <div class=\"card-body px-5\">\n      <div class=\"row\">\n        <div class=\"col\">\n          <span style=\"font-size: 0.8em\" class=\"text-uppercase text-muted\">{{\n            'AbpUiMultiTenancy::Tenant' | abpLocalization\n          }}</span\n          ><br />\n          <h6 class=\"m-0 d-inline-block\">\n            <i>{{ currentTenant.name || ('AbpUiMultiTenancy::NotSelected' | abpLocalization) }}</i>\n          </h6>\n        </div>\n        <div class=\"col-auto\">\n          <a\n            id=\"AbpTenantSwitchLink\"\n            href=\"javascript:void(0);\"\n            class=\"btn btn-sm mt-3 btn-outline-primary\"\n            (click)=\"service.onSwitch()\"\n            >{{ 'AbpUiMultiTenancy::Switch' | abpLocalization }}</a\n          >\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <abp-modal size=\"md\" [(visible)]=\"service.isModalVisible\" [busy]=\"service.modalBusy\">\n    <ng-template #abpHeader>\n      <h5>Switch Tenant</h5>\n    </ng-template>\n    <ng-template #abpBody>\n      <form (ngSubmit)=\"service.save()\">\n        <div class=\"mt-2\">\n          <div class=\"form-group\">\n            <label for=\"name\">{{ 'AbpUiMultiTenancy::Name' | abpLocalization }}</label>\n            <input\n              [(ngModel)]=\"service.name\"\n              type=\"text\"\n              id=\"name\"\n              name=\"tenant\"\n              class=\"form-control\"\n              autofocus\n            />\n          </div>\n          <p>{{ 'AbpUiMultiTenancy::SwitchTenantHint' | abpLocalization }}</p>\n        </div>\n      </form>\n    </ng-template>\n    <ng-template #abpFooter>\n      <button abpClose type=\"button\" class=\"btn btn-secondary\">\n        {{ 'AbpTenantManagement::Cancel' | abpLocalization }}\n      </button>\n      <abp-button\n        type=\"abp-button\"\n        iconClass=\"fa fa-check\"\n        (click)=\"service.save()\"\n        [disabled]=\"currentTenant?.name === service.name\"\n      >\n        <span>{{ 'AbpTenantManagement::Save' | abpLocalization }}</span>\n      </abp-button>\n    </ng-template>\n  </abp-modal>\n</ng-container>\n",
                    providers: [ng_account_core.TenantBoxService]
                },] }
    ];
    TenantBoxComponent.ctorParameters = function () { return [
        { type: ng_account_core.TenantBoxService }
    ]; };

    var ApplicationLayoutComponent = /** @class */ (function () {
        function ApplicationLayoutComponent(service) {
            this.service = service;
        }
        ApplicationLayoutComponent.prototype.ngAfterViewInit = function () {
            this.service.subscribeWindowSize();
        };
        return ApplicationLayoutComponent;
    }());
    // required for dynamic component
    ApplicationLayoutComponent.type = "application" /* application */;
    ApplicationLayoutComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-layout-application',
                    template: "<nav\n  class=\"navbar navbar-expand-lg navbar-dark bg-dark shadow-sm flex-column flex-md-row mb-4\"\n  id=\"main-navbar\"\n  style=\"min-height: 4rem\"\n>\n  <div class=\"container\">\n    <abp-logo *abpReplaceableTemplate=\"{ componentKey: service.logoComponentKey }\"></abp-logo>\n    <button\n      class=\"navbar-toggler\"\n      type=\"button\"\n      [attr.aria-expanded]=\"!service.isCollapsed\"\n      (click)=\"service.isCollapsed = !service.isCollapsed\"\n    >\n      <span class=\"navbar-toggler-icon\"></span>\n    </button>\n    <div class=\"navbar-collapse\" [class.overflow-hidden]=\"service.smallScreen\" id=\"main-navbar-collapse\">\n      <ng-container *ngTemplateOutlet=\"!service.smallScreen ? navigations : null\"></ng-container>\n\n      <div *ngIf=\"service.smallScreen\" [@collapseWithMargin]=\"service.isCollapsed ? 'collapsed' : 'expanded'\">\n        <ng-container *ngTemplateOutlet=\"navigations\"></ng-container>\n      </div>\n\n      <ng-template #navigations>\n        <abp-routes\n          *abpReplaceableTemplate=\"{\n            componentKey: service.routesComponentKey,\n            inputs: {\n              smallScreen: { value: service.smallScreen }\n            }\n          }\"\n          class=\"mx-auto\"\n          [smallScreen]=\"service.smallScreen\"\n        ></abp-routes>\n\n        <abp-nav-items\n          *abpReplaceableTemplate=\"{\n            componentKey: service.navItemsComponentKey\n          }\"\n        ></abp-nav-items>\n      </ng-template>\n    </div>\n  </div>\n</nav>\n\n<!-- [@slideFromBottom]=\"outlet.isActivated && outlet.activatedRoute?.routeConfig?.path\" TODO: throws ExpressionChangedAfterItHasBeenCheck when animation is active. It should be fixed -->\n<div class=\"container\">\n  <abp-page-alert-container></abp-page-alert-container>\n\n  <router-outlet #outlet=\"outlet\"></router-outlet>\n</div>\n",
                    animations: [ng_theme_shared.slideFromBottom, ng_theme_shared.collapseWithMargin],
                    providers: [LayoutService, ng_core.SubscriptionService]
                },] }
    ];
    ApplicationLayoutComponent.ctorParameters = function () { return [
        { type: LayoutService }
    ]; };

    var EmptyLayoutComponent = /** @class */ (function () {
        function EmptyLayoutComponent() {
        }
        return EmptyLayoutComponent;
    }());
    EmptyLayoutComponent.type = "empty" /* empty */;
    EmptyLayoutComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-layout-empty',
                    template: "\n    <router-outlet></router-outlet>\n    <abp-confirmation></abp-confirmation>\n  "
                },] }
    ];

    var LogoComponent = /** @class */ (function () {
        function LogoComponent(environment) {
            this.environment = environment;
        }
        Object.defineProperty(LogoComponent.prototype, "appInfo", {
            get: function () {
                return this.environment.getEnvironment().application;
            },
            enumerable: false,
            configurable: true
        });
        return LogoComponent;
    }());
    LogoComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-logo',
                    template: "\n    <a class=\"navbar-brand\" routerLink=\"/\">\n      <img\n        *ngIf=\"appInfo.logoUrl; else appName\"\n        [src]=\"appInfo.logoUrl\"\n        [alt]=\"appInfo.name\"\n        width=\"100%\"\n        height=\"auto\"\n      />\n    </a>\n\n    <ng-template #appName>\n      {{ appInfo.name }}\n    </ng-template>\n  "
                },] }
    ];
    LogoComponent.ctorParameters = function () { return [
        { type: ng_core.EnvironmentService }
    ]; };

    var CurrentUserComponent = /** @class */ (function () {
        function CurrentUserComponent(navigateToManageProfile, authService, configState, sessionState) {
            this.navigateToManageProfile = navigateToManageProfile;
            this.authService = authService;
            this.configState = configState;
            this.sessionState = sessionState;
            this.currentUser$ = this.configState.getOne$('currentUser');
            this.selectedTenant$ = this.sessionState.getTenant$();
        }
        Object.defineProperty(CurrentUserComponent.prototype, "smallScreen", {
            get: function () {
                return window.innerWidth < 992;
            },
            enumerable: false,
            configurable: true
        });
        CurrentUserComponent.prototype.navigateToLogin = function () {
            this.authService.navigateToLogin();
        };
        CurrentUserComponent.prototype.logout = function () {
            this.authService.logout().subscribe();
        };
        return CurrentUserComponent;
    }());
    CurrentUserComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-current-user',
                    template: "<ng-template #loginBtn>\n  <a role=\"button\" class=\"nav-link pointer\" (click)=\"navigateToLogin()\">{{\n    'AbpAccount::Login' | abpLocalization\n  }}</a>\n</ng-template>\n<div\n  *ngIf=\"(currentUser$ | async)?.isAuthenticated; else loginBtn\"\n  ngbDropdown\n  class=\"dropdown\"\n  #currentUserDropdown=\"ngbDropdown\"\n  display=\"static\"\n>\n  <a\n    ngbDropdownToggle\n    class=\"nav-link\"\n    href=\"javascript:void(0)\"\n    role=\"button\"\n    id=\"dropdownMenuLink\"\n    data-toggle=\"dropdown\"\n    aria-haspopup=\"true\"\n    aria-expanded=\"false\"\n  >\n    <small *ngIf=\"(selectedTenant$ | async)?.name as tenantName\"\n      ><i>{{ tenantName }}</i\n      >\\</small\n    >\n    <strong>{{ (currentUser$ | async)?.userName }}</strong>\n  </a>\n  <div\n    class=\"dropdown-menu dropdown-menu-right border-0 shadow-sm\"\n    aria-labelledby=\"dropdownMenuLink\"\n    [class.d-block]=\"smallScreen && currentUserDropdown.isOpen()\"\n  >\n    <a class=\"dropdown-item pointer\" (click)=\"navigateToManageProfile()\"\n      ><i class=\"fa fa-cog mr-1\"></i>{{ 'AbpAccount::MyAccount' | abpLocalization }}</a\n    >\n    <a class=\"dropdown-item\" href=\"javascript:void(0)\" (click)=\"logout()\"\n      ><i class=\"fa fa-power-off mr-1\"></i>{{ 'AbpUi::Logout' | abpLocalization }}</a\n    >\n  </div>\n</div>\n"
                },] }
    ];
    CurrentUserComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [ng_core.NAVIGATE_TO_MANAGE_PROFILE,] }] },
        { type: ng_core.AuthService },
        { type: ng_core.ConfigStateService },
        { type: ng_core.SessionStateService }
    ]; };

    var LanguagesComponent = /** @class */ (function () {
        function LanguagesComponent(sessionState, configState) {
            this.sessionState = sessionState;
            this.configState = configState;
            this.languages$ = this.configState.getDeep$('localization.languages');
        }
        Object.defineProperty(LanguagesComponent.prototype, "smallScreen", {
            get: function () {
                return window.innerWidth < 992;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LanguagesComponent.prototype, "defaultLanguage$", {
            get: function () {
                var _this = this;
                return this.languages$.pipe(operators.map(function (languages) { return snq__default['default'](function () { return languages.find(function (lang) { return lang.cultureName === _this.selectedLangCulture; }).displayName; }); }, ''));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LanguagesComponent.prototype, "dropdownLanguages$", {
            get: function () {
                var _this = this;
                return this.languages$.pipe(operators.map(function (languages) { return snq__default['default'](function () { return languages.filter(function (lang) { return lang.cultureName !== _this.selectedLangCulture; }); }); }, []));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LanguagesComponent.prototype, "selectedLangCulture", {
            get: function () {
                return this.sessionState.getLanguage();
            },
            enumerable: false,
            configurable: true
        });
        LanguagesComponent.prototype.onChangeLang = function (cultureName) {
            this.sessionState.setLanguage(cultureName);
        };
        return LanguagesComponent;
    }());
    LanguagesComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-languages',
                    // tslint:disable-next-line: component-max-inline-declarations
                    template: "\n    <div\n      *ngIf=\"(dropdownLanguages$ | async)?.length > 0\"\n      class=\"dropdown\"\n      ngbDropdown\n      #languageDropdown=\"ngbDropdown\"\n      display=\"static\"\n    >\n      <a\n        ngbDropdownToggle\n        class=\"nav-link\"\n        href=\"javascript:void(0)\"\n        role=\"button\"\n        id=\"dropdownMenuLink\"\n        data-toggle=\"dropdown\"\n        aria-haspopup=\"true\"\n        aria-expanded=\"false\"\n      >\n        {{ defaultLanguage$ | async }}\n      </a>\n      <div\n        class=\"dropdown-menu dropdown-menu-right border-0 shadow-sm\"\n        aria-labelledby=\"dropdownMenuLink\"\n        [class.d-block]=\"smallScreen && languageDropdown.isOpen()\"\n      >\n        <a\n          *ngFor=\"let lang of dropdownLanguages$ | async\"\n          href=\"javascript:void(0)\"\n          class=\"dropdown-item\"\n          (click)=\"onChangeLang(lang.cultureName)\"\n          >{{ lang?.displayName }}</a\n        >\n      </div>\n    </div>\n  "
                },] }
    ];
    LanguagesComponent.ctorParameters = function () { return [
        { type: ng_core.SessionStateService },
        { type: ng_core.ConfigStateService }
    ]; };

    var NavItemsComponent = /** @class */ (function () {
        function NavItemsComponent(navItems) {
            this.navItems = navItems;
            this.trackByFn = function (_, element) { return element.id; };
        }
        return NavItemsComponent;
    }());
    NavItemsComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-nav-items',
                    template: "<ul class=\"navbar-nav\">\n  <ng-container *ngFor=\"let item of navItems.items$ | async; trackBy: trackByFn\">\n    <ng-container *ngIf=\"item.visible()\">\n      <li class=\"nav-item d-flex align-items-center\" *abpPermission=\"item.requiredPolicy\">\n        <ng-container\n          *ngIf=\"item.component; else htmlTemplate\"\n          [ngComponentOutlet]=\"item.component\"\n        ></ng-container>\n\n        <ng-template #htmlTemplate>\n          <div [innerHTML]=\"item.html\" (click)=\"item.action ? item.action() : null\"></div>\n        </ng-template>\n      </li>\n    </ng-container>\n  </ng-container>\n</ul>\n"
                },] }
    ];
    NavItemsComponent.ctorParameters = function () { return [
        { type: ng_theme_shared.NavItemsService }
    ]; };

    var RoutesComponent = /** @class */ (function () {
        function RoutesComponent(routesService, renderer) {
            this.routesService = routesService;
            this.renderer = renderer;
            this.trackByFn = function (_, item) { return item.name; };
        }
        RoutesComponent.prototype.isDropdown = function (node) {
            return !(node === null || node === void 0 ? void 0 : node.isLeaf) || this.routesService.hasChildren(node.name);
        };
        RoutesComponent.prototype.closeDropdown = function () {
            var _this = this;
            this.childrenContainers.forEach(function (_a) {
                var nativeElement = _a.nativeElement;
                _this.renderer.addClass(nativeElement, 'd-none');
                setTimeout(function () { return _this.renderer.removeClass(nativeElement, 'd-none'); }, 0);
            });
        };
        return RoutesComponent;
    }());
    RoutesComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-routes',
                    template: "<ul class=\"navbar-nav\">\n  <ng-container\n    *ngFor=\"let route of routesService.visible$ | async; trackBy: trackByFn\"\n    [ngTemplateOutlet]=\"isDropdown(route) ? dropdownLink : defaultLink\"\n    [ngTemplateOutletContext]=\"{ $implicit: route }\"\n  >\n  </ng-container>\n\n  <ng-template #defaultLink let-route>\n    <li class=\"nav-item\" *abpPermission=\"route.requiredPolicy\">\n      <a class=\"nav-link\" [routerLink]=\"[route.path]\"\n        ><i *ngIf=\"route.iconClass\" [ngClass]=\"route.iconClass\"></i>\n        {{ route.name | abpLocalization }}</a\n      >\n    </li>\n  </ng-template>\n\n  <ng-template #dropdownLink let-route>\n    <ng-container *ngIf=\"route.children?.length\">\n      <li\n        #navbarRootDropdown\n        class=\"nav-item dropdown\"\n        display=\"static\"\n        *abpPermission=\"route.requiredPolicy\"\n        (click)=\"\n          navbarRootDropdown.expand\n            ? (navbarRootDropdown.expand = false)\n            : (navbarRootDropdown.expand = true)\n        \"\n      >\n        <a\n          class=\"nav-link dropdown-toggle\"\n          data-toggle=\"dropdown\"\n          aria-haspopup=\"true\"\n          aria-expanded=\"false\"\n          href=\"javascript:void(0)\"\n        >\n          <i *ngIf=\"route.iconClass\" [ngClass]=\"route.iconClass\"></i>\n          {{ route.name | abpLocalization }}\n        </a>\n        <div\n          #routeContainer\n          class=\"dropdown-menu border-0 shadow-sm\"\n          (click)=\"$event.preventDefault(); $event.stopPropagation()\"\n          [class.d-block]=\"smallScreen && navbarRootDropdown.expand\"\n        >\n          <ng-container\n            *ngTemplateOutlet=\"forTemplate; context: { $implicit: route }\"\n          ></ng-container>\n        </div>\n      </li>\n    </ng-container>\n  </ng-template>\n\n  <ng-template #forTemplate let-route>\n    <ng-container *ngFor=\"let child of route.children\">\n      <ng-template\n        [ngTemplateOutlet]=\"child.children?.length ? dropdownChild : defaultChild\"\n        [ngTemplateOutletContext]=\"{ $implicit: child }\"\n      ></ng-template>\n    </ng-container>\n  </ng-template>\n\n  <ng-template #defaultChild let-child>\n    <ng-container *ngIf=\"child.path\">\n      <div class=\"dropdown-submenu\" *abpPermission=\"child.requiredPolicy\">\n        <a class=\"dropdown-item\" [routerLink]=\"[child.path]\" (click)=\"closeDropdown()\">\n          <i *ngIf=\"child.iconClass\" [ngClass]=\"child.iconClass\"></i>\n          {{ child.name | abpLocalization }}</a\n        >\n      </div>\n    </ng-container>\n  </ng-template>\n\n  <ng-template #dropdownChild let-child>\n    <div\n      class=\"dropdown-submenu\"\n      ngbDropdown\n      #dropdownSubmenu=\"ngbDropdown\"\n      placement=\"right-top\"\n      [autoClose]=\"true\"\n      *abpPermission=\"child.requiredPolicy\"\n    >\n      <div ngbDropdownToggle [class.dropdown-toggle]=\"false\">\n        <a\n          abpEllipsis=\"210px\"\n          [abpEllipsisEnabled]=\"!smallScreen\"\n          role=\"button\"\n          class=\"btn d-block text-left dropdown-toggle\"\n        >\n          <i *ngIf=\"child.iconClass\" [ngClass]=\"child.iconClass\"></i>\n          {{ child.name | abpLocalization }}\n        </a>\n      </div>\n      <div\n        #childrenContainer\n        class=\"dropdown-menu border-0 shadow-sm\"\n        [class.d-block]=\"smallScreen && dropdownSubmenu.isOpen()\"\n      >\n        <ng-container *ngTemplateOutlet=\"forTemplate; context: { $implicit: child }\"></ng-container>\n      </div>\n    </div>\n  </ng-template>\n</ul>\n"
                },] }
    ];
    RoutesComponent.ctorParameters = function () { return [
        { type: ng_core.RoutesService },
        { type: core.Renderer2 }
    ]; };
    RoutesComponent.propDecorators = {
        smallScreen: [{ type: core.Input }],
        childrenContainers: [{ type: core.ViewChildren, args: ['childrenContainer',] }]
    };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var ValidationErrorComponent = /** @class */ (function (_super) {
        __extends(ValidationErrorComponent, _super);
        function ValidationErrorComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(ValidationErrorComponent.prototype, "abpErrors", {
            get: function () {
                if (!this.errors || !this.errors.length)
                    return [];
                return this.errors.map(function (error) {
                    if (!error.message)
                        return error;
                    var index = error.message.indexOf('[');
                    if (index > -1) {
                        return Object.assign(Object.assign({}, error), { message: error.message.slice(0, index), interpoliteParams: error.message.slice(index + 1, error.message.length - 1).split(',') });
                    }
                    return error;
                });
            },
            enumerable: false,
            configurable: true
        });
        return ValidationErrorComponent;
    }(core$1.ValidationErrorComponent));
    ValidationErrorComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-validation-error',
                    template: "\n    <div class=\"invalid-feedback\" *ngFor=\"let error of abpErrors; trackBy: trackByFn\">\n      {{ error.message | abpLocalization: error.interpoliteParams }}\n    </div>\n  ",
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    encapsulation: core.ViewEncapsulation.None
                },] }
    ];

    var PageAlertContainerComponent = /** @class */ (function () {
        function PageAlertContainerComponent(service) {
            this.service = service;
        }
        return PageAlertContainerComponent;
    }());
    PageAlertContainerComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'abp-page-alert-container',
                    template: "<ng-container *ngFor=\"let alert of service.alerts$ | async; let i = index\">\n  <div\n    class=\"alert alert-{{ alert.type }} fade show\"\n    [ngClass]=\"{ 'alert-dismissible fade show': alert.dismissible }\"\n    role=\"alert\"\n  >\n    <h4 class=\"alert-heading\" *ngIf=\"alert.title\">\n      {{ alert.title | abpLocalization: alert.titleLocalizationParams }}\n    </h4>\n    {{ alert.message | abpLocalization: alert.messageLocalizationParams }}\n    <button\n      *ngIf=\"alert.dismissible\"\n      type=\"button\"\n      class=\"close\"\n      data-dismiss=\"alert\"\n      aria-label=\"Close\"\n      (click)=\"service.remove(i)\"\n    >\n      <span aria-hidden=\"true\">&times;</span>\n    </button>\n  </div>\n</ng-container>\n",
                    encapsulation: core.ViewEncapsulation.None
                },] }
    ];
    PageAlertContainerComponent.ctorParameters = function () { return [
        { type: ng_theme_shared.PageAlertService }
    ]; };

    var BASIC_THEME_NAV_ITEM_PROVIDERS = [
        {
            provide: core.APP_INITIALIZER,
            useFactory: configureNavItems,
            deps: [ng_theme_shared.NavItemsService],
            multi: true,
        },
    ];
    function configureNavItems(navItems) {
        return function () {
            navItems.addItems([
                {
                    id: "Theme.LanguagesComponent" /* Languages */,
                    order: 100,
                    component: LanguagesComponent,
                },
                {
                    id: "Theme.CurrentUserComponent" /* CurrentUser */,
                    order: 100,
                    component: CurrentUserComponent,
                },
            ]);
        };
    }

    var styles = "\n.content-header-title {\n    font-size: 24px;\n}\n.entry-row {\n    margin-bottom: 15px;\n}\n#main-navbar-tools a.dropdown-toggle {\n    text-decoration: none;\n    color: #fff;\n}\n.navbar .dropdown-submenu {\n    position: relative;\n}\n.navbar .dropdown-menu {\n    margin: 0;\n    padding: 0;\n}\n.navbar .dropdown-menu a {\n    font-size: .9em;\n    padding: 10px 15px;\n    display: block;\n    min-width: 210px;\n    text-align: left;\n    border-radius: 0.25rem;\n    min-height: 44px;\n}\n[dir=rtl] .navbar .dropdown-menu a {\n    text-align: right!important;\n}\n.navbar .dropdown-submenu a::after {\n    transform: rotate(-90deg);\n    position: absolute;\n    right: 16px;\n    top: 18px;\n}\n[dir=rtl] .navbar .dropdown-submenu a::after {\n    transform: rotate(90deg);\n    left: 16px;\n    right: auto;\n    top: 20px;\n}\n.navbar .dropdown-submenu .dropdown-menu {\n    top: 0;\n    left: 100%;\n}\n.card-header .btn {\n    padding: 2px 6px;\n}\n.card-header h5 {\n    margin: 0;\n}\n.container > .card {\n    box-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.075) !important;\n}\n@media screen and (min-width: 992px) {\n    .navbar .dropdown:hover > .dropdown-menu {\n        display: block;\n    }\n\n    .navbar .dropdown-submenu:hover > .dropdown-menu {\n        display: block;\n    }\n}\n.input-validation-error {\n    border-color: #dc3545;\n}\n.field-validation-error {\n    font-size: 0.8em;\n}\n.ui-table .ui-table-tbody > tr.empty-row > div.empty-row-content {\n    border: 1px solid #c8c8c8;\n  }\n.abp-loading {\n    background: rgba(0, 0, 0, 0.05);\n}\n.modal-backdrop {\nbackground-color: rgba(0, 0, 0, 0.6);\n}\n\n.confirmation .confirmation-backdrop {\n\t background: rgba(0, 0, 0, 0.7) !important;\n}\n .confirmation .confirmation-dialog {\n\t border: none;\n\t border-radius: 10px;\n\t background-color: #fff;\n\t box-shadow: 0 0 10px -5px rgba(0, 0, 0, 0.5);\n}\n .confirmation .confirmation-dialog .icon-container .icon {\n\t stroke: #fff;\n\t color: #fff;\n}\n .confirmation .confirmation-dialog .icon-container.info .icon {\n\t stroke: #2f96b4;\n\t color: #2f96b4;\n}\n .confirmation .confirmation-dialog .icon-container.success .icon {\n\t stroke: #51a351;\n\t color: #51a351;\n}\n .confirmation .confirmation-dialog .icon-container.warning .icon {\n\t stroke: #f89406;\n\t color: #f89406;\n}\n .confirmation .confirmation-dialog .icon-container.error .icon {\n\t stroke: #bd362f;\n\t color: #bd362f;\n}\n .confirmation .confirmation-dialog .content .title {\n\t color: #222;\n}\n .confirmation .confirmation-dialog .content .message {\n\t color: #777;\n}\n .confirmation .confirmation-dialog .footer {\n\t background: transparent;\n}\n .confirmation .confirmation-dialog .footer .confirmation-button {\n\t background-color: #eee;\n\t color: #777;\n}\n .confirmation .confirmation-dialog .footer .confirmation-button:hover, .confirmation .confirmation-dialog .footer .confirmation-button:focus, .confirmation .confirmation-dialog .footer .confirmation-button:active {\n\t background-color: #bbb;\n}\n .confirmation .confirmation-dialog .footer .confirmation-button--confirm {\n\t background-color: #2f96b4;\n\t color: #fff;\n}\n .confirmation .confirmation-dialog .footer .confirmation-button--confirm:hover {\n\t background-color: #2e819b;\n}\n.ui-table .pagination-wrapper {\n    background-color: #f4f4f4;\n    border: 1px solid #c8c8c8;\n}\n.bordered .datatable-body-row {\n    border-top: 1px solid #eee;\n    margin-top: -1px;\n}\n.breadcrumb {\n    background-color: transparent;\n    padding: 0.27rem;\n}\n";

    var BASIC_THEME_STYLES_PROVIDERS = [
        {
            provide: core.APP_INITIALIZER,
            useFactory: configureStyles,
            deps: [ng_core.DomInsertionService, ng_core.ReplaceableComponentsService],
            multi: true,
        },
    ];
    function configureStyles(domInsertion, replaceableComponents) {
        return function () {
            domInsertion.insertContent(ng_core.CONTENT_STRATEGY.AppendStyleToHead(styles));
            initLayouts(replaceableComponents);
        };
    }
    function initLayouts(replaceableComponents) {
        replaceableComponents.add({
            key: "Theme.ApplicationLayoutComponent" /* ApplicationLayout */,
            component: ApplicationLayoutComponent,
        });
        replaceableComponents.add({
            key: "Theme.AccountLayoutComponent" /* AccountLayout */,
            component: AccountLayoutComponent,
        });
        replaceableComponents.add({
            key: "Theme.EmptyLayoutComponent" /* EmptyLayout */,
            component: EmptyLayoutComponent,
        });
    }

    var LAYOUTS = [ApplicationLayoutComponent, AccountLayoutComponent, EmptyLayoutComponent];
    var BaseThemeBasicModule = /** @class */ (function () {
        function BaseThemeBasicModule() {
        }
        return BaseThemeBasicModule;
    }());
    BaseThemeBasicModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: __spread(LAYOUTS, [
                        ValidationErrorComponent,
                        LogoComponent,
                        NavItemsComponent,
                        RoutesComponent,
                        CurrentUserComponent,
                        LanguagesComponent,
                        PageAlertContainerComponent,
                        TenantBoxComponent,
                        AuthWrapperComponent,
                    ]),
                    exports: __spread(LAYOUTS, [
                        ValidationErrorComponent,
                        LogoComponent,
                        NavItemsComponent,
                        RoutesComponent,
                        CurrentUserComponent,
                        LanguagesComponent,
                        PageAlertContainerComponent,
                    ]),
                    imports: [
                        ng_core.CoreModule,
                        ng_theme_shared.ThemeSharedModule,
                        ngBootstrap.NgbCollapseModule,
                        ngBootstrap.NgbDropdownModule,
                        core$1.NgxValidateCoreModule,
                    ],
                    entryComponents: __spread(LAYOUTS, [ValidationErrorComponent, CurrentUserComponent, LanguagesComponent]),
                },] }
    ];
    var ThemeBasicModule = /** @class */ (function () {
        function ThemeBasicModule() {
        }
        ThemeBasicModule.forRoot = function () {
            return {
                ngModule: ThemeBasicModule,
                providers: [
                    BASIC_THEME_NAV_ITEM_PROVIDERS,
                    BASIC_THEME_STYLES_PROVIDERS,
                    {
                        provide: core$1.VALIDATION_ERROR_TEMPLATE,
                        useValue: ValidationErrorComponent,
                    },
                    {
                        provide: core$1.VALIDATION_TARGET_SELECTOR,
                        useValue: '.form-group',
                    },
                    {
                        provide: core$1.VALIDATION_INVALID_CLASSES,
                        useValue: 'is-invalid',
                    },
                ],
            };
        };
        return ThemeBasicModule;
    }());
    ThemeBasicModule.decorators = [
        { type: core.NgModule, args: [{
                    exports: [BaseThemeBasicModule],
                    imports: [BaseThemeBasicModule],
                },] }
    ];

    /*
     * Public API Surface of theme-basic
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AccountLayoutComponent = AccountLayoutComponent;
    exports.ApplicationLayoutComponent = ApplicationLayoutComponent;
    exports.AuthWrapperComponent = AuthWrapperComponent;
    exports.BASIC_THEME_NAV_ITEM_PROVIDERS = BASIC_THEME_NAV_ITEM_PROVIDERS;
    exports.BASIC_THEME_STYLES_PROVIDERS = BASIC_THEME_STYLES_PROVIDERS;
    exports.BaseThemeBasicModule = BaseThemeBasicModule;
    exports.CurrentUserComponent = CurrentUserComponent;
    exports.EmptyLayoutComponent = EmptyLayoutComponent;
    exports.LAYOUTS = LAYOUTS;
    exports.LanguagesComponent = LanguagesComponent;
    exports.LogoComponent = LogoComponent;
    exports.NavItemsComponent = NavItemsComponent;
    exports.PageAlertContainerComponent = PageAlertContainerComponent;
    exports.RoutesComponent = RoutesComponent;
    exports.TenantBoxComponent = TenantBoxComponent;
    exports.ThemeBasicModule = ThemeBasicModule;
    exports.ValidationErrorComponent = ValidationErrorComponent;
    exports.configureNavItems = configureNavItems;
    exports.configureStyles = configureStyles;
    exports.a = ApplicationLayoutComponent;
    exports.b = LayoutService;
    exports.c = AccountLayoutComponent;
    exports.d = EmptyLayoutComponent;
    exports.e = ValidationErrorComponent;
    exports.f = LogoComponent;
    exports.g = NavItemsComponent;
    exports.h = RoutesComponent;
    exports.i = CurrentUserComponent;
    exports.j = LanguagesComponent;
    exports.k = PageAlertContainerComponent;
    exports.l = TenantBoxComponent;
    exports.m = AuthWrapperComponent;
    exports.n = BASIC_THEME_NAV_ITEM_PROVIDERS;
    exports.o = configureNavItems;
    exports.p = BASIC_THEME_STYLES_PROVIDERS;
    exports.q = configureStyles;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
