// @abp/ng.theme.shared/bundles/abp-ng.theme.shared.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/animations'), require('@abp/ng.core'), require('@angular/core'), require('@angular/router'), require('rxjs/operators'), require('rxjs'), require('snq'), require('@ng-bootstrap/ng-bootstrap'), require('@angular/common'), require('@swimlane/ngx-datatable'), require('just-clone'), require('@angular/common/http'), require('@ngxs/store'), require('@ngx-validate/core'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@abp/ng.theme.shared', ['exports', '@angular/animations', '@abp/ng.core', '@angular/core', '@angular/router', 'rxjs/operators', 'rxjs', 'snq', '@ng-bootstrap/ng-bootstrap', '@angular/common', '@swimlane/ngx-datatable', 'just-clone', '@angular/common/http', '@ngxs/store', '@ngx-validate/core', '@angular/forms'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.abp = global.abp || {}, global.abp.ng = global.abp.ng || {}, global.abp.ng.theme = global.abp.ng.theme || {}, global.abp.ng.theme.shared = {}), global.ng.animations, global.i1, global.ng.core, global.ng.router, global.rxjs.operators, global.rxjs, global.snq, global.ngBootstrap, global.ng.common, global.ngxDatatable, global.clone, global.ng.common.http, global.i1$1, global.core, global.ng.forms));
}(this, (function (exports, animations, i1, i0, router, operators, rxjs, snq, ngBootstrap, common, ngxDatatable, clone, http, i1$1, core, forms) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var snq__default = /*#__PURE__*/_interopDefaultLegacy(snq);
    var clone__default = /*#__PURE__*/_interopDefaultLegacy(clone);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);

    var bounceIn = animations.animation([
        animations.style({ opacity: '0', display: '{{ display }}' }),
        animations.animate('{{ time}} {{ easing }}', animations.keyframes([
            animations.style({ opacity: '0', transform: '{{ transform }} scale(0.0)', offset: 0 }),
            animations.style({ opacity: '0', transform: '{{ transform }} scale(0.8)', offset: 0.5 }),
            animations.style({ opacity: '1', transform: '{{ transform }} scale(1.0)', offset: 1 })
        ]))
    ], {
        params: {
            time: '350ms',
            easing: 'cubic-bezier(.7,.31,.72,1.47)',
            display: 'block',
            transform: 'translate(-50%, -50%)'
        }
    });

    var collapseY = animations.animation([
        animations.style({ height: '*', overflow: 'hidden', 'box-sizing': 'border-box' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ height: '0', padding: '0px' })),
    ], { params: { time: '350ms', easing: 'ease' } });
    var collapseYWithMargin = animations.animation([
        animations.style({ 'margin-top': '0' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ 'margin-left': '-100%' })),
    ], {
        params: { time: '500ms', easing: 'ease' },
    });
    var collapseX = animations.animation([
        animations.style({ width: '*', overflow: 'hidden', 'box-sizing': 'border-box' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ width: '0', padding: '0px' })),
    ], { params: { time: '350ms', easing: 'ease' } });
    var expandY = animations.animation([
        animations.style({ height: '0', overflow: 'hidden', 'box-sizing': 'border-box' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ height: '*', padding: '*' })),
    ], { params: { time: '350ms', easing: 'ease' } });
    var expandYWithMargin = animations.animation([
        animations.style({ 'margin-top': '-100%' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ 'margin-top': '0' })),
    ], {
        params: { time: '500ms', easing: 'ease' },
    });
    var expandX = animations.animation([
        animations.style({ width: '0', overflow: 'hidden', 'box-sizing': 'border-box' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ width: '*', padding: '*' })),
    ], { params: { time: '350ms', easing: 'ease' } });
    var collapse = animations.trigger('collapse', [
        animations.state('collapsed', animations.style({ height: '0', overflow: 'hidden' })),
        animations.state('expanded', animations.style({ height: '*', overflow: 'hidden' })),
        animations.transition('expanded => collapsed', animations.useAnimation(collapseY)),
        animations.transition('collapsed => expanded', animations.useAnimation(expandY)),
    ]);
    var collapseWithMargin = animations.trigger('collapseWithMargin', [
        animations.state('collapsed', animations.style({ 'margin-top': '-100%' })),
        animations.state('expanded', animations.style({ 'margin-top': '0' })),
        animations.transition('expanded => collapsed', animations.useAnimation(collapseYWithMargin), {
            params: { time: '400ms', easing: 'linear' },
        }),
        animations.transition('collapsed => expanded', animations.useAnimation(expandYWithMargin)),
    ]);
    var collapseLinearWithMargin = animations.trigger('collapseLinearWithMargin', [
        animations.state('collapsed', animations.style({ 'margin-top': '-100vh' })),
        animations.state('expanded', animations.style({ 'margin-top': '0' })),
        animations.transition('expanded => collapsed', animations.useAnimation(collapseYWithMargin, { params: { time: '200ms', easing: 'linear' } })),
        animations.transition('collapsed => expanded', animations.useAnimation(expandYWithMargin, { params: { time: '250ms', easing: 'linear' } })),
    ]);

    var fadeIn = animations.animation([animations.style({ opacity: '0' }), animations.animate('{{ time}} {{ easing }}', animations.style({ opacity: '1' }))], {
        params: { time: '350ms', easing: 'ease' },
    });
    var fadeOut = animations.animation([animations.style({ opacity: '1' }), animations.animate('{{ time}} {{ easing }}', animations.style({ opacity: '0' }))], { params: { time: '350ms', easing: 'ease' } });
    var fadeInDown = animations.animation([
        animations.style({ opacity: '0', transform: '{{ transform }} translateY(-20px)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '1', transform: '{{ transform }} translateY(0)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });
    var fadeInUp = animations.animation([
        animations.style({ opacity: '0', transform: '{{ transform }} translateY(20px)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '1', transform: '{{ transform }} translateY(0)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });
    var fadeInLeft = animations.animation([
        animations.style({ opacity: '0', transform: '{{ transform }} translateX(20px)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '1', transform: '{{ transform }} translateX(0)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });
    var fadeInRight = animations.animation([
        animations.style({ opacity: '0', transform: '{{ transform }} translateX(-20px)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '1', transform: '{{ transform }} translateX(0)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });
    var fadeOutDown = animations.animation([
        animations.style({ opacity: '1', transform: '{{ transform }} translateY(0)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '0', transform: '{{ transform }} translateY(20px)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });
    var fadeOutUp = animations.animation([
        animations.style({ opacity: '1', transform: '{{ transform }} translateY(0)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '0', transform: '{{ transform }} translateY(-20px)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });
    var fadeOutLeft = animations.animation([
        animations.style({ opacity: '1', transform: '{{ transform }} translateX(0)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '0', transform: '{{ transform }} translateX(20px)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });
    var fadeOutRight = animations.animation([
        animations.style({ opacity: '1', transform: '{{ transform }} translateX(0)' }),
        animations.animate('{{ time }} {{ easing }}', animations.style({ opacity: '0', transform: '{{ transform }} translateX(-20px)' })),
    ], { params: { time: '350ms', easing: 'ease', transform: '' } });

    var fadeAnimation = animations.trigger('fade', [
        animations.transition(':enter', animations.useAnimation(fadeIn)),
        animations.transition(':leave', animations.useAnimation(fadeOut)),
    ]);
    var dialogAnimation = animations.trigger('dialog', [
        animations.transition(':enter', animations.useAnimation(fadeInDown)),
        animations.transition(':leave', animations.useAnimation(fadeOut)),
    ]);

    var slideFromBottom = animations.trigger('slideFromBottom', [
        animations.transition('* <=> *', [
            animations.style({ 'margin-top': '20px', opacity: '0' }),
            animations.animate('0.2s ease-out', animations.style({ opacity: '1', 'margin-top': '0px' })),
        ]),
    ]);

    var toastInOut = animations.trigger('toastInOut', [
        animations.transition('* <=> *', [
            animations.query(':enter', [
                animations.style({ opacity: 0, transform: 'translateY(20px)' }),
                animations.animate('350ms ease', animations.style({ opacity: 1, transform: 'translateY(0)' })),
            ], { optional: true }),
            animations.query(':leave', animations.animate('450ms ease', animations.style({ opacity: 0 })), {
                optional: true,
            }),
        ]),
    ]);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var BreadcrumbComponent = /** @class */ (function () {
        function BreadcrumbComponent(cdRef, router, routes, subscription, routerEvents) {
            this.cdRef = cdRef;
            this.router = router;
            this.routes = routes;
            this.subscription = subscription;
            this.routerEvents = routerEvents;
            this.segments = [];
        }
        BreadcrumbComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.subscription.addOne(this.routerEvents.getNavigationEvents('End').pipe(
            // tslint:disable-next-line:deprecation
            operators.startWith(null), operators.map(function () { return _this.routes.search({ path: i1.getRoutePath(_this.router) }); })), function (route) {
                _this.segments = [];
                if (route) {
                    var node = { parent: route };
                    while (node.parent) {
                        node = node.parent;
                        var parent = node.parent, children = node.children, isLeaf = node.isLeaf, segment = __rest(node, ["parent", "children", "isLeaf"]);
                        if (!isAdministration(segment))
                            _this.segments.unshift(segment);
                    }
                    _this.cdRef.detectChanges();
                }
            });
        };
        return BreadcrumbComponent;
    }());
    BreadcrumbComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-breadcrumb',
                    template: "<ol class=\"breadcrumb\" *ngIf=\"segments.length\">\n  <li class=\"breadcrumb-item\">\n    <a routerLink=\"/\"><i class=\"fa fa-home\"></i> </a>\n  </li>\n  <li\n    *ngFor=\"let segment of segments; let last = last\"\n    class=\"breadcrumb-item\"\n    [class.active]=\"last\"\n    aria-current=\"page\"\n  >\n    {{ segment.name | abpLocalization }}\n  </li>\n</ol>\n",
                    changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    providers: [i1.SubscriptionService]
                },] }
    ];
    BreadcrumbComponent.ctorParameters = function () { return [
        { type: i0.ChangeDetectorRef },
        { type: router.Router },
        { type: i1.RoutesService },
        { type: i1.SubscriptionService },
        { type: i1.RouterEvents }
    ]; };
    function isAdministration(route) {
        return route.name === "AbpUiNavigation::Menu:Administration" /* Administration */;
    }

    var ButtonComponent = /** @class */ (function () {
        function ButtonComponent(renderer) {
            this.renderer = renderer;
            this.buttonId = '';
            this.buttonClass = 'btn btn-primary';
            this.buttonType = 'button';
            this.loading = false;
            this.disabled = false;
            // tslint:disable
            this.click = new i0.EventEmitter();
            this.focus = new i0.EventEmitter();
            this.blur = new i0.EventEmitter();
            // tslint:enable
            this.abpClick = new i0.EventEmitter();
            this.abpFocus = new i0.EventEmitter();
            this.abpBlur = new i0.EventEmitter();
        }
        Object.defineProperty(ButtonComponent.prototype, "icon", {
            get: function () {
                return "" + (this.loading ? 'fa fa-spinner fa-spin' : this.iconClass || 'd-none');
            },
            enumerable: false,
            configurable: true
        });
        ButtonComponent.prototype.ngOnInit = function () {
            var _this = this;
            if (this.attributes) {
                Object.keys(this.attributes).forEach(function (key) {
                    _this.renderer.setAttribute(_this.buttonRef.nativeElement, key, _this.attributes[key]);
                });
            }
        };
        return ButtonComponent;
    }());
    ButtonComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-button',
                    template: "\n    <button\n      #button\n      [id]=\"buttonId\"\n      [attr.type]=\"buttonType\"\n      [ngClass]=\"buttonClass\"\n      [disabled]=\"loading || disabled\"\n      (click.stop)=\"click.next($event); abpClick.next($event)\"\n      (focus)=\"focus.next($event); abpFocus.next($event)\"\n      (blur)=\"blur.next($event); abpBlur.next($event)\"\n    >\n      <i [ngClass]=\"icon\" class=\"mr-1\"></i><ng-content></ng-content>\n    </button>\n  "
                },] }
    ];
    ButtonComponent.ctorParameters = function () { return [
        { type: i0.Renderer2 }
    ]; };
    ButtonComponent.propDecorators = {
        buttonId: [{ type: i0.Input }],
        buttonClass: [{ type: i0.Input }],
        buttonType: [{ type: i0.Input }],
        iconClass: [{ type: i0.Input }],
        loading: [{ type: i0.Input }],
        disabled: [{ type: i0.Input }],
        attributes: [{ type: i0.Input }],
        click: [{ type: i0.Output }],
        focus: [{ type: i0.Output }],
        blur: [{ type: i0.Output }],
        abpClick: [{ type: i0.Output }],
        abpFocus: [{ type: i0.Output }],
        abpBlur: [{ type: i0.Output }],
        buttonRef: [{ type: i0.ViewChild, args: ['button', { static: true },] }]
    };

    function getRandomBackgroundColor(count) {
        var colors = [];
        for (var i = 0; i < count; i++) {
            var r = ((i + 5) * (i + 5) * 474) % 255;
            var g = ((i + 5) * (i + 5) * 1600) % 255;
            var b = ((i + 5) * (i + 5) * 84065) % 255;
            colors.push('rgba(' + r + ', ' + g + ', ' + b + ', 0.7)');
        }
        return colors;
    }
    var chartJsLoaded$ = new rxjs.ReplaySubject(1);

    var ChartComponent = /** @class */ (function () {
        function ChartComponent(el, cdRef) {
            var _this = this;
            this.el = el;
            this.cdRef = cdRef;
            this.options = {};
            this.plugins = [];
            this.responsive = true;
            // tslint:disable-next-line: no-output-on-prefix
            this.onDataSelect = new i0.EventEmitter();
            this.initialized = new rxjs.BehaviorSubject(this);
            this.onCanvasClick = function (event) {
                if (_this.chart) {
                    var element = _this.chart.getElementAtEvent(event);
                    var dataset = _this.chart.getDatasetAtEvent(event);
                    if (element && element.length && dataset) {
                        _this.onDataSelect.emit({
                            originalEvent: event,
                            element: element[0],
                            dataset: dataset,
                        });
                    }
                }
            };
            this.initChart = function () {
                var opts = _this.options || {};
                opts.responsive = _this.responsive;
                // allows chart to resize in responsive mode
                if (opts.responsive && (_this.height || _this.width)) {
                    opts.maintainAspectRatio = false;
                }
                _this.chart = new Chart(_this.canvas, {
                    type: _this.type,
                    data: _this.data,
                    options: _this.options,
                    plugins: _this.plugins,
                });
                _this.cdRef.detectChanges();
            };
            this.generateLegend = function () {
                if (_this.chart) {
                    return _this.chart.generateLegend();
                }
            };
            this.refresh = function () {
                if (_this.chart) {
                    _this.chart.update();
                    _this.cdRef.detectChanges();
                }
            };
            this.reinit = function () {
                if (_this.chart) {
                    _this.chart.destroy();
                    _this.initChart();
                }
            };
        }
        Object.defineProperty(ChartComponent.prototype, "data", {
            get: function () {
                return this._data;
            },
            set: function (val) {
                this._data = val;
                this.reinit();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ChartComponent.prototype, "canvas", {
            get: function () {
                return this.el.nativeElement.children[0].children[0];
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ChartComponent.prototype, "base64Image", {
            get: function () {
                return this.chart.toBase64Image();
            },
            enumerable: false,
            configurable: true
        });
        ChartComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            chartJsLoaded$.subscribe(function () {
                _this.testChartJs();
                _this.initChart();
                _this._initialized = true;
            });
        };
        ChartComponent.prototype.testChartJs = function () {
            try {
                // tslint:disable-next-line: no-unused-expression
                Chart;
            }
            catch (error) {
                throw new Error("Chart is not found. Import the Chart from app.module like shown below:\n      import('chart.js');\n      ");
            }
        };
        ChartComponent.prototype.ngOnDestroy = function () {
            if (this.chart) {
                this.chart.destroy();
                this._initialized = false;
                this.chart = null;
            }
        };
        return ChartComponent;
    }());
    ChartComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-chart',
                    template: "<div\n  style=\"position:relative\"\n  [style.width]=\"responsive && !width ? null : width\"\n  [style.height]=\"responsive && !height ? null : height\"\n>\n  <canvas\n    [attr.width]=\"responsive && !width ? null : width\"\n    [attr.height]=\"responsive && !height ? null : height\"\n    (click)=\"onCanvasClick($event)\"\n  ></canvas>\n</div>\n"
                },] }
    ];
    ChartComponent.ctorParameters = function () { return [
        { type: i0.ElementRef },
        { type: i0.ChangeDetectorRef }
    ]; };
    ChartComponent.propDecorators = {
        type: [{ type: i0.Input }],
        options: [{ type: i0.Input }],
        plugins: [{ type: i0.Input }],
        width: [{ type: i0.Input }],
        height: [{ type: i0.Input }],
        responsive: [{ type: i0.Input }],
        onDataSelect: [{ type: i0.Output }],
        initialized: [{ type: i0.Output }],
        data: [{ type: i0.Input }]
    };

    exports.Confirmation = void 0;
    (function (Confirmation) {
        var Status;
        (function (Status) {
            Status["confirm"] = "confirm";
            Status["reject"] = "reject";
            Status["dismiss"] = "dismiss";
        })(Status = Confirmation.Status || (Confirmation.Status = {}));
    })(exports.Confirmation || (exports.Confirmation = {}));

    var ConfirmationComponent = /** @class */ (function () {
        function ConfirmationComponent() {
            this.confirm = exports.Confirmation.Status.confirm;
            this.reject = exports.Confirmation.Status.reject;
            this.dismiss = exports.Confirmation.Status.dismiss;
        }
        ConfirmationComponent.prototype.close = function (status) {
            this.clear(status);
        };
        ConfirmationComponent.prototype.getIconClass = function (_a) {
            var severity = _a.severity;
            switch (severity) {
                case 'info':
                    return 'fa-info-circle';
                case 'success':
                    return 'fa-check-circle';
                case 'warning':
                    return 'fa-exclamation-triangle';
                case 'error':
                    return 'fa-times-circle';
                default:
                    return 'fa-question-circle';
            }
        };
        return ConfirmationComponent;
    }());
    ConfirmationComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-confirmation',
                    template: "<div class=\"confirmation\" *ngIf=\"confirmation$ | async as data\">\n  <div\n    class=\"confirmation-backdrop\"\n    (click)=\"data.options?.dismissible ? close(dismiss) : null\"\n  ></div>\n  <div class=\"confirmation-dialog\">\n    <div class=\"icon-container\" [ngClass]=\"data.severity\" *ngIf=\"data.severity\">\n      <i class=\"fa icon\" [ngClass]=\"getIconClass(data)\"></i>\n    </div>\n    <div class=\"content\">\n      <h1\n        class=\"title\"\n        *ngIf=\"data.title\"\n        [innerHTML]=\"data.title | abpLocalization: data.options?.titleLocalizationParams\"\n      ></h1>\n      <p\n        class=\"message\"\n        *ngIf=\"data.message\"\n        [innerHTML]=\"data.message | abpLocalization: data.options?.messageLocalizationParams\"\n      ></p>\n    </div>\n    <div class=\"footer\">\n      <button\n        id=\"cancel\"\n        class=\"confirmation-button confirmation-button--reject\"\n        [innerHTML]=\"data.options?.cancelText || 'AbpUi::Cancel' | abpLocalization\"\n        *ngIf=\"!data?.options?.hideCancelBtn\"\n        (click)=\"close(reject)\"\n      ></button>\n      <button\n        id=\"confirm\"\n        class=\"confirmation-button confirmation-button--approve\"\n        [innerHTML]=\"data.options?.yesText || 'AbpUi::Yes' | abpLocalization\"\n        *ngIf=\"!data?.options?.hideYesBtn\"\n        (click)=\"close(confirm)\"\n      ></button>\n    </div>\n  </div>\n</div>\n",
                    styles: [".confirmation{position:fixed;top:0;right:0;bottom:0;left:0;display:flex;align-items:center;justify-content:center;z-index:1060}.confirmation .confirmation-backdrop{position:fixed;top:0;left:0;width:100vw;height:100vh;z-index:1061!important}.confirmation .confirmation-dialog{display:flex;flex-direction:column;margin:20px auto;padding:0;width:450px;min-height:300px;z-index:1062!important}@media screen and (max-width:500px){.confirmation .confirmation-dialog{width:90vw}}.confirmation .confirmation-dialog .icon-container{display:flex;align-items:center;justify-content:center;margin:0 0 10px;padding:20px}.confirmation .confirmation-dialog .icon-container .icon{width:100px;height:100px;stroke-width:1;font-size:80px;text-align:center}.confirmation .confirmation-dialog .content{flex-grow:1;display:block}.confirmation .confirmation-dialog .content .title{display:block;margin:0;padding:0;font-size:27px;font-weight:600;text-align:center}.confirmation .confirmation-dialog .content .message{display:block;margin:10px auto;padding:20px;font-size:16px;font-weight:400;text-align:center}.confirmation .confirmation-dialog .footer{display:flex;align-items:center;justify-content:flex-end;margin:10px 0 0;padding:20px;width:100%}.confirmation .confirmation-dialog .footer .confirmation-button{display:inline-block;margin:0 5px;padding:10px 20px;border:none;border-radius:6px;font-size:14px;font-weight:600}"]
                },] }
    ];

    var HttpErrorWrapperComponent = /** @class */ (function () {
        function HttpErrorWrapperComponent(subscription) {
            this.subscription = subscription;
            this.status = 0;
            this.title = 'Oops!';
            this.details = 'Sorry, an error has occured.';
            this.customComponent = null;
            this.hideCloseIcon = false;
            this.isHomeShow = true;
        }
        Object.defineProperty(HttpErrorWrapperComponent.prototype, "statusText", {
            get: function () {
                return this.status ? "[" + this.status + "]" : '';
            },
            enumerable: false,
            configurable: true
        });
        HttpErrorWrapperComponent.prototype.ngOnInit = function () {
            this.backgroundColor =
                snq__default['default'](function () { return window.getComputedStyle(document.body).getPropertyValue('background-color'); }) ||
                    '#fff';
        };
        HttpErrorWrapperComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (this.customComponent) {
                var customComponentRef = this.cfRes
                    .resolveComponentFactory(this.customComponent)
                    .create(this.injector);
                customComponentRef.instance.errorStatus = this.status;
                customComponentRef.instance.destroy$ = this.destroy$;
                this.appRef.attachView(customComponentRef.hostView);
                this.containerRef.nativeElement.appendChild(customComponentRef.hostView.rootNodes[0]);
                customComponentRef.changeDetectorRef.detectChanges();
            }
            var keyup$ = rxjs.fromEvent(document, 'keyup').pipe(operators.debounceTime(150), operators.filter(function (key) { return key && key.key === 'Escape'; }));
            this.subscription.addOne(keyup$, function () { return _this.destroy(); });
        };
        HttpErrorWrapperComponent.prototype.ngOnDestroy = function () { };
        HttpErrorWrapperComponent.prototype.destroy = function () {
            this.destroy$.next();
            this.destroy$.complete();
        };
        return HttpErrorWrapperComponent;
    }());
    HttpErrorWrapperComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-http-error-wrapper',
                    template: "<div\n  #container\n  id=\"abp-http-error-container\"\n  class=\"error\"\n  [style.backgroundColor]=\"backgroundColor\"\n>\n  <button\n    *ngIf=\"!hideCloseIcon\"\n    id=\"abp-close-button\"\n    type=\"button\"\n    class=\"close mr-2\"\n    (click)=\"destroy()\"\n  >\n    <span aria-hidden=\"true\">&times;</span>\n  </button>\n\n  <div *ngIf=\"!customComponent\" class=\"row centered\">\n    <div class=\"col-md-12\">\n      <div class=\"error-template\">\n        <h1>{{ statusText }} {{ title | abpLocalization }}</h1>\n        <div class=\"error-details\">\n          {{ details | abpLocalization }}\n        </div>\n        <div class=\"error-actions\">\n          <a\n            *ngIf=\"isHomeShow\"\n            (click)=\"destroy()\"\n            routerLink=\"/\"\n            class=\"btn btn-primary btn-md mt-2\"\n            ><span class=\"glyphicon glyphicon-home\"></span>\n            {{ { key: '::Menu:Home', defaultValue: 'Home' } | abpLocalization }}\n          </a>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n",
                    providers: [i1.SubscriptionService],
                    styles: [".error{position:fixed;top:0;width:100vw;height:100vh;z-index:999999}.centered{position:fixed;top:50%;left:50%;transform:translate(-50%,-50%)}"]
                },] }
    ];
    HttpErrorWrapperComponent.ctorParameters = function () { return [
        { type: i1.SubscriptionService }
    ]; };
    HttpErrorWrapperComponent.propDecorators = {
        containerRef: [{ type: i0.ViewChild, args: ['container', { static: false },] }]
    };

    var LoaderBarComponent = /** @class */ (function () {
        function LoaderBarComponent(router, cdRef, subscription, httpWaitService, routerWaitService) {
            var _this = this;
            this.router = router;
            this.cdRef = cdRef;
            this.subscription = subscription;
            this.httpWaitService = httpWaitService;
            this.routerWaitService = routerWaitService;
            this.containerClass = 'abp-loader-bar';
            this.color = '#77b6ff';
            this.progressLevel = 0;
            this.interval = new rxjs.Subscription();
            this.timer = new rxjs.Subscription();
            this.intervalPeriod = 350;
            this.stopDelay = 800;
            this.clearProgress = function () {
                _this.progressLevel = 0;
                _this.cdRef.detectChanges();
            };
            this.reportProgress = function () {
                if (_this.progressLevel < 75) {
                    _this.progressLevel += 1 + Math.random() * 9;
                }
                else if (_this.progressLevel < 90) {
                    _this.progressLevel += 0.4;
                }
                else if (_this.progressLevel < 100) {
                    _this.progressLevel += 0.1;
                }
                else {
                    _this.interval.unsubscribe();
                }
                _this.cdRef.detectChanges();
            };
        }
        Object.defineProperty(LoaderBarComponent.prototype, "isLoading", {
            get: function () {
                return this._isLoading;
            },
            set: function (value) {
                this._isLoading = value;
                this.cdRef.detectChanges();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LoaderBarComponent.prototype, "boxShadow", {
            get: function () {
                return "0 0 10px rgba(" + this.color + ", 0.5)";
            },
            enumerable: false,
            configurable: true
        });
        LoaderBarComponent.prototype.ngOnInit = function () {
            this.subscribeLoading();
        };
        LoaderBarComponent.prototype.subscribeLoading = function () {
            var _this = this;
            this.subscription.addOne(rxjs.combineLatest([this.httpWaitService.getLoading$(), this.routerWaitService.getLoading$()]), function (_a) {
                var _b = __read(_a, 2), httpLoading = _b[0], routerLoading = _b[1];
                if (httpLoading || routerLoading)
                    _this.startLoading();
                else
                    _this.stopLoading();
            });
        };
        LoaderBarComponent.prototype.ngOnDestroy = function () {
            this.interval.unsubscribe();
        };
        LoaderBarComponent.prototype.startLoading = function () {
            if (this.isLoading || !this.interval.closed)
                return;
            this.isLoading = true;
            this.progressLevel = 0;
            this.cdRef.detectChanges();
            this.interval = rxjs.timer(0, this.intervalPeriod).subscribe(this.reportProgress);
            this.timer.unsubscribe();
        };
        LoaderBarComponent.prototype.stopLoading = function () {
            this.interval.unsubscribe();
            this.progressLevel = 100;
            this.isLoading = false;
            if (!this.timer.closed)
                return;
            this.timer = rxjs.timer(this.stopDelay).subscribe(this.clearProgress);
        };
        return LoaderBarComponent;
    }());
    LoaderBarComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-loader-bar',
                    template: "\n    <div id=\"abp-loader-bar\" [ngClass]=\"containerClass\" [class.is-loading]=\"isLoading\">\n      <div\n        class=\"abp-progress\"\n        [class.progressing]=\"progressLevel\"\n        [style.width.vw]=\"progressLevel\"\n        [ngStyle]=\"{\n          'background-color': color,\n          'box-shadow': boxShadow\n        }\"\n      ></div>\n    </div>\n  ",
                    providers: [i1.SubscriptionService],
                    styles: [".abp-loader-bar{left:0;opacity:0;position:fixed;top:0;transition:opacity .4s linear .4s;z-index:99999}.abp-loader-bar.is-loading{opacity:1;transition:none}.abp-loader-bar .abp-progress{height:3px;left:0;position:fixed;top:0}.abp-loader-bar .abp-progress.progressing{transition:width .4s ease}"]
                },] }
    ];
    LoaderBarComponent.ctorParameters = function () { return [
        { type: router.Router },
        { type: i0.ChangeDetectorRef },
        { type: i1.SubscriptionService },
        { type: i1.HttpWaitService },
        { type: i1.RouterWaitService }
    ]; };
    LoaderBarComponent.propDecorators = {
        isLoading: [{ type: i0.Input }],
        containerClass: [{ type: i0.Input }],
        color: [{ type: i0.Input }]
    };

    var LoadingComponent = /** @class */ (function () {
        function LoadingComponent() {
        }
        LoadingComponent.prototype.ngOnInit = function () { };
        return LoadingComponent;
    }());
    LoadingComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-loading',
                    template: "\n    <div class=\"abp-loading\">\n      <i class=\"fa fa-spinner fa-pulse abp-spinner\"></i>\n    </div>\n  ",
                    encapsulation: i0.ViewEncapsulation.None,
                    styles: ["\n      .abp-loading {\n        position: absolute;\n        width: 100%;\n        height: 100%;\n        top: 0;\n        left: 0;\n        z-index: 1040;\n      }\n\n      .abp-loading .abp-spinner {\n        position: absolute;\n        top: 50%;\n        left: 50%;\n        font-size: 14px;\n        -moz-transform: translateX(-50%) translateY(-50%);\n        -o-transform: translateX(-50%) translateY(-50%);\n        -ms-transform: translateX(-50%) translateY(-50%);\n        -webkit-transform: translateX(-50%) translateY(-50%);\n        transform: translateX(-50%) translateY(-50%);\n      }\n    "]
                },] }
    ];
    LoadingComponent.ctorParameters = function () { return []; };

    var ConfirmationService = /** @class */ (function () {
        function ConfirmationService(contentProjectionService) {
            var _this = this;
            this.contentProjectionService = contentProjectionService;
            this.confirmation$ = new rxjs.ReplaySubject(1);
            this.clear = function (status) {
                if (status === void 0) { status = exports.Confirmation.Status.dismiss; }
                _this.confirmation$.next();
                _this.status$.next(status);
            };
        }
        ConfirmationService.prototype.setContainer = function () {
            var _this = this;
            this.containerComponentRef = this.contentProjectionService.projectContent(i1.PROJECTION_STRATEGY.AppendComponentToBody(ConfirmationComponent, {
                confirmation$: this.confirmation$,
                clear: this.clear,
            }));
            setTimeout(function () {
                _this.containerComponentRef.changeDetectorRef.detectChanges();
            }, 0);
        };
        ConfirmationService.prototype.info = function (message, title, options) {
            return this.show(message, title, 'info', options);
        };
        ConfirmationService.prototype.success = function (message, title, options) {
            return this.show(message, title, 'success', options);
        };
        ConfirmationService.prototype.warn = function (message, title, options) {
            return this.show(message, title, 'warning', options);
        };
        ConfirmationService.prototype.error = function (message, title, options) {
            return this.show(message, title, 'error', options);
        };
        ConfirmationService.prototype.show = function (message, title, severity, options) {
            if (options === void 0) { options = {}; }
            if (!this.containerComponentRef)
                this.setContainer();
            this.confirmation$.next({
                message: message,
                title: title,
                severity: severity || 'neutral',
                options: options,
            });
            this.status$ = new rxjs.Subject();
            var _a = options.dismissible, dismissible = _a === void 0 ? true : _a;
            if (dismissible)
                this.listenToEscape();
            return this.status$;
        };
        ConfirmationService.prototype.listenToEscape = function () {
            var _this = this;
            rxjs.fromEvent(document, 'keyup')
                .pipe(operators.takeUntil(this.status$), operators.debounceTime(150), operators.filter(function (key) { return key && key.key === 'Escape'; }))
                .subscribe(function (_) {
                _this.clear();
            });
        };
        return ConfirmationService;
    }());
    ConfirmationService.prov = i0__namespace.defineInjectable({ factory: function ConfirmationService_Factory() { return new ConfirmationService(i0__namespace.inject(i1__namespace.ContentProjectionService)); }, token: ConfirmationService, providedIn: "root" });
    ConfirmationService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    ConfirmationService.ctorParameters = function () { return [
        { type: i1.ContentProjectionService }
    ]; };

    // TODO: Should be documented
    var SUPPRESS_UNSAVED_CHANGES_WARNING = new i0.InjectionToken('SUPPRESS_UNSAVED_CHANGES_WARNING');

    var ModalRefService = /** @class */ (function () {
        function ModalRefService() {
            this.modalRefs = [];
        }
        ModalRefService.prototype.register = function (modal) {
            this.modalRefs.push(modal);
        };
        ModalRefService.prototype.unregister = function (modal) {
            var index = this.modalRefs.indexOf(modal);
            if (index > -1) {
                this.modalRefs.splice(index, 1);
            }
        };
        ModalRefService.prototype.dismissAll = function (mode) {
            this.modalRefs.forEach(function (modal) { return modal.dismiss(mode); });
        };
        return ModalRefService;
    }());
    ModalRefService.prov = i0__namespace.defineInjectable({ factory: function ModalRefService_Factory() { return new ModalRefService(); }, token: ModalRefService, providedIn: "root" });
    ModalRefService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];

    var ModalComponent = /** @class */ (function () {
        function ModalComponent(confirmationService, subscription, suppressUnsavedChangesWarningToken, modal, modalRefService) {
            this.confirmationService = confirmationService;
            this.subscription = subscription;
            this.suppressUnsavedChangesWarningToken = suppressUnsavedChangesWarningToken;
            this.modal = modal;
            this.modalRefService = modalRefService;
            /**
             * @deprecated Use centered property of options input instead. To be deleted in v5.0.
             */
            this.centered = false;
            /**
             * @deprecated Use windowClass property of options input instead. To be deleted in v5.0.
             */
            this.modalClass = '';
            /**
             * @deprecated Use size property of options input instead. To be deleted in v5.0.
             */
            this.size = 'lg';
            this.options = {};
            this.suppressUnsavedChangesWarning = this.suppressUnsavedChangesWarningToken;
            this.visibleChange = new i0.EventEmitter();
            this.init = new i0.EventEmitter();
            this.appear = new i0.EventEmitter();
            this.disappear = new i0.EventEmitter();
            this._visible = false;
            this._busy = false;
            this.isConfirmationOpen = false;
            this.destroy$ = new rxjs.Subject();
            this.modalIdentifier = "modal-" + i1.uuid();
            this.toggle$ = new rxjs.Subject();
            this.initToggleStream();
        }
        Object.defineProperty(ModalComponent.prototype, "visible", {
            get: function () {
                return this._visible;
            },
            set: function (value) {
                if (typeof value !== 'boolean')
                    return;
                this.toggle$.next(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ModalComponent.prototype, "busy", {
            get: function () {
                return this._busy;
            },
            set: function (value) {
                if (this.abpSubmit && this.abpSubmit instanceof ButtonComponent) {
                    this.abpSubmit.loading = value;
                }
                this._busy = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ModalComponent.prototype, "modalWindowRef", {
            get: function () {
                return document.querySelector("ngb-modal-window." + this.modalIdentifier);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ModalComponent.prototype, "isFormDirty", {
            get: function () {
                return Boolean(this.modalWindowRef.querySelector('.ng-dirty'));
            },
            enumerable: false,
            configurable: true
        });
        ModalComponent.prototype.ngOnInit = function () {
            this.modalRefService.register(this);
        };
        ModalComponent.prototype.dismiss = function (mode) {
            switch (mode) {
                case 'hard':
                    this.visible = false;
                    break;
                case 'soft':
                    this.close();
                    break;
                default:
                    break;
            }
        };
        ModalComponent.prototype.initToggleStream = function () {
            var _this = this;
            this.subscription.addOne(this.toggle$.pipe(operators.debounceTime(0), operators.distinctUntilChanged()), function (value) { return _this.toggle(value); });
        };
        ModalComponent.prototype.toggle = function (value) {
            var _this = this;
            var _a;
            this._visible = value;
            this.visibleChange.emit(value);
            if (!value) {
                (_a = this.modalRef) === null || _a === void 0 ? void 0 : _a.dismiss();
                this.disappear.emit();
                this.destroy$.next();
                return;
            }
            setTimeout(function () { return _this.listen(); }, 0);
            this.modalRef = this.modal.open(this.modalContent, Object.assign(Object.assign({
                // TODO: set size to 'lg' when removed the size variable
                size: this.size, centered: this.centered, keyboard: false, scrollable: true,
                beforeDismiss: function () {
                    if (!_this.visible)
                        return true;
                    _this.close();
                    return !_this.visible;
                }
            }, this.options), { windowClass: this.modalClass + " " + (this.options.windowClass || '') + " " + this.modalIdentifier }));
            this.appear.emit();
        };
        ModalComponent.prototype.ngOnDestroy = function () {
            this.modalRefService.unregister(this);
            this.toggle(false);
            this.destroy$.next();
        };
        ModalComponent.prototype.close = function () {
            var _this = this;
            if (this.busy)
                return;
            if (this.isFormDirty && !this.suppressUnsavedChangesWarning) {
                if (this.isConfirmationOpen)
                    return;
                this.isConfirmationOpen = true;
                this.confirmationService
                    .warn('AbpAccount::AreYouSureYouWantToCancelEditingWarningMessage', 'AbpAccount::AreYouSure', { dismissible: false })
                    .subscribe(function (status) {
                    _this.isConfirmationOpen = false;
                    if (status === exports.Confirmation.Status.confirm) {
                        _this.visible = false;
                    }
                });
            }
            else {
                this.visible = false;
            }
        };
        ModalComponent.prototype.listen = function () {
            var _this = this;
            rxjs.fromEvent(this.modalWindowRef, 'keyup')
                .pipe(operators.takeUntil(this.destroy$), operators.debounceTime(150), operators.filter(function (key) { return key && key.key === 'Escape'; }))
                .subscribe(function () { return _this.close(); });
            rxjs.fromEvent(window, 'beforeunload')
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe(function (event) {
                event.preventDefault();
                if (_this.isFormDirty && !_this.suppressUnsavedChangesWarning) {
                    event.returnValue = true;
                }
                else {
                    delete event.returnValue;
                }
            });
            setTimeout(function () {
                if (!_this.abpClose)
                    return;
                _this.warnForDeprecatedClose();
                rxjs.fromEvent(_this.abpClose.nativeElement, 'click')
                    .pipe(operators.takeUntil(_this.destroy$))
                    .subscribe(function () { return _this.close(); });
            }, 0);
            this.init.emit();
        };
        ModalComponent.prototype.warnForDeprecatedClose = function () {
            if (i0.isDevMode()) {
                console.warn('Please use abpClose directive instead of #abpClose template variable. #abpClose will be removed in v5.0');
            }
        };
        return ModalComponent;
    }());
    ModalComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-modal',
                    template: "<ng-content></ng-content>\n\n<ng-template #modalContent let-modal>\n  <div id=\"abp-modal-header\" class=\"modal-header\">\n    <ng-container *ngTemplateOutlet=\"abpHeader\"></ng-container>\n    \u200B\n    <button\n      id=\"abp-modal-close-button\"\n      type=\"button\"\n      class=\"close\"\n      aria-label=\"Close\"\n      (click)=\"modal.dismiss()\"\n    >\n      <span aria-hidden=\"true\">&times;</span>\n    </button>\n  </div>\n  <div id=\"abp-modal-body\" class=\"modal-body\">\n    <ng-container *ngTemplateOutlet=\"abpBody\"></ng-container>\n  </div>\n  <div *ngIf=\"abpFooter\" id=\"abp-modal-footer\" class=\"modal-footer\">\n    <ng-container *ngTemplateOutlet=\"abpFooter\"></ng-container>\n  </div>\n</ng-template>\n",
                    providers: [i1.SubscriptionService],
                    styles: [".modal.show{display:block!important}.modal-backdrop{opacity:.8}.modal::-webkit-scrollbar{width:7px}.modal::-webkit-scrollbar-track{background:#ddd}.modal::-webkit-scrollbar-thumb{background:#8a8686}.modal-dialog{z-index:1050}"]
                },] }
    ];
    ModalComponent.ctorParameters = function () { return [
        { type: ConfirmationService },
        { type: i1.SubscriptionService },
        { type: Boolean, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [SUPPRESS_UNSAVED_CHANGES_WARNING,] }] },
        { type: ngBootstrap.NgbModal },
        { type: ModalRefService }
    ]; };
    ModalComponent.propDecorators = {
        centered: [{ type: i0.Input }],
        modalClass: [{ type: i0.Input }],
        size: [{ type: i0.Input }],
        visible: [{ type: i0.Input }],
        busy: [{ type: i0.Input }],
        options: [{ type: i0.Input }],
        suppressUnsavedChangesWarning: [{ type: i0.Input }],
        modalContent: [{ type: i0.ViewChild, args: ['modalContent',] }],
        abpHeader: [{ type: i0.ContentChild, args: ['abpHeader', { static: false },] }],
        abpBody: [{ type: i0.ContentChild, args: ['abpBody', { static: false },] }],
        abpFooter: [{ type: i0.ContentChild, args: ['abpFooter', { static: false },] }],
        abpSubmit: [{ type: i0.ContentChild, args: [ButtonComponent, { static: false, read: ButtonComponent },] }],
        abpClose: [{ type: i0.ContentChild, args: ['abpClose', { static: false, read: i0.ElementRef },] }],
        visibleChange: [{ type: i0.Output }],
        init: [{ type: i0.Output }],
        appear: [{ type: i0.Output }],
        disappear: [{ type: i0.Output }]
    };

    var ModalCloseDirective = /** @class */ (function () {
        function ModalCloseDirective(modal) {
            this.modal = modal;
            if (!modal) {
                console.error('Please use abpClose within an abp-modal');
            }
        }
        ModalCloseDirective.prototype.onClick = function () {
            var _a;
            (_a = this.modal) === null || _a === void 0 ? void 0 : _a.close();
        };
        return ModalCloseDirective;
    }());
    ModalCloseDirective.decorators = [
        { type: i0.Directive, args: [{ selector: '[abpClose]' },] }
    ];
    ModalCloseDirective.ctorParameters = function () { return [
        { type: ModalComponent, decorators: [{ type: i0.Optional }] }
    ]; };
    ModalCloseDirective.propDecorators = {
        onClick: [{ type: i0.HostListener, args: ['click',] }]
    };

    /**
     * @deprecated To be deleted in v5.0. Use ngx-datatale instead.
     */
    var SortOrderIconComponent = /** @class */ (function () {
        function SortOrderIconComponent() {
            this.orderChange = new i0.EventEmitter();
            this.selectedSortKeyChange = new i0.EventEmitter();
        }
        Object.defineProperty(SortOrderIconComponent.prototype, "selectedSortKey", {
            get: function () {
                return this._selectedSortKey;
            },
            set: function (value) {
                this._selectedSortKey = value;
                this.selectedSortKeyChange.emit(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SortOrderIconComponent.prototype, "order", {
            get: function () {
                return this._order;
            },
            set: function (value) {
                this._order = value;
                this.orderChange.emit(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SortOrderIconComponent.prototype, "icon", {
            get: function () {
                if (this.selectedSortKey === this.sortKey)
                    return "sorting_" + this.order;
                else
                    return 'sorting';
            },
            enumerable: false,
            configurable: true
        });
        SortOrderIconComponent.prototype.sort = function (key) {
            this.selectedSortKey = key;
            switch (this.order) {
                case '':
                    this.order = 'asc';
                    this.orderChange.emit('asc');
                    break;
                case 'asc':
                    this.order = 'desc';
                    this.orderChange.emit('desc');
                    break;
                case 'desc':
                    this.order = '';
                    this.orderChange.emit('');
                    break;
            }
        };
        return SortOrderIconComponent;
    }());
    SortOrderIconComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-sort-order-icon',
                    template: "<div class=\"float-right {{ iconClass }}\">\n  <span class=\"{{ icon }}\"></span>\n</div>\n"
                },] }
    ];
    SortOrderIconComponent.propDecorators = {
        sortKey: [{ type: i0.Input }],
        selectedSortKey: [{ type: i0.Input }],
        order: [{ type: i0.Input }],
        orderChange: [{ type: i0.Output }],
        selectedSortKeyChange: [{ type: i0.Output }],
        iconClass: [{ type: i0.Input }]
    };

    var TableEmptyMessageComponent = /** @class */ (function () {
        function TableEmptyMessageComponent() {
            this.colspan = 2;
            this.localizationResource = 'AbpAccount';
            this.localizationProp = 'NoDataAvailableInDatatable';
        }
        Object.defineProperty(TableEmptyMessageComponent.prototype, "emptyMessage", {
            get: function () {
                return this.message || this.localizationResource + "::" + this.localizationProp;
            },
            enumerable: false,
            configurable: true
        });
        return TableEmptyMessageComponent;
    }());
    TableEmptyMessageComponent.decorators = [
        { type: i0.Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: '[abp-table-empty-message]',
                    template: "\n    <td class=\"text-center\" [attr.colspan]=\"colspan\">\n      {{ emptyMessage | abpLocalization }}\n    </td>\n  "
                },] }
    ];
    TableEmptyMessageComponent.propDecorators = {
        colspan: [{ type: i0.Input }],
        message: [{ type: i0.Input }],
        localizationResource: [{ type: i0.Input }],
        localizationProp: [{ type: i0.Input }]
    };

    /**
     *
     * @deprecated To be deleted in v5.0. Use ngx-datatale instead.
     */
    var TableComponent = /** @class */ (function () {
        function TableComponent() {
            var _this = this;
            this.bodyScrollLeft = 0;
            this.page = 1;
            this.trackingProp = 'id';
            this.emptyMessage = 'AbpAccount::NoDataAvailableInDatatable';
            this.pageChange = new i0.EventEmitter();
            this.trackByFn = function (_, value) {
                return typeof value === 'object' ? value[_this.trackingProp] || value : value;
            };
        }
        Object.defineProperty(TableComponent.prototype, "totalRecords", {
            get: function () {
                return this._totalRecords || this.value.length;
            },
            set: function (newValue) {
                if (newValue < 0)
                    this._totalRecords = 0;
                this._totalRecords = newValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TableComponent.prototype, "totalPages", {
            get: function () {
                if (!this.rows) {
                    return;
                }
                return Math.ceil(this.totalRecords / this.rows);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TableComponent.prototype, "slicedValue", {
            get: function () {
                if (!this.rows || this.rows >= this.value.length) {
                    return this.value;
                }
                var start = (this.page - 1) * this.rows;
                return this.value.slice(start, start + this.rows);
            },
            enumerable: false,
            configurable: true
        });
        TableComponent.prototype.ngOnInit = function () {
            this.marginCalculator = document.body.dir === 'rtl' ? rtlCalculator : ltrCalculator;
        };
        return TableComponent;
    }());
    TableComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-table',
                    template: "<div #wrapper class=\"ui-table ui-widget\">\n  <div class=\"ui-table-wrapper\">\n    <ng-container\n      *ngTemplateOutlet=\"scrollable ? scrollableTemplate : defaultTemplate\"\n    ></ng-container>\n    <div class=\"pagination-wrapper\">\n      <ngb-pagination\n        [class.op-0]=\"!totalPages\"\n        [collectionSize]=\"totalPages\"\n        [pageSize]=\"1\"\n        [page]=\"page\"\n        (pageChange)=\"pageChange.emit($event)\"\n        [maxSize]=\"3\"\n        [rotate]=\"true\"\n      ></ngb-pagination>\n    </div>\n  </div>\n</div>\n\n<ng-template #scrollableTemplate>\n  <div class=\"ui-table-scrollable-wrapper\">\n    <div class=\"ui-table-scrollable-view\"></div>\n    <div class=\"ui-table-scrollable-header ui-widget-header\">\n      <div #header class=\"ui-table-scrollable-header-box\">\n        <table class=\"ui-table-scrollable-header-table\">\n          <ng-container *ngTemplateOutlet=\"colGroup\"></ng-container>\n          <ng-container *ngTemplateOutlet=\"head\"></ng-container>\n          <tbody></tbody>\n        </table>\n      </div>\n    </div>\n    <div\n      #scrollableBody\n      (scroll)=\"header.style.margin = marginCalculator(scrollableBody)\"\n      class=\"ui-table-scrollable-body\"\n      [style.max-height]=\"scrollHeight\"\n    >\n      <table class=\"ui-table-scrollable-body-table\">\n        <ng-container *ngTemplateOutlet=\"colGroup\"></ng-container>\n        <ng-container *ngTemplateOutlet=\"body\"></ng-container>\n      </table>\n    </div>\n  </div>\n</ng-template>\n\n<ng-template #defaultTemplate>\n  <table>\n    <ng-container *ngTemplateOutlet=\"colGroup\"></ng-container>\n    <ng-container *ngTemplateOutlet=\"head\"></ng-container>\n    <ng-container *ngTemplateOutlet=\"body\"></ng-container>\n  </table>\n</ng-template>\n\n<ng-template #colGroup>\n  <ng-container *ngTemplateOutlet=\"colgroupTemplate\"></ng-container>\n</ng-template>\n\n<ng-template #head>\n  <thead class=\"ui-table-thead\">\n    <ng-container *ngTemplateOutlet=\"headerTemplate\"></ng-container>\n  </thead>\n</ng-template>\n\n<ng-template #body>\n  <tbody class=\"ui-table-tbody\" *ngIf=\"value && value.length; else emptyTemplate\">\n    <ng-template\n      #bodyTemplateWrapper\n      *ngFor=\"let val of slicedValue; let index = index; trackBy: trackByFn\"\n      [ngTemplateOutlet]=\"bodyTemplate\"\n      [ngTemplateOutletContext]=\"{ $implicit: val, rowIndex: index }\"\n    ></ng-template>\n  </tbody>\n</ng-template>\n\n<ng-template #emptyTemplate>\n  <caption class=\"ui-table-empty\">\n    {{\n      emptyMessage | abpLocalization\n    }}\n  </caption>\n</ng-template>\n",
                    encapsulation: i0.ViewEncapsulation.None,
                    styles: [".ui-table{position:relative}.ui-table .ui-table-tbody>tr:hover,.ui-table .ui-table-tbody>tr:nth-child(2n):hover{filter:brightness(90%)}.ui-table .ui-table-empty{padding:20px 0;text-align:center;border:1px solid #e0e0e0;border-top:0 solid #e0e0e0}.ui-table .ui-table-caption,.ui-table .ui-table-summary{background-color:#f4f4f4;color:#333;border:1px solid #c8c8c8;padding:.571em 1em}.ui-table .ui-table-caption,.ui-table .ui-table-summary{font-weight:700}.ui-table .ui-table-thead>tr>th{padding:.571em .857em;border:1px solid #c8c8c8;font-weight:700;color:#333;background-color:#f4f4f4}.ui-table .ui-table-tbody>tr>td{padding:.571em .857em}.ui-table .ui-table-tfoot>tr>td{padding:.571em .857em;border:1px solid #c8c8c8;font-weight:700;color:#333;background-color:#fff}.ui-table .ui-sortable-column{transition:box-shadow .2s}.ui-table .ui-sortable-column:focus{outline:0 none;outline-offset:0;box-shadow:inset 0 0 0 .2em #8dcdff}.ui-table .ui-sortable-column .ui-sortable-column-icon{color:#848484}.ui-table .ui-sortable-column:not(.ui-state-highlight):hover{background-color:#e0e0e0;color:#333}.ui-table .ui-sortable-column:not(.ui-state-highlight):hover .ui-sortable-column-icon{color:#333}.ui-table .ui-sortable-column.ui-state-highlight{background-color:#007ad9;color:#fff}.ui-table .ui-sortable-column.ui-state-highlight .ui-sortable-column-icon{color:#fff}.ui-table .ui-editable-column input{font-size:14px;font-family:Open Sans,Helvetica Neue,sans-serif}.ui-table .ui-editable-column input:focus{outline:1px solid #007ad9;outline-offset:2px}.ui-table .ui-table-tbody>tr{background-color:#fff;color:#333}.ui-table .ui-table-tbody>tr>td{background-color:inherit;border:1px solid #c8c8c8}.ui-table .ui-table-tbody>tr.ui-state-highlight{background-color:#007ad9;color:#fff}.ui-table .ui-table-tbody>tr.ui-state-highlight a{color:#fff}.ui-table .ui-table-tbody>tr.ui-contextmenu-selected{background-color:#007ad9;color:#fff}.ui-table .ui-table-tbody>tr.ui-table-dragpoint-top>td{box-shadow:inset 0 2px 0 0 #007ad9}.ui-table .ui-table-tbody>tr.ui-table-dragpoint-bottom>td{box-shadow:inset 0 -2px 0 0 #007ad9}.ui-table .ui-table-tbody>tr:nth-child(2n){background-color:#f9f9f9}.ui-table .ui-table-tbody>tr:nth-child(2n).ui-state-highlight{background-color:#007ad9;color:#fff}.ui-table .ui-table-tbody>tr:nth-child(2n).ui-state-highlight a{color:#fff}.ui-table .ui-table-tbody>tr:nth-child(2n).ui-contextmenu-selected{background-color:#007ad9;color:#fff}.ui-table.ui-table-hoverable-rows .ui-table-tbody>tr.ui-selectable-row:not(.ui-state-highlight):not(.ui-contextmenu-selected):hover{cursor:pointer;background-color:#eaeaea;color:#333}.ui-table .ui-column-resizer-helper{background-color:#007ad9}@media screen and (max-width:40em){.ui-table.ui-table-responsive .ui-table-tbody>tr>td{border:0}}.ui-table table{border-collapse:collapse;width:100%;table-layout:fixed}.ui-table .ui-table-tbody>tr>td,.ui-table .ui-table-tfoot>tr>td,.ui-table .ui-table-thead>tr>th{padding:.571em .857em}.ui-table .ui-sortable-column{cursor:pointer}.ui-table p-sorticon{vertical-align:middle}.ui-table .ui-table-auto-layout>.ui-table-wrapper{overflow-x:auto}.ui-table .ui-table-auto-layout>.ui-table-wrapper>table{table-layout:auto}.ui-table .ui-table-caption,.ui-table .ui-table-summary{padding:.25em .5em;text-align:center;font-weight:700}.ui-table .ui-table-caption{border-bottom:0}.ui-table .ui-table-summary{border-top:0}.ui-table .ui-table-scrollable-wrapper{position:relative}.ui-table .ui-table-scrollable-footer,.ui-table .ui-table-scrollable-header{overflow:hidden;border:0}.ui-table .ui-table-scrollable-body{overflow:auto;position:relative}.ui-table .ui-table-virtual-table{position:absolute}.ui-table .ui-table-loading-virtual-table{display:none}.ui-table .ui-table-frozen-view .ui-table-scrollable-body{overflow:hidden}.ui-table .ui-table-frozen-view>.ui-table-scrollable-body>table>.ui-table-tbody>tr>td:last-child{border-right:0}.ui-table .ui-table-unfrozen-view{position:absolute;top:0}.ui-table .ui-table-resizable>.ui-table-wrapper{overflow-x:auto}.ui-table .ui-table-resizable .ui-table-tbody>tr>td,.ui-table .ui-table-resizable .ui-table-tfoot>tr>td,.ui-table .ui-table-resizable .ui-table-thead>tr>th{overflow:hidden}.ui-table .ui-table-resizable .ui-resizable-column{background-clip:padding-box;position:relative}.ui-table .ui-table-resizable-fit .ui-resizable-column:last-child .ui-column-resizer{display:none}.ui-table .ui-column-resizer{display:block;position:absolute!important;top:0;right:0;margin:0;width:.5em;height:100%;padding:0;cursor:col-resize;border:1px solid transparent}.ui-table .ui-column-resizer-helper{width:1px;position:absolute;z-index:10;display:none}.ui-table .ui-table-tbody>tr>td.ui-editing-cell{padding:0}.ui-table .ui-table-tbody>tr>td.ui-editing-cell p-celleditor>*{width:100%}.ui-table .ui-table-reorder-indicator-down,.ui-table .ui-table-reorder-indicator-up{position:absolute;display:none}.ui-table .ui-table-responsive .ui-table-tbody>tr>td .ui-column-title{display:none}@media screen and (max-width:40em){.ui-table .ui-table-responsive .ui-table-tfoot>tr>td,.ui-table .ui-table-responsive .ui-table-thead>tr>th,.ui-table .ui-table-responsive colgroup{display:none!important}.ui-table .ui-table-responsive .ui-table-tbody>tr>td{text-align:left;display:block;border:0;width:100%!important;box-sizing:border-box;float:left;clear:left}.ui-table .ui-table-responsive .ui-table-tbody>tr>td .ui-column-title{padding:.4em;min-width:30%;display:inline-block;margin:-.4em 1em -.4em -.4em;font-weight:700}}.ui-table .ui-widget{font-family:Open Sans,Helvetica Neue,sans-serif;font-size:14px;text-decoration:none}.ui-table .page-item.disabled .page-link,.ui-table .page-link{background-color:transparent;border:none}.ui-table .page-item.disabled .page-link{box-shadow:none}.ui-table .pagination{margin-bottom:0}.ui-table .pagination-wrapper{display:flex;justify-content:center;border-top:0;padding:0}.ui-table .op-0{opacity:0}"]
                },] }
    ];
    TableComponent.propDecorators = {
        value: [{ type: i0.Input }],
        headerTemplate: [{ type: i0.Input }],
        bodyTemplate: [{ type: i0.Input }],
        colgroupTemplate: [{ type: i0.Input }],
        scrollHeight: [{ type: i0.Input }],
        scrollable: [{ type: i0.Input }],
        rows: [{ type: i0.Input }],
        page: [{ type: i0.Input }],
        trackingProp: [{ type: i0.Input }],
        emptyMessage: [{ type: i0.Input }],
        pageChange: [{ type: i0.Output }],
        wrapperRef: [{ type: i0.ViewChild, args: ['wrapper', { read: i0.ElementRef },] }],
        totalRecords: [{ type: i0.Input }]
    };
    function ltrCalculator(div) {
        return "0 auto 0 -" + div.scrollLeft + "px";
    }
    function rtlCalculator(div) {
        return "0 " + -(div.scrollWidth - div.clientWidth - div.scrollLeft) + "px 0 auto";
    }

    var ToastContainerComponent = /** @class */ (function () {
        function ToastContainerComponent() {
            this.toasts = [];
            this.right = '30px';
            this.bottom = '30px';
        }
        ToastContainerComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.toasts$.subscribe(function (toasts) {
                _this.toasts = _this.toastKey
                    ? toasts.filter(function (t) {
                        return t.options && t.options.containerKey !== _this.toastKey;
                    })
                    : toasts;
            });
        };
        ToastContainerComponent.prototype.trackByFunc = function (index, toast) {
            if (!toast)
                return null;
            return toast.options.id;
        };
        return ToastContainerComponent;
    }());
    ToastContainerComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-toast-container',
                    template: "<div\n  class=\"abp-toast-container\"\n  [style.top]=\"top || 'auto'\"\n  [style.right]=\"right || 'auto'\"\n  [style.bottom]=\"bottom || 'auto'\"\n  [style.left]=\"left || 'auto'\"\n  [style.display]=\"toasts.length ? 'flex' : 'none'\"\n  [@toastInOut]=\"toasts.length\"\n>\n  <abp-toast [toast]=\"toast\" *ngFor=\"let toast of toasts; trackBy: trackByFunc\"></abp-toast>\n</div>\n",
                    animations: [toastInOut],
                    styles: [".abp-toast-container{position:fixed;display:flex;flex-direction:column;align-items:center;justify-content:flex-end;min-width:350px;min-height:80px;z-index:1900}.abp-toast-container.new-on-top{flex-direction:column-reverse}"]
                },] }
    ];
    ToastContainerComponent.propDecorators = {
        top: [{ type: i0.Input }],
        right: [{ type: i0.Input }],
        bottom: [{ type: i0.Input }],
        left: [{ type: i0.Input }],
        toastKey: [{ type: i0.Input }]
    };

    var ToasterService = /** @class */ (function () {
        function ToasterService(contentProjectionService) {
            this.contentProjectionService = contentProjectionService;
            this.toasts$ = new rxjs.ReplaySubject(1);
            this.lastId = -1;
            this.toasts = [];
        }
        ToasterService.prototype.setContainer = function () {
            this.containerComponentRef = this.contentProjectionService.projectContent(i1.PROJECTION_STRATEGY.AppendComponentToBody(ToastContainerComponent, { toasts$: this.toasts$ }));
            this.containerComponentRef.changeDetectorRef.detectChanges();
        };
        /**
         * Creates an info toast with given parameters.
         * @param message Content of the toast
         * @param title Title of the toast
         * @param options Spesific style or structural options for individual toast
         */
        ToasterService.prototype.info = function (message, title, options) {
            return this.show(message, title, 'info', options);
        };
        /**
         * Creates a success toast with given parameters.
         * @param message Content of the toast
         * @param title Title of the toast
         * @param options Spesific style or structural options for individual toast
         */
        ToasterService.prototype.success = function (message, title, options) {
            return this.show(message, title, 'success', options);
        };
        /**
         * Creates a warning toast with given parameters.
         * @param message Content of the toast
         * @param title Title of the toast
         * @param options Spesific style or structural options for individual toast
         */
        ToasterService.prototype.warn = function (message, title, options) {
            return this.show(message, title, 'warning', options);
        };
        /**
         * Creates an error toast with given parameters.
         * @param message Content of the toast
         * @param title Title of the toast
         * @param options Spesific style or structural options for individual toast
         */
        ToasterService.prototype.error = function (message, title, options) {
            return this.show(message, title, 'error', options);
        };
        /**
         * Creates a toast with given parameters.
         * @param message Content of the toast
         * @param title Title of the toast
         * @param severity Sets color of the toast. "success", "warning" etc.
         * @param options Spesific style or structural options for individual toast
         */
        ToasterService.prototype.show = function (message, title, severity, options) {
            if (title === void 0) { title = null; }
            if (severity === void 0) { severity = 'neutral'; }
            if (options === void 0) { options = {}; }
            if (!this.containerComponentRef)
                this.setContainer();
            var id = ++this.lastId;
            this.toasts.push({
                message: message,
                title: title,
                severity: severity,
                options: Object.assign({ closable: true, id: id }, options),
            });
            this.toasts$.next(this.toasts);
            return id;
        };
        /**
         * Removes the toast with given id.
         * @param id ID of the toast to be removed.
         */
        ToasterService.prototype.remove = function (id) {
            this.toasts = this.toasts.filter(function (toast) { return snq__default['default'](function () { return toast.options.id; }) !== id; });
            this.toasts$.next(this.toasts);
        };
        /**
         * Removes all open toasts at once.
         */
        ToasterService.prototype.clear = function (containerKey) {
            this.toasts = !containerKey
                ? []
                : this.toasts.filter(function (toast) { return snq__default['default'](function () { return toast.options.containerKey; }) !== containerKey; });
            this.toasts$.next(this.toasts);
        };
        return ToasterService;
    }());
    ToasterService.prov = i0__namespace.defineInjectable({ factory: function ToasterService_Factory() { return new ToasterService(i0__namespace.inject(i1__namespace.ContentProjectionService)); }, token: ToasterService, providedIn: "root" });
    ToasterService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    ToasterService.ctorParameters = function () { return [
        { type: i1.ContentProjectionService }
    ]; };

    var ToastComponent = /** @class */ (function () {
        function ToastComponent(toasterService) {
            this.toasterService = toasterService;
        }
        Object.defineProperty(ToastComponent.prototype, "severityClass", {
            get: function () {
                if (!this.toast || !this.toast.severity)
                    return '';
                return "abp-toast-" + this.toast.severity;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ToastComponent.prototype, "iconClass", {
            get: function () {
                switch (this.toast.severity) {
                    case 'success':
                        return 'fa-check-circle';
                    case 'info':
                        return 'fa-info-circle';
                    case 'warning':
                        return 'fa-exclamation-triangle';
                    case 'error':
                        return 'fa-times-circle';
                    default:
                        return 'fa-exclamation-circle';
                }
            },
            enumerable: false,
            configurable: true
        });
        ToastComponent.prototype.ngOnInit = function () {
            var _this = this;
            if (snq__default['default'](function () { return _this.toast.options.sticky; }))
                return;
            var timeout = snq__default['default'](function () { return _this.toast.options.life; }) || 5000;
            setTimeout(function () {
                _this.close();
            }, timeout);
        };
        ToastComponent.prototype.close = function () {
            this.toasterService.remove(this.toast.options.id);
        };
        ToastComponent.prototype.tap = function () {
            var _a;
            if ((_a = this.toast.options) === null || _a === void 0 ? void 0 : _a.tapToDismiss)
                this.close();
        };
        return ToastComponent;
    }());
    ToastComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-toast',
                    template: "<div class=\"abp-toast\" [ngClass]=\"severityClass\" (click)=\"tap()\">\n  <div class=\"abp-toast-icon\">\n    <i class=\"fa icon\" [ngClass]=\"iconClass\"></i>\n  </div>\n  <div class=\"abp-toast-content\">\n    <button class=\"abp-toast-close-button\" (click)=\"close()\" *ngIf=\"toast.options.closable\">\n      <i class=\"fa fa-times\"></i>\n    </button>\n    <div class=\"abp-toast-title\">\n      {{ toast.title | abpLocalization: toast.options?.titleLocalizationParams }}\n    </div>\n    <p\n      class=\"abp-toast-message\"\n      [innerHTML]=\"toast.message | abpLocalization: toast.options?.messageLocalizationParams\"\n    ></p>\n  </div>\n</div>\n",
                    styles: [".abp-toast{display:grid;grid-template-columns:50px 1fr;gap:10px;margin:5px 0;padding:10px;border-radius:0;width:350px;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;z-index:9999;border:2px solid #f0f0f0;background-color:#f0f0f0;color:#000;box-shadow:0 0 10px -5px rgba(0,0,0,.4);opacity:1}.abp-toast:hover{border:2px solid #e3e3e3;background-color:#e3e3e3;box-shadow:0 0 15px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-success{border:2px solid #51a351;background-color:#51a351;color:#fff;box-shadow:0 0 10px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-success:hover{border:2px solid #499249;background-color:#499249;box-shadow:0 0 15px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-info{border:2px solid #2f96b4;background-color:#2f96b4;color:#fff;box-shadow:0 0 10px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-info:hover{border:2px solid #2a85a0;background-color:#2a85a0;box-shadow:0 0 15px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-warning{border:2px solid #f89406;background-color:#f89406;color:#fff;box-shadow:0 0 10px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-warning:hover{border:2px solid #df8505;background-color:#df8505;box-shadow:0 0 15px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-error{border:2px solid #bd362f;background-color:#bd362f;color:#fff;box-shadow:0 0 10px -5px rgba(0,0,0,.4)}.abp-toast.abp-toast-error:hover{border:2px solid #a9302a;background-color:#a9302a;box-shadow:0 0 15px -5px rgba(0,0,0,.4)}.abp-toast .abp-toast-icon{display:flex;align-items:center;justify-content:center}.abp-toast .abp-toast-icon .icon{font-size:36px}.abp-toast .abp-toast-content{position:relative}.abp-toast .abp-toast-content .abp-toast-close-button{position:absolute;top:0;right:0;display:flex;align-items:center;justify-content:center;margin:0;padding:0 5px 0 0;width:25px;height:25px;border:none;border-radius:50%;background:transparent;color:inherit}.abp-toast .abp-toast-content .abp-toast-close-button:focus{outline:none}.abp-toast .abp-toast-content .abp-toast-title{margin:0;padding:0;font-size:1rem;font-weight:600}.abp-toast .abp-toast-content .abp-toast-message{margin:0;padding:0;max-width:240px}"]
                },] }
    ];
    ToastComponent.ctorParameters = function () { return [
        { type: ToasterService }
    ]; };
    ToastComponent.propDecorators = {
        toast: [{ type: i0.Input }]
    };

    var styles = "\n.is-invalid .form-control {\n  border-color: #dc3545;\n  border-style: solid !important;\n  padding-right: calc(1.5em + .75rem);\n  background-image: url(data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' width='12' height='12' fill='none' stroke='%23dc3545' viewBox='0 0 12 12'%3e%3ccircle cx='6' cy='6' r='4.5'/%3e%3cpath stroke-linejoin='round' d='M5.8 3.6h.4L6 6.5z'/%3e%3ccircle cx='6' cy='8.2' r='.6' fill='%23dc3545' stroke='none'/%3e%3c/svg%3e);\n  background-repeat: no-repeat;\n  background-position: right calc(.375em + .1875rem) center;\n  background-size: calc(.75em + .375rem) calc(.75em + .375rem);\n}\n\n.is-invalid .invalid-feedback,\n.is-invalid + * .invalid-feedback {\n  display: block;\n}\n\n.data-tables-filter {\n  text-align: right;\n}\n\n[dir=rtl] .data-tables-filter {\n  text-align: left;\n}\n\n.pointer {\n  cursor: pointer;\n}\n\n.navbar .dropdown-submenu a::after {\n  transform: rotate(-90deg);\n  position: absolute;\n  right: 16px;\n  top: 18px;\n}\n\n.navbar .dropdown-menu {\n  min-width: 215px;\n}\n\n.datatable-scroll {\n  margin-bottom: 5px !important;\n  width: unset !important;\n}\n\n.ui-table-scrollable-body::-webkit-scrollbar {\n  height: 5px !important;\n  width: 5px !important;\n}\n\n.ui-table-scrollable-body::-webkit-scrollbar-track {\n  background: #ddd;\n}\n\n.ui-table-scrollable-body::-webkit-scrollbar-thumb {\n  background: #8a8686;\n}\n\n.abp-ellipsis-inline {\n  display: inline-block;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n}\n\n.abp-ellipsis {\n  overflow: hidden !important;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n}\n\n.ui-widget-overlay {\n  z-index: 1000;\n}\n\n.color-white {\n  color: #FFF !important;\n}\n\n.custom-checkbox > label {\n  cursor: pointer;\n}\n\n/* <animations */\n\n.fade-in-top {\n  animation: fadeInTop 0.2s ease-in-out;\n}\n\n.fade-out-top {\n  animation: fadeOutTop 0.2s ease-in-out;\n}\n\n.abp-collapsed-height {\n  -moz-transition: max-height linear 0.35s;\n  -ms-transition: max-height linear 0.35s;\n  -o-transition: max-height linear 0.35s;\n  -webkit-transition: max-height linear 0.35s;\n  overflow:hidden;\n  transition:max-height 0.35s linear;\n  height:auto;\n  max-height: 0;\n}\n\n.abp-mh-25 {\n  max-height: 25vh;\n}\n\n.abp-mh-50 {\n  transition:max-height 0.65s linear;\n  max-height: 50vh;\n}\n\n.abp-mh-75 {\n  transition:max-height 0.85s linear;\n  max-height: 75vh;\n}\n\n.abp-mh-100 {\n  transition:max-height 1s linear;\n  max-height: 100vh;\n}\n\n[class^=\"sorting\"] {\n  opacity: .3;\n  cursor: pointer;\n}\n[class^=\"sorting\"]:before {\n  right: 0.5rem;\n  content: \"\u2191\";\n}\n[class^=\"sorting\"]:after {\n  right: 0.5rem;\n  content: \"\u2193\";\n}\n\n.sorting_desc {\n  opacity: 1;\n}\n.sorting_desc:before {\n  opacity: .3;\n}\n\n.sorting_asc {\n  opacity: 1;\n}\n.sorting_asc:after {\n  opacity: .3;\n}\n.ngx-datatable.material {\n  box-shadow: none;\n}\nngb-typeahead-window, ngb-typeahead-window.dropdown-menu {\n  max-height: 25em;\n  overflow-y: scroll !important;\n  z-index: 1050;\n}\n\n@keyframes fadeInTop {\n  from {\n    transform: translateY(-5px);\n    opacity: 0;\n  }\n\n  to {\n    transform: translateY(0px);\n    opacity: 1;\n  }\n}\n\n@keyframes fadeOutTop {\n  to {\n    transform: translateY(-5px);\n    opacity: 0;\n  }\n}\n\n/* </animations */\n\n";
    var BOOTSTRAP = 'bootstrap-{{dir}}.min.css';

    var DEFAULT_VALIDATION_BLUEPRINTS = {
        creditCard: 'AbpValidation::ThisFieldIsNotAValidCreditCardNumber.',
        email: 'AbpValidation::ThisFieldIsNotAValidEmailAddress.',
        invalid: 'AbpValidation::ThisFieldIsNotValid.',
        max: 'AbpValidation::ThisFieldMustBeBetween{0}And{1}[{{ min }},{{ max }}]',
        maxlength: 'AbpValidation::ThisFieldMustBeAStringOrArrayTypeWithAMaximumLengthOf{0}[{{ requiredLength }}]',
        min: 'AbpValidation::ThisFieldMustBeBetween{0}And{1}[{{ min }},{{ max }}]',
        minlength: 'AbpValidation::ThisFieldMustBeAStringOrArrayTypeWithAMinimumLengthOf{0}[{{ requiredLength }}]',
        ngbDate: 'AbpValidation::ThisFieldIsNotValid.',
        passwordMismatch: 'AbpIdentity::Volo.Abp.Identity:PasswordConfirmationFailed',
        range: 'AbpValidation::ThisFieldMustBeBetween{0}And{1}[{{ min }},{{ max }}]',
        required: 'AbpValidation::ThisFieldIsRequired.',
        url: 'AbpValidation::ThisFieldIsNotAValidFullyQualifiedHttpHttpsOrFtpUrl',
    };

    var LoadingDirective = /** @class */ (function () {
        function LoadingDirective(elRef, vcRef, cdRes, injector, renderer) {
            this.elRef = elRef;
            this.vcRef = vcRef;
            this.cdRes = cdRes;
            this.injector = injector;
            this.renderer = renderer;
            this.position = 'relative';
            this.delay = 0;
        }
        Object.defineProperty(LoadingDirective.prototype, "loading", {
            get: function () {
                return this._loading;
            },
            set: function (newValue) {
                var _this = this;
                setTimeout(function () {
                    if (!newValue && _this.timerSubscription) {
                        _this.timerSubscription.unsubscribe();
                        _this.timerSubscription = null;
                        _this._loading = newValue;
                        if (_this.rootNode) {
                            _this.renderer.removeChild(_this.rootNode.parentElement, _this.rootNode);
                            _this.rootNode = null;
                        }
                        return;
                    }
                    _this.timerSubscription = rxjs.timer(_this.delay)
                        .pipe(operators.take(1))
                        .subscribe(function () {
                        if (!_this.componentRef) {
                            _this.componentRef = _this.cdRes
                                .resolveComponentFactory(LoadingComponent)
                                .create(_this.injector);
                        }
                        if (newValue && !_this.rootNode) {
                            _this.rootNode = _this.componentRef.hostView.rootNodes[0];
                            _this.targetElement.appendChild(_this.rootNode);
                        }
                        else if (_this.rootNode) {
                            _this.renderer.removeChild(_this.rootNode.parentElement, _this.rootNode);
                            _this.rootNode = null;
                        }
                        _this._loading = newValue;
                        _this.timerSubscription = null;
                    });
                }, 0);
            },
            enumerable: false,
            configurable: true
        });
        LoadingDirective.prototype.ngOnInit = function () {
            if (!this.targetElement) {
                var _a = this.elRef.nativeElement, offsetHeight = _a.offsetHeight, offsetWidth = _a.offsetWidth;
                if (!offsetHeight && !offsetWidth && this.elRef.nativeElement.children.length) {
                    this.targetElement = this.elRef.nativeElement.children[0];
                }
                else {
                    this.targetElement = this.elRef.nativeElement;
                }
            }
        };
        LoadingDirective.prototype.ngOnDestroy = function () {
            if (this.timerSubscription) {
                this.timerSubscription.unsubscribe();
            }
        };
        return LoadingDirective;
    }());
    LoadingDirective.decorators = [
        { type: i0.Directive, args: [{ selector: '[abpLoading]' },] }
    ];
    LoadingDirective.ctorParameters = function () { return [
        { type: i0.ElementRef },
        { type: i0.ViewContainerRef },
        { type: i0.ComponentFactoryResolver },
        { type: i0.Injector },
        { type: i0.Renderer2 }
    ]; };
    LoadingDirective.propDecorators = {
        position: [{ type: i0.HostBinding, args: ['style.position',] }],
        loading: [{ type: i0.Input, args: ['abpLoading',] }],
        targetElement: [{ type: i0.Input, args: ['abpLoadingTargetElement',] }],
        delay: [{ type: i0.Input, args: ['abpLoadingDelay',] }]
    };

    var NgxDatatableDefaultDirective = /** @class */ (function () {
        function NgxDatatableDefaultDirective(table, document) {
            this.table = table;
            this.document = document;
            this.subscription = new rxjs.Subscription();
            this.resizeDiff = 0;
            this.class = 'material bordered';
            this.table.columnMode = ngxDatatable.ColumnMode.force;
            this.table.footerHeight = 50;
            this.table.headerHeight = 50;
            this.table.rowHeight = 'auto';
            this.table.scrollbarH = true;
            this.table.virtualization = false;
        }
        Object.defineProperty(NgxDatatableDefaultDirective.prototype, "classes", {
            get: function () {
                return "ngx-datatable " + this.class;
            },
            enumerable: false,
            configurable: true
        });
        NgxDatatableDefaultDirective.prototype.fixHorizontalGap = function (scroller) {
            var _a = this.document, body = _a.body, documentElement = _a.documentElement;
            if (documentElement.scrollHeight !== documentElement.clientHeight) {
                if (this.resizeDiff === 0) {
                    this.resizeDiff = window.innerWidth - body.offsetWidth;
                    scroller.scrollWidth -= this.resizeDiff;
                }
            }
            else {
                scroller.scrollWidth += this.resizeDiff;
                this.resizeDiff = 0;
            }
        };
        NgxDatatableDefaultDirective.prototype.fixStyleOnWindowResize = function () {
            var _this = this;
            // avoided @HostListener('window:resize') in favor of performance
            var subscription = rxjs.fromEvent(window, 'resize')
                .pipe(operators.debounceTime(500))
                .subscribe(function () {
                var scroller = _this.table.bodyComponent.scroller;
                if (!scroller)
                    return;
                _this.fixHorizontalGap(scroller);
            });
            this.subscription.add(subscription);
        };
        NgxDatatableDefaultDirective.prototype.ngAfterViewInit = function () {
            this.fixStyleOnWindowResize();
        };
        NgxDatatableDefaultDirective.prototype.ngOnDestroy = function () {
            this.subscription.unsubscribe();
        };
        return NgxDatatableDefaultDirective;
    }());
    NgxDatatableDefaultDirective.decorators = [
        { type: i0.Directive, args: [{
                    // tslint:disable-next-line
                    selector: 'ngx-datatable[default]',
                    exportAs: 'ngxDatatableDefault',
                },] }
    ];
    NgxDatatableDefaultDirective.ctorParameters = function () { return [
        { type: ngxDatatable.DatatableComponent },
        { type: undefined, decorators: [{ type: i0.Inject, args: [common.DOCUMENT,] }] }
    ]; };
    NgxDatatableDefaultDirective.propDecorators = {
        class: [{ type: i0.Input }],
        classes: [{ type: i0.HostBinding, args: ['class',] }]
    };

    var defaultNgxDatatableMessages = {
        emptyMessage: 'AbpUi::NoDataAvailableInDatatable',
        totalMessage: 'AbpUi::Total',
        selectedMessage: 'AbpUi::Selected',
    };
    var NGX_DATATABLE_MESSAGES = new i0.InjectionToken('NGX_DATATABLE_MESSAGES');

    var NgxDatatableListDirective = /** @class */ (function () {
        function NgxDatatableListDirective(table, cdRef, localizationService, ngxDatatableMessages) {
            this.table = table;
            this.cdRef = cdRef;
            this.localizationService = localizationService;
            this.ngxDatatableMessages = ngxDatatableMessages;
            this.subscription = new rxjs.Subscription();
            this.querySubscription = new rxjs.Subscription();
            this.setInitialValues();
        }
        NgxDatatableListDirective.prototype.setInitialValues = function () {
            this.table.externalPaging = true;
            this.table.externalSorting = true;
            var _a = this.ngxDatatableMessages || defaultNgxDatatableMessages, emptyMessage = _a.emptyMessage, selectedMessage = _a.selectedMessage, totalMessage = _a.totalMessage;
            this.table.messages = {
                emptyMessage: this.localizationService.instant(emptyMessage),
                totalMessage: this.localizationService.instant(totalMessage),
                selectedMessage: this.localizationService.instant(selectedMessage),
            };
        };
        NgxDatatableListDirective.prototype.subscribeToPage = function () {
            var _this = this;
            var sub = this.table.page.subscribe(function (_a) {
                var offset = _a.offset;
                _this.list.page = offset;
                _this.table.offset = offset;
            });
            this.subscription.add(sub);
        };
        NgxDatatableListDirective.prototype.subscribeToSort = function () {
            var _this = this;
            var sub = this.table.sort.subscribe(function (_a) {
                var _b = __read(_a.sorts, 1), _c = _b[0], prop = _c.prop, dir = _c.dir;
                if (prop === _this.list.sortKey && _this.list.sortOrder === 'desc') {
                    _this.list.sortKey = '';
                    _this.list.sortOrder = '';
                    _this.table.sorts = [];
                    _this.cdRef.detectChanges();
                }
                else {
                    _this.list.sortKey = prop;
                    _this.list.sortOrder = dir;
                }
            });
            this.subscription.add(sub);
        };
        NgxDatatableListDirective.prototype.subscribeToQuery = function () {
            var _this = this;
            if (!this.querySubscription.closed)
                this.querySubscription.unsubscribe();
            this.querySubscription = this.list.query$.subscribe(function () {
                var offset = _this.list.page;
                if (_this.table.offset !== offset)
                    _this.table.offset = offset;
            });
        };
        NgxDatatableListDirective.prototype.ngOnChanges = function (_a) {
            var list = _a.list;
            this.subscribeToQuery();
            if (!list.firstChange)
                return;
            var _b = list.currentValue, maxResultCount = _b.maxResultCount, page = _b.page;
            this.table.limit = maxResultCount;
            this.table.offset = page;
        };
        NgxDatatableListDirective.prototype.ngOnDestroy = function () {
            this.subscription.unsubscribe();
            this.querySubscription.unsubscribe();
        };
        NgxDatatableListDirective.prototype.ngOnInit = function () {
            this.subscribeToPage();
            this.subscribeToSort();
        };
        return NgxDatatableListDirective;
    }());
    NgxDatatableListDirective.decorators = [
        { type: i0.Directive, args: [{
                    // tslint:disable-next-line
                    selector: 'ngx-datatable[list]',
                    exportAs: 'ngxDatatableList',
                },] }
    ];
    NgxDatatableListDirective.ctorParameters = function () { return [
        { type: ngxDatatable.DatatableComponent },
        { type: i0.ChangeDetectorRef },
        { type: i1.LocalizationService },
        { type: undefined, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [NGX_DATATABLE_MESSAGES,] }] }
    ]; };
    NgxDatatableListDirective.propDecorators = {
        list: [{ type: i0.Input }]
    };

    /**
     *
     * @deprecated To be deleted in v5.0
     */
    var TableSortDirective = /** @class */ (function () {
        function TableSortDirective(abpTable, sortPipe, cdRef) {
            this.abpTable = abpTable;
            this.sortPipe = sortPipe;
            this.cdRef = cdRef;
            this.value = [];
        }
        Object.defineProperty(TableSortDirective.prototype, "table", {
            get: function () {
                var _this = this;
                return (this.abpTable || snq__default['default'](function () { return _this.cdRef['_view'].component; }) || snq__default['default'](function () { return _this.cdRef['context']; }) // 'context' for ivy
                );
            },
            enumerable: false,
            configurable: true
        });
        TableSortDirective.prototype.ngOnChanges = function (_a) {
            var value = _a.value, abpTableSort = _a.abpTableSort;
            if (this.table && (value || abpTableSort)) {
                this.abpTableSort = this.abpTableSort || {};
                this.table.value = this.sortPipe.transform(clone__default['default'](this.value), this.abpTableSort.order, this.abpTableSort.key);
            }
        };
        return TableSortDirective;
    }());
    TableSortDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[abpTableSort]',
                    providers: [i1.SortPipe],
                },] }
    ];
    TableSortDirective.ctorParameters = function () { return [
        { type: TableComponent, decorators: [{ type: i0.Host }, { type: i0.Optional }, { type: i0.Self }] },
        { type: i1.SortPipe },
        { type: i0.ChangeDetectorRef }
    ]; };
    TableSortDirective.propDecorators = {
        abpTableSort: [{ type: i0.Input }],
        value: [{ type: i0.Input }]
    };

    var EllipsisDirective = /** @class */ (function () {
        function EllipsisDirective(cdRef, elRef) {
            this.cdRef = cdRef;
            this.elRef = elRef;
            this.enabled = true;
        }
        Object.defineProperty(EllipsisDirective.prototype, "inlineClass", {
            get: function () {
                return this.enabled && this.width;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(EllipsisDirective.prototype, "class", {
            get: function () {
                return this.enabled && !this.width;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(EllipsisDirective.prototype, "maxWidth", {
            get: function () {
                return this.enabled && this.width ? this.width || '170px' : undefined;
            },
            enumerable: false,
            configurable: true
        });
        EllipsisDirective.prototype.ngAfterViewInit = function () {
            this.title = this.title || this.elRef.nativeElement.innerText;
            this.cdRef.detectChanges();
        };
        return EllipsisDirective;
    }());
    EllipsisDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[abpEllipsis]',
                },] }
    ];
    EllipsisDirective.ctorParameters = function () { return [
        { type: i0.ChangeDetectorRef },
        { type: i0.ElementRef }
    ]; };
    EllipsisDirective.propDecorators = {
        width: [{ type: i0.Input, args: ['abpEllipsis',] }],
        title: [{ type: i0.HostBinding, args: ['title',] }, { type: i0.Input }],
        enabled: [{ type: i0.Input, args: ['abpEllipsisEnabled',] }],
        inlineClass: [{ type: i0.HostBinding, args: ['class.abp-ellipsis-inline',] }],
        class: [{ type: i0.HostBinding, args: ['class.abp-ellipsis',] }],
        maxWidth: [{ type: i0.HostBinding, args: ['style.max-width',] }]
    };
    var EllipsisModule = /** @class */ (function () {
        function EllipsisModule() {
        }
        return EllipsisModule;
    }());
    EllipsisModule.decorators = [
        { type: i0.NgModule, args: [{
                    exports: [EllipsisDirective],
                    declarations: [EllipsisDirective],
                },] }
    ];

    function httpErrorConfigFactory(config) {
        if (config === void 0) { config = {}; }
        if (config.errorScreen && config.errorScreen.component && !config.errorScreen.forWhichErrors) {
            config.errorScreen.forWhichErrors = [401, 403, 404, 500];
        }
        return Object.assign({ skipHandledErrorCodes: [], errorScreen: {} }, config);
    }
    var HTTP_ERROR_CONFIG = new i0.InjectionToken('HTTP_ERROR_CONFIG');
    var HTTP_ERROR_HANDLER = new i0.InjectionToken('HTTP_ERROR_HANDLER');

    var DEFAULT_ERROR_MESSAGES = {
        defaultError: {
            title: 'An error has occurred!',
            details: 'Error detail not sent by server.',
        },
        defaultError401: {
            title: 'You are not authenticated!',
            details: 'You should be authenticated (sign in) in order to perform this operation.',
        },
        defaultError403: {
            title: 'You are not authorized!',
            details: 'You are not allowed to perform this operation.',
        },
        defaultError404: {
            title: 'Resource not found!',
            details: 'The resource requested could not found on the server.',
        },
        defaultError500: {
            title: 'Internal server error',
            details: 'Error detail not sent by server.',
        },
    };
    var DEFAULT_ERROR_LOCALIZATIONS = {
        defaultError: {
            title: 'AbpUi::DefaultErrorMessage',
            details: 'AbpUi::DefaultErrorMessageDetail',
        },
        defaultError401: {
            title: 'AbpUi::DefaultErrorMessage401',
            details: 'AbpUi::DefaultErrorMessage401Detail',
        },
        defaultError403: {
            title: 'AbpUi::DefaultErrorMessage403',
            details: 'AbpUi::DefaultErrorMessage403Detail',
        },
        defaultError404: {
            title: 'AbpUi::DefaultErrorMessage404',
            details: 'AbpUi::DefaultErrorMessage404Detail',
        },
        defaultError500: {
            title: 'AbpUi::500Message',
            details: 'AbpUi::DefaultErrorMessage',
        },
    };
    var ErrorHandler = /** @class */ (function () {
        function ErrorHandler(actions, routerEvents, confirmationService, cfRes, rendererFactory, injector, httpErrorConfig) {
            var _this = this;
            this.actions = actions;
            this.routerEvents = routerEvents;
            this.confirmationService = confirmationService;
            this.cfRes = cfRes;
            this.rendererFactory = rendererFactory;
            this.injector = injector;
            this.httpErrorConfig = httpErrorConfig;
            this.httpErrorHandler = this.injector.get(HTTP_ERROR_HANDLER, function (_, err) { return rxjs.throwError(err); });
            this.executeErrorHandler = function (error) {
                var returnValue = _this.httpErrorHandler(_this.injector, error);
                return (returnValue instanceof rxjs.Observable ? returnValue : rxjs.of(null)).pipe(operators.catchError(function (err) {
                    _this.handleError(err);
                    return rxjs.of(null);
                }));
            };
            this.filterRestErrors = function (_a) {
                var status = _a.status;
                if (typeof status !== 'number')
                    return false;
                return _this.httpErrorConfig.skipHandledErrorCodes.findIndex(function (code) { return code === status; }) < 0;
            };
            this.filterRouteErrors = function (navigationError) {
                return (snq__default['default'](function () { return navigationError.error.message.indexOf('Cannot match') > -1; }) &&
                    _this.httpErrorConfig.skipHandledErrorCodes.findIndex(function (code) { return code === 404; }) < 0);
            };
            this.listenToRestError();
            this.listenToRouterError();
            this.listenToRouterDataResolved();
        }
        ErrorHandler.prototype.listenToRouterError = function () {
            var _this = this;
            this.routerEvents
                .getNavigationEvents('Error')
                .pipe(operators.filter(this.filterRouteErrors))
                .subscribe(function () { return _this.show404Page(); });
        };
        ErrorHandler.prototype.listenToRouterDataResolved = function () {
            var _this = this;
            this.routerEvents
                .getEvents(router.ResolveEnd)
                .pipe(operators.filter(function () { return !!_this.componentRef; }))
                .subscribe(function () {
                _this.componentRef.destroy();
                _this.componentRef = null;
            });
        };
        ErrorHandler.prototype.listenToRestError = function () {
            this.actions
                .pipe(i1$1.ofActionSuccessful(i1.RestOccurError), operators.map(function (action) { return action.payload; }), operators.filter(this.filterRestErrors), operators.switchMap(this.executeErrorHandler))
                .subscribe();
        };
        ErrorHandler.prototype.handleError = function (err) {
            var _this = this;
            var body = snq__default['default'](function () { return err.error.error; }, {
                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError.title,
                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError.title,
            });
            if (err instanceof http.HttpErrorResponse && err.headers.get('_AbpErrorFormat')) {
                var confirmation$ = this.showError(null, null, body);
                if (err.status === 401) {
                    confirmation$.subscribe(function () {
                        _this.navigateToLogin();
                    });
                }
            }
            else {
                switch (err.status) {
                    case 401:
                        this.canCreateCustomError(401)
                            ? this.show401Page()
                            : this.showError({
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError401.title,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError401.title,
                            }, {
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError401.details,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError401.details,
                            }).subscribe(function () { return _this.navigateToLogin(); });
                        break;
                    case 403:
                        this.createErrorComponent({
                            title: {
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError403.title,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError403.title,
                            },
                            details: {
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError403.details,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError403.details,
                            },
                            status: 403,
                        });
                        break;
                    case 404:
                        this.canCreateCustomError(404)
                            ? this.show404Page()
                            : this.showError({
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError404.details,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError404.details,
                            }, {
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError404.title,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError404.title,
                            });
                        break;
                    case 500:
                        this.createErrorComponent({
                            title: {
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError500.title,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError500.title,
                            },
                            details: {
                                key: DEFAULT_ERROR_LOCALIZATIONS.defaultError500.details,
                                defaultValue: DEFAULT_ERROR_MESSAGES.defaultError500.details,
                            },
                            status: 500,
                        });
                        break;
                    case 0:
                        if (err.statusText === 'Unknown Error') {
                            this.createErrorComponent({
                                title: {
                                    key: DEFAULT_ERROR_LOCALIZATIONS.defaultError.title,
                                    defaultValue: DEFAULT_ERROR_MESSAGES.defaultError.title,
                                },
                                details: err.message,
                                isHomeShow: false,
                            });
                        }
                        break;
                    default:
                        this.showError({
                            key: DEFAULT_ERROR_LOCALIZATIONS.defaultError.details,
                            defaultValue: DEFAULT_ERROR_MESSAGES.defaultError.details,
                        }, {
                            key: DEFAULT_ERROR_LOCALIZATIONS.defaultError.title,
                            defaultValue: DEFAULT_ERROR_MESSAGES.defaultError.title,
                        });
                        break;
                }
            }
        };
        ErrorHandler.prototype.show401Page = function () {
            this.createErrorComponent({
                title: {
                    key: DEFAULT_ERROR_LOCALIZATIONS.defaultError401.title,
                    defaultValue: DEFAULT_ERROR_MESSAGES.defaultError401.title,
                },
                status: 401,
            });
        };
        ErrorHandler.prototype.show404Page = function () {
            this.createErrorComponent({
                title: {
                    key: DEFAULT_ERROR_LOCALIZATIONS.defaultError404.title,
                    defaultValue: DEFAULT_ERROR_MESSAGES.defaultError404.title,
                },
                status: 404,
            });
        };
        ErrorHandler.prototype.showError = function (message, title, body) {
            if (body) {
                if (body.details) {
                    message = body.details;
                    title = body.message;
                }
                else if (body.message) {
                    title = {
                        key: DEFAULT_ERROR_LOCALIZATIONS.defaultError.title,
                        defaultValue: DEFAULT_ERROR_MESSAGES.defaultError.title,
                    };
                    message = body.message;
                }
                else {
                    message = body.message || {
                        key: DEFAULT_ERROR_LOCALIZATIONS.defaultError.title,
                        defaultValue: DEFAULT_ERROR_MESSAGES.defaultError.title,
                    };
                }
            }
            return this.confirmationService.error(message, title, {
                hideCancelBtn: true,
                yesText: 'AbpAccount::Close',
            });
        };
        ErrorHandler.prototype.navigateToLogin = function () {
            this.injector.get(i1.AuthService).navigateToLogin();
        };
        ErrorHandler.prototype.createErrorComponent = function (instance) {
            var _this = this;
            var renderer = this.rendererFactory.createRenderer(null, null);
            var host = renderer.selectRootElement(document.body, true);
            this.componentRef = this.cfRes
                .resolveComponentFactory(HttpErrorWrapperComponent)
                .create(this.injector);
            for (var key in instance) {
                /* istanbul ignore else */
                if (this.componentRef.instance.hasOwnProperty(key)) {
                    this.componentRef.instance[key] = instance[key];
                }
            }
            this.componentRef.instance.hideCloseIcon = this.httpErrorConfig.errorScreen.hideCloseIcon;
            var appRef = this.injector.get(i0.ApplicationRef);
            if (this.canCreateCustomError(instance.status)) {
                this.componentRef.instance.cfRes = this.cfRes;
                this.componentRef.instance.appRef = appRef;
                this.componentRef.instance.injector = this.injector;
                this.componentRef.instance.customComponent = this.httpErrorConfig.errorScreen.component;
            }
            appRef.attachView(this.componentRef.hostView);
            renderer.appendChild(host, this.componentRef.hostView.rootNodes[0]);
            var destroy$ = new rxjs.Subject();
            this.componentRef.instance.destroy$ = destroy$;
            destroy$.subscribe(function () {
                _this.componentRef.destroy();
                _this.componentRef = null;
            });
        };
        ErrorHandler.prototype.canCreateCustomError = function (status) {
            var _this = this;
            return snq__default['default'](function () { return _this.httpErrorConfig.errorScreen.component &&
                _this.httpErrorConfig.errorScreen.forWhichErrors.indexOf(status) > -1; });
        };
        return ErrorHandler;
    }());
    ErrorHandler.prov = i0__namespace.defineInjectable({ factory: function ErrorHandler_Factory() { return new ErrorHandler(i0__namespace.inject(i1__namespace$1.Actions), i0__namespace.inject(i1__namespace.RouterEvents), i0__namespace.inject(ConfirmationService), i0__namespace.inject(i0__namespace.ComponentFactoryResolver), i0__namespace.inject(i0__namespace.RendererFactory2), i0__namespace.inject(i0__namespace.INJECTOR), i0__namespace.inject("HTTP_ERROR_CONFIG")); }, token: ErrorHandler, providedIn: "root" });
    ErrorHandler.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    ErrorHandler.ctorParameters = function () { return [
        { type: i1$1.Actions },
        { type: i1.RouterEvents },
        { type: ConfirmationService },
        { type: i0.ComponentFactoryResolver },
        { type: i0.RendererFactory2 },
        { type: i0.Injector },
        { type: undefined, decorators: [{ type: i0.Inject, args: ['HTTP_ERROR_CONFIG',] }] }
    ]; };

    var LAZY_STYLES = new i0.InjectionToken('LAZY_STYLES');

    var LazyStyleHandler = /** @class */ (function () {
        function LazyStyleHandler(injector) {
            this._dir = 'ltr';
            this.loaded = new Map();
            this.setStyles(injector);
            this.setLazyLoad(injector);
            this.listenToLanguageChanges(injector);
        }
        Object.defineProperty(LazyStyleHandler.prototype, "dir", {
            get: function () {
                return this._dir;
            },
            set: function (dir) {
                if (dir === this._dir)
                    return;
                this.switchCSS(dir);
                this.setBodyDir(dir);
                this._dir = dir;
            },
            enumerable: false,
            configurable: true
        });
        LazyStyleHandler.prototype.getHrefFromLink = function (link) {
            if (!link)
                return '';
            var a = document.createElement('a');
            a.href = link.href;
            return a.pathname.replace(/^\//, '');
        };
        LazyStyleHandler.prototype.getLoadedBootstrap = function () {
            var href = createLazyStyleHref(BOOTSTRAP, this.dir);
            var selector = "[href*=\"" + href.replace(/\.css$/, '') + "\"]";
            var link = document.querySelector(selector);
            return { href: href, link: link };
        };
        LazyStyleHandler.prototype.listenToLanguageChanges = function (injector) {
            var _this = this;
            var l10n = injector.get(i1.LocalizationService);
            // will always listen, no need to unsubscribe
            l10n.languageChange$.pipe(operators.startWith(l10n.currentLang)).subscribe(function (locale) {
                _this.dir = i1.getLocaleDirection(locale);
            });
        };
        LazyStyleHandler.prototype.setBodyDir = function (dir) {
            document.body.dir = dir;
        };
        LazyStyleHandler.prototype.setLazyLoad = function (injector) {
            this.lazyLoad = injector.get(i1.LazyLoadService);
            var _a = this.getLoadedBootstrap(), href = _a.href, link = _a.link;
            this.lazyLoad.loaded.set(href, link);
        };
        LazyStyleHandler.prototype.setStyles = function (injector) {
            this.styles = injector.get(LAZY_STYLES, [BOOTSTRAP]);
        };
        LazyStyleHandler.prototype.switchCSS = function (dir) {
            var _this = this;
            this.styles.forEach(function (style) {
                var oldHref = createLazyStyleHref(style, _this.dir);
                var newHref = createLazyStyleHref(style, dir);
                var link = _this.loaded.get(newHref);
                var href = _this.getHrefFromLink(link) || newHref;
                var strategy = i1.LOADING_STRATEGY.PrependAnonymousStyleToHead(href);
                _this.lazyLoad.load(strategy).subscribe(function () {
                    var oldLink = _this.lazyLoad.loaded.get(oldHref);
                    _this.loaded.delete(newHref);
                    _this.loaded.set(oldHref, oldLink);
                    var newLink = _this.lazyLoad.loaded.get(href);
                    _this.lazyLoad.loaded.delete(href);
                    _this.lazyLoad.loaded.set(newHref, newLink);
                    _this.lazyLoad.remove(oldHref);
                });
            });
        };
        return LazyStyleHandler;
    }());
    LazyStyleHandler.prov = i0__namespace.defineInjectable({ factory: function LazyStyleHandler_Factory() { return new LazyStyleHandler(i0__namespace.inject(i0__namespace.INJECTOR)); }, token: LazyStyleHandler, providedIn: "root" });
    LazyStyleHandler.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    LazyStyleHandler.ctorParameters = function () { return [
        { type: i0.Injector }
    ]; };
    function createLazyStyleHref(style, dir) {
        return style.replace(/{{\s*dir\s*}}/g, dir);
    }
    function initLazyStyleHandler(injector) {
        return function () { return new LazyStyleHandler(injector); };
    }

    var NavItem = /** @class */ (function () {
        function NavItem(props) {
            props = Object.assign(Object.assign({}, props), { visible: props.visible || (function () { return true; }) });
            Object.assign(this, props);
        }
        return NavItem;
    }());

    var NG_BOOTSTRAP_CONFIG_PROVIDERS = [
        {
            provide: i0.APP_INITIALIZER,
            useFactory: configureNgBootstrap,
            deps: [ngBootstrap.NgbInputDatepickerConfig, ngBootstrap.NgbTypeaheadConfig],
            multi: true,
        },
    ];
    function configureNgBootstrap(datepicker, typeahead) {
        return function () {
            datepicker.container = 'body';
            typeahead.container = 'body';
        };
    }

    var THEME_SHARED_ROUTE_PROVIDERS = [
        { provide: i0.APP_INITIALIZER, useFactory: configureRoutes, deps: [i1.RoutesService], multi: true },
    ];
    function configureRoutes(routesService) {
        return function () {
            routesService.add([
                {
                    path: undefined,
                    name: "AbpUiNavigation::Menu:Administration" /* Administration */,
                    iconClass: 'fa fa-wrench',
                    order: 100,
                },
            ]);
        };
    }

    /**
     * @deprecated To be removed in v5.0
     */
    var ModalContainerComponent = /** @class */ (function () {
        function ModalContainerComponent() {
        }
        return ModalContainerComponent;
    }());
    ModalContainerComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'abp-modal-container',
                    template: '<ng-container #container></ng-container>'
                },] }
    ];
    ModalContainerComponent.propDecorators = {
        container: [{ type: i0.ViewChild, args: ['container', { static: true, read: i0.ViewContainerRef },] }]
    };

    /**
     * @deprecated Use ng-bootstrap modal. To be deleted in v5.0.
     */
    var ModalService = /** @class */ (function () {
        function ModalService(contentProjectionService) {
            this.contentProjectionService = contentProjectionService;
            this.setContainer();
        }
        ModalService.prototype.setContainer = function () {
            this.containerComponentRef = this.contentProjectionService.projectContent(i1.PROJECTION_STRATEGY.AppendComponentToBody(ModalContainerComponent));
            this.containerComponentRef.changeDetectorRef.detectChanges();
        };
        ModalService.prototype.clearModal = function () {
            this.getContainer().clear();
            this.detectChanges();
        };
        ModalService.prototype.detectChanges = function () {
            this.containerComponentRef.changeDetectorRef.detectChanges();
        };
        ModalService.prototype.getContainer = function () {
            return this.containerComponentRef.instance.container;
        };
        ModalService.prototype.renderTemplate = function (template, context) {
            var containerRef = this.getContainer();
            var strategy = i1.PROJECTION_STRATEGY.ProjectTemplateToContainer(template, containerRef, context);
            this.contentProjectionService.projectContent(strategy);
        };
        ModalService.prototype.ngOnDestroy = function () {
            this.containerComponentRef.destroy();
        };
        return ModalService;
    }());
    ModalService.prov = i0__namespace.defineInjectable({ factory: function ModalService_Factory() { return new ModalService(i0__namespace.inject(i1__namespace.ContentProjectionService)); }, token: ModalService, providedIn: "root" });
    ModalService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    ModalService.ctorParameters = function () { return [
        { type: i1.ContentProjectionService }
    ]; };

    var NavItemsService = /** @class */ (function () {
        function NavItemsService() {
            this._items$ = new rxjs.BehaviorSubject([]);
        }
        Object.defineProperty(NavItemsService.prototype, "items", {
            get: function () {
                return this._items$.value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(NavItemsService.prototype, "items$", {
            get: function () {
                return this._items$.asObservable();
            },
            enumerable: false,
            configurable: true
        });
        NavItemsService.prototype.addItems = function (newItems) {
            var items = __spread(this.items);
            newItems.forEach(function (item) { return items.push(new NavItem(item)); });
            items.sort(sortItems);
            this._items$.next(items);
        };
        NavItemsService.prototype.removeItem = function (id) {
            var index = this.items.findIndex(function (item) { return item.id === id; });
            if (index < 0)
                return;
            var items = __spread(this.items.slice(0, index), this.items.slice(index + 1));
            this._items$.next(items);
        };
        NavItemsService.prototype.patchItem = function (id, item) {
            var index = this.items.findIndex(function (i) { return i.id === id; });
            if (index < 0)
                return;
            var items = __spread(this.items);
            items[index] = new NavItem(Object.assign(Object.assign({}, items[index]), item));
            items.sort(sortItems);
            this._items$.next(items);
        };
        return NavItemsService;
    }());
    NavItemsService.prov = i0__namespace.defineInjectable({ factory: function NavItemsService_Factory() { return new NavItemsService(); }, token: NavItemsService, providedIn: "root" });
    NavItemsService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    function sortItems(a, b) {
        if (!a.order)
            return 1;
        if (!b.order)
            return -1;
        return a.order - b.order;
    }

    var PageAlertService = /** @class */ (function () {
        function PageAlertService() {
            this.alerts = new i1.InternalStore([]);
            this.alerts$ = this.alerts.sliceState(function (state) { return state; });
        }
        PageAlertService.prototype.show = function (alert) {
            var _a;
            var newAlert = Object.assign(Object.assign({}, alert), { dismissible: (_a = alert.dismissible) !== null && _a !== void 0 ? _a : true });
            this.alerts.set(__spread([newAlert], this.alerts.state));
        };
        PageAlertService.prototype.remove = function (index) {
            var alerts = __spread(this.alerts.state);
            alerts.splice(index, 1);
            this.alerts.set(alerts);
        };
        return PageAlertService;
    }());
    PageAlertService.prov = i0__namespace.defineInjectable({ factory: function PageAlertService_Factory() { return new PageAlertService(); }, token: PageAlertService, providedIn: "root" });
    PageAlertService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    PageAlertService.ctorParameters = function () { return []; };

    var THEME_SHARED_APPEND_CONTENT = new i0.InjectionToken('THEME_SHARED_APPEND_CONTENT', {
        providedIn: 'root',
        factory: function () {
            var domInsertion = i0.inject(i1.DomInsertionService);
            domInsertion.insertContent(i1.CONTENT_STRATEGY.AppendStyleToHead(styles));
            import('chart.js').then(function () { return chartJsLoaded$.next(true); });
        },
    });

    function isNumber(value) {
        return !isNaN(toInteger(value));
    }
    function toInteger(value) {
        return parseInt("" + value, 10);
    }
    var DateParserFormatter = /** @class */ (function (_super) {
        __extends(DateParserFormatter, _super);
        function DateParserFormatter(configState, locale) {
            var _this = _super.call(this) || this;
            _this.configState = configState;
            _this.locale = locale;
            return _this;
        }
        DateParserFormatter.prototype.parse = function (value) {
            if (value) {
                var dateParts = value.trim().split('-');
                if (dateParts.length === 1 && isNumber(dateParts[0])) {
                    return { year: toInteger(dateParts[0]), month: null, day: null };
                }
                else if (dateParts.length === 2 && isNumber(dateParts[0]) && isNumber(dateParts[1])) {
                    return { year: toInteger(dateParts[0]), month: toInteger(dateParts[1]), day: null };
                }
                else if (dateParts.length === 3 &&
                    isNumber(dateParts[0]) &&
                    isNumber(dateParts[1]) &&
                    isNumber(dateParts[2])) {
                    return {
                        year: toInteger(dateParts[0]),
                        month: toInteger(dateParts[1]),
                        day: toInteger(dateParts[2]),
                    };
                }
            }
            return null;
        };
        DateParserFormatter.prototype.format = function (date) {
            if (!date)
                return '';
            var localization = this.configState.getOne('localization');
            var dateFormat = snq__default['default'](function () { return localization.currentCulture.dateTimeFormat.shortDatePattern; }) || 'yyyy-MM-dd';
            return common.formatDate(new Date(date.year, date.month - 1, date.day), dateFormat, this.locale);
        };
        return DateParserFormatter;
    }(ngBootstrap.NgbDateParserFormatter));
    DateParserFormatter.decorators = [
        { type: i0.Injectable }
    ];
    DateParserFormatter.ctorParameters = function () { return [
        { type: i1.ConfigStateService },
        { type: String, decorators: [{ type: i0.Inject, args: [i0.LOCALE_ID,] }] }
    ]; };

    var declarationsWithExports = [
        BreadcrumbComponent,
        ButtonComponent,
        ChartComponent,
        ConfirmationComponent,
        LoaderBarComponent,
        LoadingComponent,
        ModalComponent,
        TableComponent,
        TableEmptyMessageComponent,
        ToastComponent,
        ToastContainerComponent,
        SortOrderIconComponent,
        NgxDatatableDefaultDirective,
        NgxDatatableListDirective,
        LoadingDirective,
        TableSortDirective,
        ModalCloseDirective,
    ];
    var BaseThemeSharedModule = /** @class */ (function () {
        function BaseThemeSharedModule() {
        }
        return BaseThemeSharedModule;
    }());
    BaseThemeSharedModule.decorators = [
        { type: i0.NgModule, args: [{
                    imports: [
                        i1.CoreModule,
                        ngxDatatable.NgxDatatableModule,
                        core.NgxValidateCoreModule,
                        ngBootstrap.NgbPaginationModule,
                        EllipsisModule,
                    ],
                    declarations: __spread(declarationsWithExports, [HttpErrorWrapperComponent, ModalContainerComponent]),
                    exports: __spread([ngxDatatable.NgxDatatableModule, EllipsisModule], declarationsWithExports),
                    providers: [common.DatePipe],
                    entryComponents: [
                        HttpErrorWrapperComponent,
                        LoadingComponent,
                        ModalContainerComponent,
                        ToastContainerComponent,
                        ConfirmationComponent,
                    ],
                },] }
    ];
    var ThemeSharedModule = /** @class */ (function () {
        function ThemeSharedModule() {
        }
        ThemeSharedModule.forRoot = function (_a) {
            var _b = _a === void 0 ? {} : _a, httpErrorConfig = _b.httpErrorConfig, _c = _b.validation, validation = _c === void 0 ? {} : _c;
            return {
                ngModule: ThemeSharedModule,
                providers: [
                    {
                        provide: i0.APP_INITIALIZER,
                        multi: true,
                        deps: [ErrorHandler],
                        useFactory: i1.noop,
                    },
                    THEME_SHARED_ROUTE_PROVIDERS,
                    {
                        provide: i0.APP_INITIALIZER,
                        multi: true,
                        deps: [THEME_SHARED_APPEND_CONTENT],
                        useFactory: i1.noop,
                    },
                    {
                        provide: i0.APP_INITIALIZER,
                        multi: true,
                        deps: [i0.Injector],
                        useFactory: initLazyStyleHandler,
                    },
                    { provide: HTTP_ERROR_CONFIG, useValue: httpErrorConfig },
                    {
                        provide: 'HTTP_ERROR_CONFIG',
                        useFactory: httpErrorConfigFactory,
                        deps: [HTTP_ERROR_CONFIG],
                    },
                    { provide: ngBootstrap.NgbDateParserFormatter, useClass: DateParserFormatter },
                    NG_BOOTSTRAP_CONFIG_PROVIDERS,
                    {
                        provide: core.VALIDATION_BLUEPRINTS,
                        useValue: Object.assign(Object.assign({}, DEFAULT_VALIDATION_BLUEPRINTS), (validation.blueprints || {})),
                    },
                    {
                        provide: core.VALIDATION_MAP_ERRORS_FN,
                        useValue: validation.mapErrorsFn || core.defaultMapErrorsFn,
                    },
                    {
                        provide: core.VALIDATION_VALIDATE_ON_SUBMIT,
                        useValue: validation.validateOnSubmit,
                    },
                ],
            };
        };
        return ThemeSharedModule;
    }());
    ThemeSharedModule.decorators = [
        { type: i0.NgModule, args: [{
                    imports: [BaseThemeSharedModule],
                    exports: [BaseThemeSharedModule],
                },] }
    ];

    var minLength = forms.Validators.minLength, maxLength = forms.Validators.maxLength;
    function getPasswordValidators(injector) {
        var getRule = getRuleFn(injector);
        var passwordRulesArr = [];
        var requiredLength = 1;
        if (getRule('RequireDigit') === 'true') {
            passwordRulesArr.push('number');
        }
        if (getRule('RequireLowercase') === 'true') {
            passwordRulesArr.push('small');
        }
        if (getRule('RequireUppercase') === 'true') {
            passwordRulesArr.push('capital');
        }
        if (getRule('RequireNonAlphanumeric') === 'true') {
            passwordRulesArr.push('special');
        }
        if (Number.isInteger(+getRule('RequiredLength'))) {
            requiredLength = +getRule('RequiredLength');
        }
        return [core.validatePassword(passwordRulesArr), minLength(requiredLength), maxLength(128)];
    }
    function getRuleFn(injector) {
        var configState = injector.get(i1.ConfigStateService);
        return function (key) {
            var passwordRules = configState.getSettings('Identity.Password');
            return (passwordRules["Abp.Identity.Password." + key] || '').toLowerCase();
        };
    }

    /*
     * Public API Surface of theme-shared
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.BOOTSTRAP = BOOTSTRAP;
    exports.BaseThemeSharedModule = BaseThemeSharedModule;
    exports.BreadcrumbComponent = BreadcrumbComponent;
    exports.ButtonComponent = ButtonComponent;
    exports.ChartComponent = ChartComponent;
    exports.ConfirmationComponent = ConfirmationComponent;
    exports.ConfirmationService = ConfirmationService;
    exports.DEFAULT_ERROR_LOCALIZATIONS = DEFAULT_ERROR_LOCALIZATIONS;
    exports.DEFAULT_ERROR_MESSAGES = DEFAULT_ERROR_MESSAGES;
    exports.DEFAULT_VALIDATION_BLUEPRINTS = DEFAULT_VALIDATION_BLUEPRINTS;
    exports.DateParserFormatter = DateParserFormatter;
    exports.EllipsisDirective = EllipsisDirective;
    exports.EllipsisModule = EllipsisModule;
    exports.ErrorHandler = ErrorHandler;
    exports.HTTP_ERROR_CONFIG = HTTP_ERROR_CONFIG;
    exports.HTTP_ERROR_HANDLER = HTTP_ERROR_HANDLER;
    exports.HttpErrorWrapperComponent = HttpErrorWrapperComponent;
    exports.LAZY_STYLES = LAZY_STYLES;
    exports.LazyStyleHandler = LazyStyleHandler;
    exports.LoaderBarComponent = LoaderBarComponent;
    exports.LoadingComponent = LoadingComponent;
    exports.LoadingDirective = LoadingDirective;
    exports.ModalCloseDirective = ModalCloseDirective;
    exports.ModalComponent = ModalComponent;
    exports.ModalRefService = ModalRefService;
    exports.ModalService = ModalService;
    exports.NGX_DATATABLE_MESSAGES = NGX_DATATABLE_MESSAGES;
    exports.NG_BOOTSTRAP_CONFIG_PROVIDERS = NG_BOOTSTRAP_CONFIG_PROVIDERS;
    exports.NavItem = NavItem;
    exports.NavItemsService = NavItemsService;
    exports.NgxDatatableDefaultDirective = NgxDatatableDefaultDirective;
    exports.NgxDatatableListDirective = NgxDatatableListDirective;
    exports.PageAlertService = PageAlertService;
    exports.SUPPRESS_UNSAVED_CHANGES_WARNING = SUPPRESS_UNSAVED_CHANGES_WARNING;
    exports.SortOrderIconComponent = SortOrderIconComponent;
    exports.THEME_SHARED_APPEND_CONTENT = THEME_SHARED_APPEND_CONTENT;
    exports.THEME_SHARED_ROUTE_PROVIDERS = THEME_SHARED_ROUTE_PROVIDERS;
    exports.TableComponent = TableComponent;
    exports.TableEmptyMessageComponent = TableEmptyMessageComponent;
    exports.TableSortDirective = TableSortDirective;
    exports.ThemeSharedModule = ThemeSharedModule;
    exports.ToastComponent = ToastComponent;
    exports.ToastContainerComponent = ToastContainerComponent;
    exports.ToasterService = ToasterService;
    exports.bounceIn = bounceIn;
    exports.chartJsLoaded$ = chartJsLoaded$;
    exports.collapse = collapse;
    exports.collapseLinearWithMargin = collapseLinearWithMargin;
    exports.collapseWithMargin = collapseWithMargin;
    exports.collapseX = collapseX;
    exports.collapseY = collapseY;
    exports.collapseYWithMargin = collapseYWithMargin;
    exports.configureNgBootstrap = configureNgBootstrap;
    exports.configureRoutes = configureRoutes;
    exports.createLazyStyleHref = createLazyStyleHref;
    exports.defaultNgxDatatableMessages = defaultNgxDatatableMessages;
    exports.dialogAnimation = dialogAnimation;
    exports.expandX = expandX;
    exports.expandY = expandY;
    exports.expandYWithMargin = expandYWithMargin;
    exports.fadeAnimation = fadeAnimation;
    exports.fadeIn = fadeIn;
    exports.fadeInDown = fadeInDown;
    exports.fadeInLeft = fadeInLeft;
    exports.fadeInRight = fadeInRight;
    exports.fadeInUp = fadeInUp;
    exports.fadeOut = fadeOut;
    exports.fadeOutDown = fadeOutDown;
    exports.fadeOutLeft = fadeOutLeft;
    exports.fadeOutRight = fadeOutRight;
    exports.fadeOutUp = fadeOutUp;
    exports.getPasswordValidators = getPasswordValidators;
    exports.getRandomBackgroundColor = getRandomBackgroundColor;
    exports.httpErrorConfigFactory = httpErrorConfigFactory;
    exports.initLazyStyleHandler = initLazyStyleHandler;
    exports.slideFromBottom = slideFromBottom;
    exports.toastInOut = toastInOut;
    exports.a = EllipsisDirective;
    exports.b = EllipsisModule;
    exports.ba = HttpErrorWrapperComponent;
    exports.bb = ModalContainerComponent;
    exports.bc = ErrorHandler;
    exports.be = THEME_SHARED_ROUTE_PROVIDERS;
    exports.bf = configureRoutes;
    exports.bg = THEME_SHARED_APPEND_CONTENT;
    exports.bh = initLazyStyleHandler;
    exports.bi = httpErrorConfigFactory;
    exports.bj = HTTP_ERROR_CONFIG;
    exports.bk = DateParserFormatter;
    exports.bl = NG_BOOTSTRAP_CONFIG_PROVIDERS;
    exports.c = BreadcrumbComponent;
    exports.d = ButtonComponent;
    exports.e = ChartComponent;
    exports.f = ConfirmationComponent;
    exports.g = LoaderBarComponent;
    exports.h = LoadingComponent;
    exports.i = ModalComponent;
    exports.j = SUPPRESS_UNSAVED_CHANGES_WARNING;
    exports.k = ConfirmationService;
    exports.l = ModalRefService;
    exports.m = TableComponent;
    exports.n = TableEmptyMessageComponent;
    exports.o = ToastComponent;
    exports.p = ToasterService;
    exports.q = ToastContainerComponent;
    exports.r = toastInOut;
    exports.s = SortOrderIconComponent;
    exports.t = NgxDatatableDefaultDirective;
    exports.u = NgxDatatableListDirective;
    exports.w = NGX_DATATABLE_MESSAGES;
    exports.x = LoadingDirective;
    exports.y = TableSortDirective;
    exports.z = ModalCloseDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
